---
title: "Amplicon_analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set work directory
```{r}
setwd('e:/thermokarst_gully/')
wd_16s <- file.path(getwd(),"data/16S/rdp")
# if (!dir.exists(wd_16s)) {
#   dir.create(wd_16s)
# }
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```
Loading packages
```{r}
library(phyloseq)
library(ape)
library(vegan)
library(Biostrings)
library(microbiome)
library(tidyverse)
```
Data input
```{r}
source("script/read_data_rdp.R")
```

## Alpha diversity

Determine the alpha diversity including **Observed**, **Chao1**, **Shannon** and **Simpson**.

```{r}
alpha_div <- estimate_richness(phylo_rare, measures = c("Observed", "Chao1", 'Shannon', 'Simpson'))
library(picante)
pd <- pd(t(otu), tree, include.root = F) # estimate the phylogenetic diversity
alpha_div <- cbind(metadata[, c('Gully_id', 'Group')], alpha_div, Faith = pd$PD,
                   Evenness = alpha_div$Shannon/log(alpha_div$Observed)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')))
```

Box plot for alpha index using ggplot2

```{r, fig.align='center', fig.width=7.5, fig.height=6.2}
library(ggpubr)
library(ggplot2)
main_theme = theme_linedraw() + 
  theme(panel.grid=element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 10),
        strip.background = element_rect(colour = 'grey', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 12),
        axis.ticks = element_line(color = "black", linewidth = 1),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour = 'black', size = 10),
        axis.text.x = element_text(colour = 'black', size = 10, angle = 45, hjust = 1),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        legend.key = element_blank(),
        legend.background = element_blank())
# box plot
my_comparisons <- list(c('Control', 'Collapsed'))
alpha_div_plot <- alpha_div %>% 
  select(c("Gully_id", "Group", "Chao1", "Evenness", "Shannon", "Faith")) %>%
  gather(diversity, values, -c("Gully_id", "Group")) %>%
  ggplot(aes(x = Group, y = values)) + 
  geom_boxplot(width = 0.5, aes(fill = Group), outlier.shape = NA) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.15))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.5,
                     size = 4.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = 'Group', y = NULL, fill= 'Group') +
  facet_grid(diversity ~ Gully_id, scales = "free", space = "free_x") +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  main_theme +
  theme(panel.spacing = unit(0, "lines"))
# ggsave(file.path(save.dir, "./figs/alpha/alpha_boxplot_evenness.pdf"), 
#        p, width = 89, height = 89, units = "mm")
alpha_div_plot
```

Linear mixed models test the effect of collapsed and gully_id

```{r}
library(lme4)
library(car)
fm1 <- lmer(Chao1 ~ Group + (1 | Gully_id), data = alpha_div)
summary(fm1)
car::Anova(fm1, type = 2)

fm2 <- lmer(Evenness ~ Group + (1 | Gully_id), data = alpha_div)
summary(fm2)
car::Anova(fm2, type = 2)

fm3 <- lmer(Shannon ~ Group + (1 | Gully_id), data = alpha_div)
summary(fm3)
car::Anova(fm3, type = 2)

fm4 <- lmer(Faith ~ Group + (1 | Gully_id), data = alpha_div)
summary(fm4)
car::Anova(fm4, type = 2)
```
## Plot
```{r, fig.align='center', fig.width=5, fig.height=6}
library(gghalves)
alpha_div %>% select(-c("Gully_id", "Observed", "se.chao1", "Simpson")) %>%
  gather(diversity, value, -c("Group")) %>% 
  # mutate(diversity = factor(diversity, levels(c("Chao1", "Evenness", "Shannon", "Faith")))) %>%
  ggplot(aes(Group, value, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 1, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 2.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.5,
                     size = 4.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~diversity, scales = "free_y", ncol = 2) +
  theme_bw() +
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(color = "black", size = 10),
        legend.position = "none",
        axis.ticks = element_line(color = "black", linewidth = 1))
```
## Beta diversity
```{r, fig.align='center', fig.width=6, fig.height=3}
library(vegan)
# determine the dissimilarity matrix based on the bray-curties distance
tax_dist <-vegdist(t(otu), "bray" )
# difference in taxonomic variance among Group
my_comparisons <- list(c('Control', 'Collapsed'))
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
sapply(vars, function(x) usedist::dist_subset(tax_dist, grep(x, metadata$Sample_name, value = T))) %>%
  data.frame() %>%  gather("tem_group", "distance") %>% 
  cbind(Gully_id = c(rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20)), 
        Group = c(rep(c('Control', 'Collapsed'), each = 10, times = 6))) %>%
  select(., -c('tem_group')) %>% 
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'))) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Group, y = distance)) + 
  geom_boxplot(width = 0.5, aes(fill = Group)) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.5,
                     size = 4.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = 'Group', y = 'Bray-Curties distance', fill = 'Group') +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_grid(~ Gully_id, scales = "free_x", space = "free_x") +
  main_theme
# ggsave(file.path(save.dir, "./figs/beta/tax_variance.pdf"),
#        beta_tax_plot, width = 89, height = 89, units = "mm")
```

```{r}
# permanova, ANOSIM and MRPP analysis
library(vegan)
adonis2(tax_dist ~ Group, data = metadata)
mrpp(tax_dist, metadata$Group, perm = 999)
anosim(tax_dist, metadata$Group, perm = 999)
```

### PCoA plot with bray-curties as distance
```{r, fig.align='center', fig.width=4, fig.height=4}
# taxa
ord.tax <-  cmdscale(tax_dist,  k = 2, eig = T, add = T)
pcoa_tax_plot <- data.frame(Group = metadata$Group, scores(ord.tax)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_color_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x = paste("PCoA1 (", format(100 * ord.tax$eig[1] / sum(ord.tax$eig), digits = 3), "%)", sep = ""),
       y = paste("PCoA2 (", format(100 * ord.tax$eig[2] / sum(ord.tax$eig), digits = 3), "%)", sep = "")) +
 theme_bw() +
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(color = "black", size = 10),
        legend.position = c(0.15, 0.1),
        legend.background = element_blank(),
        legend.key = element_blank(),
        axis.ticks = element_line(color = "black", linewidth = 1))
# ggsave(file.path(save.dir, "./figs/beta/PCoA_tax_bray.pdf"),
#        pcoa_tax_plot, width = 89, height = 59, units = "mm")
pcoa_tax_plot
```

### Difference in taxonomic variance among layers
```{r, fig.align='center', fig.width=6, fig.height=3}
my_comparisons <- list(c('Control', 'Collapsed'))
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
beta_tax_plot <- sapply(vars, function(x) usedist::dist_subset(tax_dist, grep(x, metadata$Sample_name, value = T))) %>%
  data.frame() %>%  gather("tem_group", "distance") %>% 
  cbind(Gully_id = c(rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20)), 
        Group = c(rep(c('Control', 'Collapsed'), each = 10, times = 6))) %>%
  select(., -c('tem_group')) %>% 
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'))) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Group, y = distance)) + 
  geom_boxplot(width = 0.5, aes(fill = Group)) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.5,
                     size = 4.5, tip.length = 0.00, method = "wilcox.test") +
  # scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  facet_grid(~ Gully_id, scales = "free_x", space = "free_x") +
  labs(x = 'Group', y = 'Bray-Curties distance', fill='Group') +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  main_theme
# ggsave(file.path(save.dir, "./figs/beta/tax_variance.pdf"),
#        beta_tax_plot, width = 89, height = 89, units = "mm")
beta_tax_plot
```

## Composition
### Difference in the taxonomic composition between group control and Collapsed.
```{r, fig.align='center', fig.width=8, fig.height=4}
# extract the taxa at phylum level
subphylo <- tax_glom(phylo, 'Phylum', NArm = F)
subphylo.rel  = transform_sample_counts(subphylo, function(x) x / sum(x))
ntaxa(subphylo.rel)
ra.tab <- otu_table(subphylo.rel)
sum(ra.tab[, 1])
subtaxa_tab <- tax_table(subphylo.rel)[, 2]

# extract the taxa at class level
subphylo_class <- tax_glom(phylo, 'Class', NArm = F)
subphylo.class.rel  = transform_sample_counts(subphylo_class, function(x) x / sum(x))
ntaxa(subphylo.class.rel)
ra.class.tab <- otu_table(subphylo.class.rel)
sum(ra.class.tab[, 1])
subtaxa_class_tab <- tax_table(subphylo.class.rel)[, 3]

#ecombine the class within proteobacteria and other phyla into one abundance table
otu_final_tab <- rbind(data.frame(subtaxa_class_tab, ra.class.tab) %>%
  filter(Class %in% c("Alphaproteobacteria", "Betaproteobacteria",
                      "Gammaproteobacteria", "Deltaproteobacteria")) %>%
  rename(Phylum = Class), data.frame(subtaxa_tab, ra.tab) %>%
  filter(!Phylum %in% c("Proteobacteria", "Unassigned")))

# boxplot shows the community composition
box_plot <- otu_final_tab %>%
  filter(!Phylum %in% c("Unassigned")) %>% 
  group_by(Phylum) %>% 
  summarise(across(everything(), sum)) %>% 
  mutate(MRA = rowMeans(select(., colnames(ra.tab)))) %>%
  arrange(desc(MRA)) %>% dplyr::top_n(20, MRA) %>%
  select(., -c('MRA')) %>% 
  mutate(Phylum = factor(Phylum, levels = Phylum)) %>%
  tidyr::pivot_longer(cols = -c(Phylum), names_to = "sample_id", values_to = 'rel_abun') %>%
  mutate(Group = sapply(stringr::str_split(sample_id, "_",  n = 2), `[`, 2)) %>%
  mutate(Group = sapply(stringr::str_split(Group, "",  n = 2), `[`, 1)) %>%
  mutate(Group = factor(Group, levels = c('C', 'T'))) %>%
  ggplot(aes(Phylum, rel_abun*100)) + 
  #erro bar
  stat_boxplot(aes(x = Phylum, y = rel_abun*100, fill = Group), 
               position = position_dodge(width = .5), 
               geom = 'errorbar', width = 0.25) +
  #boxplot
  geom_boxplot(outlier.shape = NA, width = 0.5, aes(fill = Group)) +
  stat_compare_means(aes(group = Group),  paired = F, 
                     p.adjust.method = "BH", label = "p.signif",
                     label.y.npc = 0.9) +
  scale_y_continuous(limits=c(0, 50)) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x = 'Phylum', y = 'Relative abundance (%)', fill = 'Group') +
  main_theme
# ggsave(file.path(save.dir, "./figs/composition/tax_composition_10.pdf"), box_plot, width = 189, height = 95, units = "mm")
box_plot
```

### Microbial composition in each group and treatments
```{r, fig.align='center', fig.width=8, fig.height=4}
bar_plot <- otu_final_tab %>% 
  mutate(MRA = rowMeans(select(., colnames(ra.tab)))) %>%
  arrange(desc(MRA)) %>% dplyr::top_n(11, MRA) %>%
  select(., -c('MRA')) %>% 
  bind_rows(summarise_all(., ~if(is.numeric(.)) 1-sum(.) else "Other")) %>%
  mutate(Phylum = factor(Phylum, levels = Phylum)) %>%
  pivot_longer(cols = -c(Phylum), names_to = "Sample_name", values_to = 'rel_abun') %>%
  right_join(metadata[, c('Sample_name', 'Gully_id', 'Group')], by = c("Sample_name")) %>%
  select(., -c('Sample_name')) %>% 
  group_by(Gully_id, Group, Phylum) %>%
  dplyr::summarise(across(, mean, na.rm = TRUE)) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'))) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Group, y = 100*rel_abun, fill = Phylum))+
  geom_bar(stat = "identity") +
  scale_fill_brewer(palette = "Set3")+
  scale_y_continuous(expand = c(0,0))+
  labs(x = 'Sample', y = 'Mean relative abundance (%)') +
  guides(fill = guide_legend(keywidth = 0.5, keyheight = 0.5)) +
  facet_grid(~ Gully_id, scales = "free_x", space = "free_x") + 
  main_theme
# ggsave(file.path(save.dir, "./figs/composition/tax_composition_bar.pdf"), bar_plot, width = 120, height = 65, units = "mm")
bar_plot
```

### determine the average relative abundance of each phylum within each group
```{r, fig.align='center', fig.width=4.5, fig.height=6.5}
phyla_tab <- otu_final_tab %>%
  remove_rownames() %>%
  column_to_rownames(var = "Phylum") %>%
  t()

vars <- colnames(phyla_tab)
phyla_tab <- data.frame(metadata[, c(3, 4)], phyla_tab)

# determine the mean, standard deviations, standard errors for each phylum
avg_abun_in_control <- phyla_tab %>% select(-c("Gully_id")) %>%
    group_by(Group) %>% 
    summarise(across(where(is.numeric), list(mean = ~mean(., na.rm = T), 
                                  sd = ~sd(., na.rm = T), 
                                  n = ~n()))) %>% 
    pivot_longer(cols = -c(Group), 
                 names_to = c('Phylum', '.value'), 
                 names_sep = '_') %>% 
    mutate(se = sd/sqrt(n)) %>%
  filter(Group == "Control") %>%
  arrange(mean) %>%
  mutate(Phylum = factor(Phylum, levels = Phylum))

avg_abun_in_control_plot <- avg_abun_in_control%>%
  ggplot(aes(x = Phylum, y = mean)) +
  geom_bar(position = 'dodge', stat = 'identity', 
           fill= '#FFF6E1', colour = 'black', width = 0.7) +
  geom_errorbar(aes(ymin = mean, ymax = mean + se), width=.2,  position = position_dodge(0.7)) +
  labs(x = NULL, y = 'Mean relative abundance (%)') +
  theme_bw() + coord_flip() + scale_y_reverse() +
  theme(legend.position = NULL,
        panel.grid = element_blank(), 
        axis.ticks.y = element_blank(),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_blank(),
        axis.text.x = element_text(colour = 'black', size = 12)
  )
```
### test the difference in abundance of main taxa using liner models
```{r}
# determine the effect size of the permafrost thawing on each phylum
phyla_tab_scale <- phyla_tab %>% 
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed")))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
phylaS1 <- sapply(3:ncol(phyla_tab_scale), function(j) {
    message("Now j=", j, " in ", ncol(phyla_tab_scale), ". ", date())
    if (length(unique(phyla_tab_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(phyla_tab_scale[, j] ~ Group + (1 | Gully_id), data = phyla_tab_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(phylaS1)<-colnames(phyla_tab_scale)[-c(1,2)]
data.frame(phylaS1)
```
### effect size plot
```{r, fig.align='center', fig.width=4.5, fig.height=6.5}
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
eff_siz_tax_plot <- phylaS1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "Phylum") %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(Phylum = factor(Phylum, levels = avg_abun_in_control$Phylum)) %>%
  ggplot(aes(x = Phylum, y = GroupCollapsed.mean, color = Phylum)) +
  geom_hline(aes(yintercept =0), size=0.7,  colour="gray2")+
  geom_point(size = 4) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = Phylum, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size")+
  # scale_color_manual(values=c("#D55E00","#446DA9","#446DA9","#446DA9","#446DA9","#D55E00","#D55E00")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0, xmax = 24.5, ymin = -2, ymax = 2, alpha = 0.2,fill="#FAE3AD") +
  theme(strip.background = element_rect(fill=c("#FFF6E1")), 
        strip.text = element_text(size = 12,face = 'bold',colour = "#B8733A"),
        axis.text = element_text(colour = 'black', size = 12),
        panel.grid = element_blank(),
        legend.position = 'none')
```

Combine plots
```{r, fig.align='center', fig.width=6.5, fig.height=6.5}
library(cowplot)
# ggdraw() +
#   draw_plot(eff_siz_tax_plot, x = 0, y = 0, width = 0.65, height = 1) +
#   draw_plot(avg_abun_in_control_plot, x = 0.65, y = 0, width = 0.35, height = 1)
library(aplot) # Decorate a 'ggplot' with Associated Information
avg_abun_in_control_plot %>% insert_right(eff_siz_tax_plot, width = 2)
```

##  Network analysis
```{r}
# loading packages
pacman::p_load(microeco, file2meco, igraph, WGCNA, rgexf)
# construct the phyloseq projects for control and collapsed groups
ctrl_phylo <- subset_samples(phylo, Group == "Control")
ctrl_phylo <- prune_taxa(taxa_sums(ctrl_phylo) > 0, ctrl_phylo) 
collap_phylo <- subset_samples(phylo, Group == "Collapsed")
collap_phylo <- prune_taxa(taxa_sums(collap_phylo) > 0, collap_phylo)

# transform the phyloseq objects to microeco objects
meco_ctrl_obj <- phyloseq2meco(ctrl_phylo)
meco_collap_obj <- phyloseq2meco(collap_phylo)

# determine the adjacency matrixes
t1 <- trans_network$new(dataset = meco_ctrl_obj, cal_cor = "WGCNA", taxa_level = "OTU", filter_thres = 0.0001, cor_method = "spearman")
t2 <- trans_network$new(dataset = meco_collap_obj, cal_cor = "WGCNA", taxa_level = "OTU", filter_thres = 0.0001, cor_method = "spearman")

# use arbitrary coefficient threshold to contruct network
t1$cal_network(COR_p_thres = 0.01, COR_cut = 0.7)
t2$cal_network(COR_p_thres = 0.01, COR_cut = 0.7)

# invoke igraph cluster_fast_greedy function for this undirected network 
t1$cal_module(method = "cluster_fast_greedy")
t2$cal_module(method = "cluster_fast_greedy")

# require rgexf package to be installed
t1$save_network(filepath = file.path(save.dir, "./figs/ctrl_network.gexf"))
t2$save_network(filepath = file.path(save.dir, "./figs/collapsed_network.gexf"))

# calculate network attributes
t1$cal_network_attr()
t2$res_network_attr()
```


```{r, fig.align='center', fig.width=6.5, fig.height=4.5}
# get node properties
t1$get_node_table(node_roles = TRUE)
t2$get_node_table(node_roles = TRUE)
# return t1$res_node_table

# get edge properties
t1$get_edge_table()
t2$get_edge_table()
# return t1$res_edge_table 
t1$get_adjacency_matrix()
t2$get_adjacency_matrix()
# return t1$res_adjacency_matrix

# add_label = TRUE can be used to directly add text label for points
t1$plot_taxa_roles(use_type = 1)
t2$plot_taxa_roles(use_type = 1)
```

```{r, fig.align='center', fig.width=6.5, fig.height=5}
# plot node roles with phylum information
t1$plot_taxa_roles(use_type = 2)
t2$plot_taxa_roles(use_type = 2)
```
### Network robustness test
```{r}
# Network attack analysis
# Libraries
pacman::p_load(SpiecEasi, reshape2, phyloseq, igraph)
# get the igraph objects
ig.ctrl <- t1$res_network
ig.collapsed <- t2$res_network
# calculating the natural connectivity from adjacency matrix
ncc <- function(ig) {
  evals <- eigen(ig)$value
  nc <- log(mean(exp(evals)))
}

# calculating the natural connectivity from adjacency matrix of a graph
natcon <- function(ig) {
  adj <- get.adjacency(ig)
  evals <- eigen(adj)$value
  nc <- log(mean(exp(evals)))
}

# targeted attack ordered by betweenness
nc.attackbetweenness <- function(ig) {
  hubord <- order(rank(betweenness(ig)), decreasing=TRUE)
  sapply(1:round(vcount(ig)*.8), function(i) {
    ind <- hubord[1:i]
    tmp <- delete_vertices(ig, V(ig)$name[ind])
    natcon(tmp)
  })
}

# Targeted attack ordered by node degree.
nc.attackdegree <- function(ig) {
  hubord <- order(rank(degree(ig)), decreasing=TRUE)
  sapply(1:round(vcount(ig)*.8), function(i) {
    ind <- hubord[1:i]
    tmp <- delete_vertices(ig, V(ig)$name[ind])
    natcon(tmp)
  })
}


# Node removals
attack<-function (adj.mat, node.sup)
{
  n.nodes <- dim(adj.mat)[1]
  adj.mat[node.sup, ] <- rep(0, n.nodes)
  adj.mat[, node.sup] <- rep(0, n.nodes)
  nc<-ncc(adj.mat)
  list(new.mat = adj.mat, nc=nc)
}


ctrl.nc.deg.rmt <- nc.attackdegree(ig.ctrl)
ctrl.nc.bet.rmt <- nc.attackbetweenness(ig.ctrl)

collapsed.nc.deg.rmt <- nc.attackdegree(ig.collapsed)
collapsed.nc.bet.rmt <- nc.attackbetweenness(ig.collapsed)

ctrl.nc.deg.rmt <- cbind(rm_pro = c(1:length(ctrl.nc.deg.rmt))/vcount(ig.ctrl), nat.connet = ctrl.nc.deg.rmt)
ctrl.nc.bet.rmt <- cbind(rm_pro = c(1:length(ctrl.nc.bet.rmt))/vcount(ig.ctrl), nat.connet = ctrl.nc.bet.rmt)

collapsed.nc.deg.rmt <- cbind(rm_pro = c(1:length(collapsed.nc.deg.rmt))/vcount(ig.collapsed), nat.connet = collapsed.nc.deg.rmt)
collapsed.nc.bet.rmt <- cbind(rm_pro = c(1:length(collapsed.nc.bet.rmt))/vcount(ig.collapsed), nat.connet = collapsed.nc.bet.rmt)


write.table(ctrl.nc.deg.rmt, file.path(save.dir, "./tables/ctrl-deg-att-rmt.txt"),
            sep="\t", quote=F, row.names = T, col.names = T)
write.table(ctrl.nc.bet.rmt, file.path(save.dir, "./tables/ctrl-bet-att-rmt.txt"),
            sep="\t", quote=F, row.names = T, col.names = T)
write.table(collapsed.nc.deg.rmt, file.path(save.dir, "./tables/collapsed-deg-att-rmt.txt"), 
            sep="\t", quote=F, row.names = T, col.names = T)
write.table(collapsed.nc.bet.rmt, file.path(save.dir, "./tables/collapsed-bet-att-rmt.txt"),
            sep="\t", quote=F, row.names = T, col.names = T)
```

```{r, fig.align='center', fig.width=6, fig.height=3.5}
robut.df.rmt <- data.frame(rbind(cbind(Group = rep('Control', sum(nrow(ctrl.nc.deg.rmt), nrow(ctrl.nc.bet.rmt))),
                                       type = c(rep('Proportion of removed nodes', nrow(ctrl.nc.deg.rmt)), 
                                                rep('Proportion of removed betweenness', nrow(ctrl.nc.bet.rmt))),
                                       nat.connet = rbind(ctrl.nc.deg.rmt, ctrl.nc.bet.rmt)),
                                 cbind(Group = rep('Collapsed', sum(nrow(collapsed.nc.deg.rmt), nrow(collapsed.nc.bet.rmt))),
                                       type = c(rep('Proportion of removed nodes', nrow(collapsed.nc.deg.rmt)), 
                                                rep('Proportion of removed betweenness', nrow(collapsed.nc.bet.rmt))),
                                       nat.connet = rbind(collapsed.nc.deg.rmt, collapsed.nc.bet.rmt))))
robut.df.rmt$Group <- factor(robut.df.rmt$Group, ordered = T, 
                              levels = c('Control', 'Collapsed'))
robut.df.rmt$type <- factor(robut.df.rmt$type, ordered = T, 
                            levels = c('Proportion of removed nodes', 'Proportion of removed betweenness'))

robut.df.rmt$rm_pro <- as.numeric(robut.df.rmt$rm_pro)
robut.df.rmt$nat.connet <- as.numeric(robut.df.rmt$nat.connet)

robutness_test_plot_rmt <- ggplot(robut.df.rmt, aes(x = rm_pro, y = nat.connet, colour = Group))+
  geom_point() +
  scale_color_manual(values= c('#1b9e77', '#d95f02')) +
  labs(x = NULL, y = 'Natural connectivity') +
  facet_wrap(~type, scales = 'free') +
  theme_bw() +
  theme(strip.text = element_text(size = 10),
        axis.line = element_line(colour = "black"),
        axis.title = element_text(size = 12, colour = 'black'),
        axis.text = element_text(size = 10, colour = 'black'),
        axis.text.x = element_text(colour='black', size = 10, 
                                   angle = 45, hjust = 1),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 10),
        legend.position = c(0.9, 0.8),
        legend.background = element_blank(),
        panel.grid = element_blank())
robutness_test_plot_rmt
```
