---
title: "Composition and variations analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set working and saving directory
```{r, message=FALSE}
setwd('e:/thermokarst_gully/')
save.dir <- file.path(getwd(),"result")
```
Loading packages and reading data
```{r}
pacman::p_load(phyloseq, ape, vegan, Biostrings, microbiome, tidyverse, rstatix)
source("script/read_data_all.R")
```

Composition
```{r}
# Extrac the OTU table with the relative abundance higher than 1%
subextract_fun <- function(phylo, taxa_level, perc) {
  subphylo <- tax_glom(phylo, taxa_level, NArm = F)
  subphylo.rel  = transform_sample_counts(subphylo, function(x) x / sum(x))
  ra.tab <- otu_table(subphylo.rel)
  subtaxa_tab <- tax_table(subphylo.rel)[, taxa_level]
  otu_final_tab <- data.frame(subtaxa_tab, ra.tab) %>%
    mutate(MRA = rowMeans(select(., colnames(ra.tab)))) %>%
    filter(MRA >= perc) %>%
    arrange(desc(MRA)) %>%
    select(., -c('MRA')) # %>%
    # bind_rows(summarise_all(., ~if(is.numeric(.)) 1-sum(.) else "Others"))
  return(otu_final_tab)
}

otu_rel_abun_1perc_phyla_16s <- subextract_fun(phylo_16s, "Phylum", 0.01)
# extract the taxa at class level
otu_rel_abun_1perc_class_16s <- subextract_fun(phylo_16s, "Class", 0.01)

#ecombine the class within proteobacteria and other phyla into one abundance table
otu_rel_abun_1perc_16s <- rbind(otu_rel_abun_1perc_class_16s %>%
                         filter(Class %in% c("Alphaproteobacteria", "Betaproteobacteria",
                                             "Gammaproteobacteria", "Deltaproteobacteria")) %>%
                         dplyr::rename(Phylum = Class), otu_rel_abun_1perc_phyla_16s %>%
                         filter(!Phylum %in% c("Proteobacteria", "Unassigned")))

# extrac the Fungal phyla with the relative abundance higher than 1%
otu_rel_abun_1perc_its <- subextract_fun(phylo_its, "Phylum", 0.01) %>%
  filter(!Phylum %in% c("Unassigned"))

# extrac the Protistan phyla with the relative abundance higher than 1%
otu_rel_abun_1perc_pro <- subextract_fun(phylo_protist, "Phylum", 0.01) %>%
  filter(!Phylum %in% c("Unassigned"))

# extrac the Animal phyla with the relative abundance higher than 1%
otu_rel_abun_1perc_anim <- subextract_fun(phylo_animal, "Class", 0.01) %>%
  filter(!Class %in% c("Unassigned"))
```

Compute summary statistics for the relative abundance of all taxa.
```{r}
## arrange the data for the lmm test
otu_rel_abun_1perc_all <- bind_rows(otu_rel_abun_1perc_16s, otu_rel_abun_1perc_its,
                                otu_rel_abun_1perc_pro, 
                                otu_rel_abun_1perc_anim %>%  dplyr::rename(Phylum = Class)) %>%
  as.tibble() %>% column_to_rownames("Phylum") %>% t() %>% 
  as.data.frame() %>% rownames_to_column(var = "Sample_id") %>%
  mutate(Group = c(case_when(grepl("_C", Sample_id) ~ "Uncollapsed",
                             grepl("_T", Sample_id) ~ "Collapsed")),
         Gully_id = c(case_when(grepl("G1", Sample_id) ~ "G1",
                                grepl("G2", Sample_id) ~ "G2",
                                grepl("G3", Sample_id) ~ "G3",
                                grepl("G4", Sample_id) ~ "G4",
                                grepl("G5", Sample_id) ~ "G5",
                                grepl("G6", Sample_id) ~ "G6")))

tax_name <- colnames(otu_rel_abun_1perc_all)[-c(1, 29, 30)] #remove the "Sample_id", "Group", "Gully_id"
tax_stats <- otu_rel_abun_1perc_all %>% group_by(Group) %>%
  get_summary_stats(tax_name , type = "common") %>% 
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  arrange(variable, Group) %>%
  select(Group, variable, n, mean, sd, se)
tax_stats
```
Test the difference of each dominant taxa between the uncollapsed and collapsed sample using lmm analysis.
```{r}
## lmm test
library(lme4)
library(lmerTest)
lmm_tax_modes <- lapply(tax_name, function(x) {
  lmer(substitute(i ~ Group + (1|Gully_id), list(i = as.name(x))), data = otu_rel_abun_1perc_all)})
summary.model <- function(model){
  F.value <- anova(model)$'F value'
  p.value <- anova(model)$'Pr(>F)'
  p.stars <- function(p.values) {
    unclass(symnum(p.values, corr = FALSE, 
                   na = FALSE, cutpoints = c(0,0.001, 0.01, 0.05, 0.1, 1),
                   symbols = c("***", "**", "*", ".", " ")))}
  sig <- p.stars(p.value)
  results<-data.frame(F.value, p.value, sig)
  return(results)
}
df <- NULL
for(i in 1:length(tax_name)) {
  tmp <- summary.model(lmm_tax_modes[[i]])
  if (is.null(df)){
    df <- tmp
  } else {
    df <- rbind(df, tmp)
  }
}

tax_result_lmm <-data.frame(Variables = tax_name, group1 = rep("Uncollapsed", length(tax_name)),
                            group2 = rep("Collapsed", length(tax_name)), df)
tax_result_lmm
```


Arrange the data for sankey plot 
```{r}
# Combine all data
compo_df <- rbind(otu_rel_abun_1perc_16s %>%
  pivot_longer(-Phylum, names_to = "Sample_id", values_to = "Rela_abun") %>%
    mutate(Group = rep("Bacteria", nrow(.))),
  otu_rel_abun_1perc_its %>%
    pivot_longer(-Phylum, names_to = "Sample_id", values_to = "Rela_abun") %>%
    mutate(Group = rep("Fungi", nrow(.))),
  otu_rel_abun_1perc_pro %>%
    pivot_longer(-Phylum, names_to = "Sample_id", values_to = "Rela_abun") %>%
    mutate(Group = rep("Protist", nrow(.))),
  otu_rel_abun_1perc_anim %>%
    pivot_longer(-Class, names_to = "Sample_id", values_to = "Rela_abun") %>%
    dplyr::rename(c(Phylum = Class)) %>% 
    mutate(Group = rep("Animal", nrow(.)))) %>%
  dplyr::rename(c(Taxa = Phylum)) %>%
  mutate(Type = ifelse(grepl("_C", Sample_id), "Uncollapsed",
                       ifelse(grepl("_T", Sample_id), "Collapsed", ""))) %>%
  select(-Sample_id) %>%
  group_by(Type, Group, Taxa) %>%
  summarise(across(everything(), sum)) %>%
  mutate(Type_group = case_when(Type == "Uncollapsed" ~ "a",
                                Type == "Collapsed" ~ "b"))
# prepare the table for sankey plot
sankey <- data.table::rbindlist(list(compo_df[c("Type", "Group", "Rela_abun")],
                                     compo_df[c("Group", "Taxa", "Rela_abun")]))

names(sankey) <- c('source', 'target', 'value')
knitr::kable(sankey[1:5, ])

# Make a connection data frame
links <- sankey 
# From these flows we need to create a node data frame: it lists every entities involved in the flow
nodes <- data.frame(
  name = c("Uncollapsed", "Collapsed", "Bacteria", "Fungi", "Protist", "Animal", 
           otu_rel_abun_1perc_16s$Phylum,
           otu_rel_abun_1perc_its$Phylum, otu_rel_abun_1perc_pro$Phylum,
           otu_rel_abun_1perc_anim$Class)
)

# With networkD3, connection must be provided using id, not using real name like in the links dataframe.. So we need to reformat it.
links$IDsource <- match(links$source, nodes$name)-1 
links$IDtarget <- match(links$target, nodes$name)-1

# Set color of connections
# Add a 'group' column to each connection:
links_datset <- mutate(links,
                       group = rep(c(rep("a", 27), rep("b", 27)), 2)
                       )
# Add a 'group' column to each node. Here I decide to put all of them in the same group to make them grey
nodes$group <- as.factor(nodes$name)
links_datset[1:5, ]
```
Sanke plot
```{r}
library(networkD3)
# Give a color for each group
my_colors = c("#e95f5c", "#79ceb8", "#1F77B4", "#AEC7E8", "#FF7F0E", "#FFBB78", 
              "#2CA02C", "#98DF8A", "#D62728", "#FF9896", "#9467BD", "#C5B0D5", 
              "#8C564B", "#C49C94", "#E377C2", "#F7B6D2", "#7F7F7F", "#C7C7C7", 
              "#BCBD22", "#DBDB8D", "#17BECF", "#9EDAE5", "#1F77B4", "#AEC7E8", 
              "#FF7F0E", "#FFBB78", "#2CA02C", "#98DF8A", "#D62728", "#FF9896", 
              "#9467BD", "#C5B0D5", "#9457eb", "#f56991", "#32cd32")

my_colors <- paste(my_colors, collapse = '", "')
colorJS <- paste('d3.scaleOrdinal(["', my_colors, '"])')


# my_color <- 'd3.scaleOrdinal() .domain(["a", "b", "my_unique_group"]) .range(["#79ceb8", "#e95f5c", "#c0c0c0"])'

# Make the Network. I call my colour scale with the colourScale argument
sankey_plot <- sankeyNetwork(Links = links_datset, Nodes = nodes, 
                   Source = "IDsource", Target = "IDtarget", 
                   Value = "value", NodeID = "name", 
                   colourScale = colorJS,
                   LinkGroup = "group", NodeGroup = "group",
                   iterations = 0, 
                   fontFamily = 'Arial', fontSize = 10, 
                   nodeWidth = 10, nodePadding = 10, height = 600, width = 300,
                   sinksRight = F)
# save the widget
# recommend htmlwidgets
# library(htmlwidgets)
# saveWidget(sankey_plot, file = paste0(save.dir, "/figs/composition/composition_sankey_plot.html"))
# library(webshot)
# install phantom:
# webshot::install_phantomjs()
# Make a webshot in pdf : high quality but can not choose printed zone
# webshot(paste0(sankey.plots.folder, "/sankey_plot.html"), delay = 0.2)
sankey_plot
```

Unique otus profile among three layers
```{r, fig.width = 4, fig.height = 6.5}
# library
library(ggvenn)
# Make the venn plot for 16S
x_16s = list(
  Uncollapsed = otu_table(phylo_16s) %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('_C', metadata$Sample_name, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames(),
  Collapsed = otu_table(phylo_16s) %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('_T', metadata$Sample_name, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames()
  )

p1 <- ggvenn(
  x_16s, 
  fill_color = c("#79ceb8", "#e95f5c"),
  stroke_size = ,
  stroke_color = NA,
  set_name_size = 3,
  text_size = 2.5,
  show_percentage = T,
  auto_scale = F
)
# Make the venn plot for ITS
x_its = list(
  Uncollapsed = otu_table(phylo_its) %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('_C', metadata$Sample_name, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames(),
  Collapsed = otu_table(phylo_its) %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('_T', metadata$Sample_name, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames()
  )

p2 <- ggvenn(
  x_its, 
  fill_color = c("#79ceb8", "#e95f5c"),
  stroke_color = NA,
  set_name_size = 3,
  text_size = 2.5,
  show_percentage = T,
  auto_scale = F
)
# Make the venn plot for protist
x_pro = list(
  Uncollapsed = otu_table(phylo_protist) %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('_C', metadata$Sample_name, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames(),
  Collapsed = otu_table(phylo_protist) %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('_T', metadata$Sample_name, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames()
  )

p3 <- ggvenn(
  x_pro, 
  fill_color = c("#79ceb8", "#e95f5c"),
  stroke_color = NA,
  set_name_size = 3,
  text_size = 2.5,
  show_percentage = T,
  auto_scale = F
)
# Make the venn plot for soil animal
x_anim = list(
  Uncollapsed = otu_table(phylo_animal) %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('_C', metadata$Sample_name, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames(),
  Collapsed = otu_table(phylo_animal) %>% data.frame() %>%
      mutate(rowsum = rowSums(select(., grep('_T', metadata$Sample_name, value = T)))) %>%
      filter(rowsum > 0) %>%
      rownames()
  )

p4 <- ggvenn(
  x_anim, 
  fill_color = c("#79ceb8", "#e95f5c"),
  stroke_color = NA,
  set_name_size = 3,
  text_size = 2.5,
  show_percentage = T,
  auto_scale = F
)
library(cowplot)
venn_plots <- plot_grid(p1, p2, p3, p4, ncol = 1, align = "v")
venn_plots
```



## Test the overall difference in the taxonomic and functional composition.
```{r}
library(vegan)
#determine the dissimilarity matrix based on the bray-curties distance
tax_16s_dist <-vegdist(t(otu_16s), "bray")
tax_its_dist <-vegdist(t(otu_its), "bray")
tax_pro_dist <-vegdist(t(otu_pro), "bray")
tax_anim_dist <-vegdist(t(otu_anim), "bray")
fun_dist <-vegdist(t(ko_count_table), "bray")
#permanova test the difference in compositional variance
adonis2(tax_16s_dist ~ Group, data = metadata)
adonis2(tax_its_dist ~ Group, data = metadata)
adonis2(tax_pro_dist ~ Group, data = metadata)
adonis2(tax_anim_dist ~ Group, data = metadata)
adonis2(fun_dist ~ Group, data = metadata)
```


Visualization for the overall difference by PCoA plot.
```{r, fig.align='center', fig.width=8.9, fig.height=2.5}
#16S
ord.16s <-  cmdscale(tax_16s_dist,  k = 2, eig = T, add = T)
pcoa_16s_plot <- data.frame(Group = metadata$Group, scores(ord.16s)) %>%
  mutate(Group = factor(Group, levels = c('Uncollapsed', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.16s$eig[1] / sum(ord.16s$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.16s$eig[2] / sum(ord.16s$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none", 
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
#ITS
ord.its <-  cmdscale(tax_its_dist,  k = 2, eig = T, add = T)
pcoa_its_plot <- data.frame(Group = metadata$Group, scores(ord.its)) %>%
  mutate(Group = factor(Group, levels = c('Uncollapsed', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.its$eig[1] / sum(ord.its$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.its$eig[2] / sum(ord.its$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none", 
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
#protist
ord.pro <-  cmdscale(tax_pro_dist,  k = 2, eig = T, add = T)
pcoa_pro_plot <- data.frame(Group = metadata$Group, scores(ord.pro)) %>%
  mutate(Group = factor(Group, levels = c('Uncollapsed', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.pro$eig[1] / sum(ord.pro$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.pro$eig[2] / sum(ord.pro$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none", 
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
#Animal
ord.anim <-  cmdscale(tax_anim_dist,  k = 2, eig = T, add = T)
pcoa_anim_plot <- data.frame(Group = metadata$Group, scores(ord.anim)) %>%
  mutate(Group = factor(Group, levels = c('Uncollapsed', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.anim$eig[1] / sum(ord.anim$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.anim$eig[2] / sum(ord.anim$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none", 
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
#Function
ord.fun <-  cmdscale(fun_dist,  k = 2, eig = T, add = T)
pcoa_fun_plot <- data.frame(Group = metadata$Group, scores(ord.fun)) %>%
  mutate(Group = factor(Group, levels = c('Uncollapsed', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.fun$eig[1] / sum(ord.fun$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.fun$eig[2] / sum(ord.fun$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none",
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
cowplot::plot_grid(pcoa_16s_plot, pcoa_its_plot, pcoa_pro_plot, pcoa_anim_plot, pcoa_fun_plot, ncol = 5)
```

## Explore the biotic homogenization
```{r, message=FALSE}
# difference in taxonomic variance among Group
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
# Assuming vars is defined somewhere earlier in your code
similar_16s_data <- lapply(vars, function(x) usedist::dist_subset(tax_16s_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_its_data <- lapply(vars, function(x) usedist::dist_subset(tax_its_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_pro_data <- lapply(vars, function(x) usedist::dist_subset(tax_pro_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_anim_data <- lapply(vars, function(x) usedist::dist_subset(tax_anim_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))
similar_data <- data.frame(Group = similar_16s_data$Group,
                            Gully_id = similar_16s_data$Gully_id, 
                            distance_16s = similar_16s_data$distance,
                            distance_its = similar_its_data$distance,
                            distance_pro = similar_pro_data$distance, 
                            distance_anima = similar_anim_data$distance)
```

## Identify the generalists, specialists, and opportunists
```{r}
#determine the generalist and specialist species
library(EcolUtils)
spec_gen_16s <- spec.gen(phylo_16s@otu_table %>% t(), niche.width.method = "levins", perm.method = "quasiswap",
                         n = 1000, probs = c(0.05, 0.95))
# write.table(spec_gen_16s, file.path(save.dir, "/tables/spe_gen/spec_gen_16s_95.csv"), sep = ",")
spec_gen_its <- spec.gen(phylo_its@otu_table %>% t(), niche.width.method = "levins", perm.method = "quasiswap",
                         n = 1000, probs = c(0.05, 0.95))
# write.table(spec_gen_its, file.path(save.dir, "/tables/spe_gen/spec_gen_its_95.csv"), sep = ",")
spec_gen_pro <- spec.gen(phylo_protist@otu_table %>% t(), niche.width.method = "levins", perm.method = "quasiswap",
                         n = 1000, probs = c(0.05, 0.95))
# write.table(spec_gen_pro, file.path(save.dir, "/tables/spe_gen/spec_pro_95.csv"), sep = ",")
spec_gen_animal <- spec.gen(phylo_animal@otu_table %>% t(), niche.width.method = "levins", perm.method = "quasiswap",
                            n = 1000, probs = c(0.05, 0.95))
# write.table(spec_gen_animal, file.path(save.dir, "/tables/spe_gen/spec_animal_95.csv"), sep = ",")
```

## Explore the effect size of permafrost collapse on the microbial structure
```{r}
#Pull the ids of the generalists, specialists, and opportunists from bacterial community
spec_gen_16s <- read.table(file.path(save.dir, "/tables/spe_gen/spec_gen_16s_95.csv"), sep = ",", header = T)
spec_id_16s <- spec_gen_16s %>% filter(sign == "SPECIALIST") %>% pull(OTU)
gen_id_16s <- spec_gen_16s %>% filter(sign == "GENERALIST") %>% pull(OTU)
neu_id_16s <- spec_gen_16s %>% filter(sign == "NON SIGNIFICANT") %>% pull(OTU)

spec_16s_comm <- phylo_16s@otu_table[rownames(phylo_16s@otu_table) %in% spec_id_16s, ]
gen_16s_comm <- phylo_16s@otu_table[rownames(phylo_16s@otu_table) %in% gen_id_16s, ]
neu_16s_comm <- phylo_16s@otu_table[rownames(phylo_16s@otu_table) %in% neu_id_16s, ]

library(vegan)
#determine the dissimilarity matrix based on the bray-curties distance
tax_spec_16s_dist <-vegdist(t(spec_16s_comm), "bray")
tax_gen_16s_dist <-vegdist(t(gen_16s_comm), "bray")
tax_neu_16s_dist <-vegdist(t(neu_16s_comm), "bray")

# difference in taxonomic variance among Group
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
similar_spec_16s_data <- lapply(vars, function(x) usedist::dist_subset(tax_spec_16s_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_gen_16s_data <- lapply(vars, function(x) usedist::dist_subset(tax_gen_16s_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_neu_16s_data <- lapply(vars, function(x) usedist::dist_subset(tax_neu_16s_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))
#Combine all group datasets
similar_16s_data <- data.frame(Group = similar_spec_16s_data$Group,
                           Gully_id = similar_spec_16s_data$Gully_id, 
                           distance_16s_spec = similar_spec_16s_data$distance,
                           distance_16s_gen = similar_gen_16s_data$distance,
                           distance_16s_neu = similar_neu_16s_data$distance)
```

## Fungi
```{r}
#explore the biotic homogenelization and heteronazition
spec_gen_its <- read.table(file.path(save.dir, "/tables/spe_gen/spec_gen_its_95.csv"), sep = ",", header = T)
spec_id_its <- spec_gen_its %>% filter(sign == "SPECIALIST") %>% pull(OTU)
gen_id_its <- spec_gen_its %>% filter(sign == "GENERALIST") %>% pull(OTU)
neu_id_its <- spec_gen_its %>% filter(sign == "NON SIGNIFICANT") %>% pull(OTU)

spec_its_comm <- phylo_its@otu_table[rownames(phylo_its@otu_table) %in% spec_id_its, ]
gen_its_comm <- phylo_its@otu_table[rownames(phylo_its@otu_table) %in% gen_id_its, ]
neu_its_comm <- phylo_its@otu_table[rownames(phylo_its@otu_table) %in% neu_id_its, ]

library(vegan)
#determine the dissimilarity matrix based on the bray-curties distance
tax_spec_its_dist <-vegdist(t(spec_its_comm), "bray")
tax_gen_its_dist <-vegdist(t(gen_its_comm), "bray")
tax_neu_its_dist <-vegdist(t(neu_its_comm), "bray")

# difference in taxonomic variance among Group
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
similar_spec_its_data <- lapply(vars, function(x) usedist::dist_subset(tax_spec_its_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_gen_its_data <- lapply(vars, function(x) usedist::dist_subset(tax_gen_its_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_neu_its_data <- lapply(vars, function(x) usedist::dist_subset(tax_neu_its_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_its_data <- data.frame(Group = similar_spec_its_data$Group,
                               Gully_id = similar_spec_its_data$Gully_id, 
                               distance_its_spec = similar_spec_its_data$distance,
                               distance_its_gen = similar_gen_its_data$distance,
                               distance_its_neu = similar_neu_its_data$distance)
```

```{r}
#explore the biotic homogenelization and heteronazition
spec_gen_pro <- read.table(file.path(save.dir, "/tables/spe_gen/spec_gen_pro_95.csv"), sep = ",", header = T)
spec_id_pro <- spec_gen_pro %>% filter(sign == "SPECIALIST") %>% pull(OTU)
gen_id_pro <- spec_gen_pro %>% filter(sign == "GENERALIST") %>% pull(OTU)
neu_id_pro <- spec_gen_pro %>% filter(sign == "NON SIGNIFICANT") %>% pull(OTU)

spec_pro_comm <- phylo_protist@otu_table[rownames(phylo_protist@otu_table) %in% spec_id_pro, ]
gen_pro_comm <- phylo_protist@otu_table[rownames(phylo_protist@otu_table) %in% gen_id_pro, ]
neu_pro_comm <- phylo_protist@otu_table[rownames(phylo_protist@otu_table) %in% neu_id_pro, ]


library(vegan)
#determine the dissimilarity matrix based on the bray-curties distance
tax_spec_pro_dist <-vegdist(t(spec_pro_comm), "bray")
tax_gen_pro_dist <-vegdist(t(gen_pro_comm), "bray")
tax_neu_pro_dist <-vegdist(t(neu_pro_comm), "bray")

# difference in taxonomic variance among Group
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
# Assuming vars is defined somewhere earlier in your code
similar_spec_protist_data <- lapply(vars, function(x) usedist::dist_subset(tax_spec_protist_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_gen_protist_data <- lapply(vars, function(x) usedist::dist_subset(tax_gen_protist_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_neu_protist_data <- lapply(vars, function(x) usedist::dist_subset(tax_neu_protist_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_protist_data <- data.frame(Group = similar_spec_protist_data$Group,
                               Gully_id = similar_spec_protist_data$Gully_id, 
                               distance_protist_spec = similar_spec_protist_data$distance,
                               distance_protist_gen = similar_gen_protist_data$distance,
                               distance_protist_neu = similar_neu_protist_data$distance)

```


```{r}
#explore the biotic homogenelization and heteronazition
spec_gen_anim <- read.table(file.path(save.dir, "/tables/spe_gen/spec_gen_animal_95.csv"), sep = ",", header = T)
spec_id_anim <- spec_gen_anim %>% filter(sign == "SPECIALIST") %>% pull(OTU)
gen_id_anim <- spec_gen_anim %>% filter(sign == "GENERALIST") %>% pull(OTU)
neu_id_anim <- spec_gen_anim %>% filter(sign == "NON SIGNIFICANT") %>% pull(OTU)

spec_anim_comm <- phylo_animal@otu_table[rownames(phylo_animal@otu_table) %in% spec_id_anim, ]
gen_anim_comm <- phylo_animal@otu_table[rownames(phylo_animal@otu_table) %in% gen_id_anim, ]
neu_anim_comm <- phylo_animal@otu_table[rownames(phylo_animal@otu_table) %in% neu_id_anim, ]


library(vegan)
#determine the dissimilarity matrix based on the bray-curties distance
tax_spec_anim_dist <-vegdist(t(spec_anim_comm), "bray")
tax_gen_anim_dist <-vegdist(t(gen_anim_comm), "bray")
tax_neu_anim_dist <-vegdist(t(neu_anim_comm), "bray")

# difference in taxonomic variance among Group
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
# Assuming vars is defined somewhere earlier in your code
similar_spec_anim_data <- lapply(vars, function(x) usedist::dist_subset(tax_spec_anim_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_gen_anim_data <- lapply(vars, function(x) usedist::dist_subset(tax_gen_anim_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_neu_anim_data <- lapply(vars, function(x) usedist::dist_subset(tax_neu_anim_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))

similar_anim_data <- data.frame(Group = similar_spec_anim_data$Group,
                                   Gully_id = similar_spec_anim_data$Gully_id, 
                                   distance_anim_spec = similar_spec_anim_data$distance,
                                   distance_anim_gen = similar_gen_anim_data$distance,
                                   distance_anim_neu = similar_neu_anim_data$distance)
```


```{r}
similar_all_data <- data.frame(similar_data, similar_16s_data[, c(3:5)],
                               similar_its_data[, c(3:5)],
                               similar_protist_data[, c(3:5)],
                               similar_anim_data[, c(3:5)])
distance_index <- c("distance_16s", "distance_16s_gen", "distance_16s_spec", "distance_16s_neu",
                    "distance_its", "distance_its_gen", "distance_its_spec", "distance_its_neu",
                    "distance_pro", "distance_protist_gen", "distance_protist_spec", "distance_protist_neu",
                    "distance_anima", "distance_anim_gen", "distance_anim_spec", "distance_anim_neu")
distance_stats <- similar_all_data %>% group_by(Group) %>%
  get_summary_stats(distance_index, type = "common") %>% #or using type = "mean_sd"
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  arrange(variable, Group)
distance_stats
```


```{r}
library(lme4)
library(lmerTest)
distance_index <- unique(distance_stats$variable)
# determine the effect size of the permafrost thawing for the diversity indexes
dist_scale <- similar_all_data %>% 
  select(c("Group", "Gully_id", distance_index)) %>%
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
dist_S1 <- sapply(3:ncol(dist_scale), function(j) {
  message("Now j=", j, " in ", ncol(dist_scale), ". ", date())
  if (length(unique(dist_scale[, j])) < 3) {
    result <- rep(NA, 10)
  } else {
    fm1 <- lmer(dist_scale[, j] ~ Group + (1 | Gully_id), data = dist_scale)
    
    presult <- car::Anova(fm1, type = 2)
    coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
    names(coefs) <- paste0(names(coefs), ".mean")
    
    SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
    names(SEvalues) <- paste0(names(SEvalues), ".se")
    
    tvalues <- coef(summary(fm1))[, "t value"]  ##t values
    names(tvalues) <- paste0(names(tvalues), ".t")
    
    chisqP <- c(presult[, 1], presult[, 3])
    names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))
    
    result <- c(coefs, tvalues, SEvalues, chisqP)
  }
  result
})
colnames(dist_S1)<-colnames(dist_scale)[-c(1:2)]
data.frame(dist_S1)
```

```{r, fig.align='center', fig.height=5.5, fig.width=3.5}
#plot
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 7),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 7),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.title = element_text(colour = 'black', size = 7),
        legend.text = element_text(colour = 'black', size = 6),
        legend.key.size = unit(0.5, 'cm'))
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
single_dist_comparison <- dist_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  mutate(Taxa = factor(c(rep("Bacteria", 4), rep("Fungi", 4),
                  rep("Protist", 4), rep("Animal", 4)), 
                  levels = c("Bacteria", "Fungi", "Protist", "Animal"))) %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = rev(distance_index))) %>%
  mutate(colour = case_when(GroupCollapsed.mean <= 0 & Group.P <= 0.05 ~ "Negative",
                            GroupCollapsed.mean > 0 & Group.P <= 0.05 ~ "Positvie",
                            Group.P > 0.05 ~ "Neutral")) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, colour = colour)) +
  geom_hline(aes(yintercept = 0), size = 0.7,  colour = "gray2")+
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_color_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  facet_wrap(~Taxa, scales = "free_y", ncol = 1) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "bottom") +
  main_theme + theme(legend.position = "none")

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir.multifunc, "./single_div_comparison.pdf"),
#        single_div_comparison, width = 2.7, height = 5, units = "in")
single_dist_comparison
```
# Explore the effect size of permafrost collapse on the microbial richness and abundance in each group
```{r}
#The ASV orders are the same in tax_table() and spec_gen_anim
ps <- microViz::tax_mutate(phylo_animal, Group = spec_gen_anim$sign)
GPf = tax_glom(ps, "Group") %>% transform_sample_counts(function(x) {x * 100/sum(x)})

aun.table <- data.frame(Group = tax_table(GPf)[,"Group"], otu_table(GPf), row.names = NULL) %>%
  group_by(Group) %>%
  summarise(across(, sum, na.rm = TRUE)) %>% column_to_rownames("Group") %>%
  t() %>% as.data.frame() %>% rownames_to_column("Sample_id") %>% as_tibble() %>% 
  mutate(Gully_id = sapply(stringr::str_split(Sample_id, "_",  n = 2), `[`, 1)) %>%
  mutate(Group = case_when(grepl("_C", Sample_id) ~ "Uncollapsed",
                              grepl("_T", Sample_id) ~ "Collapsed"))

aun.table
```


```{r}
library(lme4)
library(lmerTest)
group_taxa_name <- c("GENERALIST", "SPECIALIST", "NON SIGNIFICANT")
# determine the effect size of the permafrost thawing for the diversity indexes
abun_scale <- aun.table %>% 
  select(c("Gully_id", "Group", group_taxa_name)) %>%
  mutate(across(where(is.numeric), scale)) %>% as.data.frame() %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
abun_S1 <- sapply(3:ncol(abun_scale), function(j) {
  if (length(unique(abun_scale[, j])) < 3) {
    result <- rep(NA, 10)
  } else {
    fm1 <- lmer(abun_scale[, j] ~ Group + (1 | Gully_id), data = abun_scale)
    
    presult <- car::Anova(fm1, type = 2)
    coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
    names(coefs) <- paste0(names(coefs), ".mean")
    
    SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
    names(SEvalues) <- paste0(names(SEvalues), ".se")
    
    tvalues <- coef(summary(fm1))[, "t value"]  ##t values
    names(tvalues) <- paste0(names(tvalues), ".t")
    
    chisqP <- c(presult[, 1], presult[, 3])
    names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))
    
    result <- c(coefs, tvalues, SEvalues, chisqP)
  }
  result
})
colnames(abun_S1)<-colnames(abun_scale)[-c(1:2)]
data.frame(abun_S1)
```


# Resistance analysis
```{r}
gp <- paste(metadata$Gully_id, metadata$Group, sep = "_")
extract_fun <- function(df) {
  simi_group_df <- dist_groups(1-df, gp)
  between_df <- simi_group_df %>% filter(Label %in% grep("^Between", simi_group_df$Label, value = T))
  simi_between_group_df <- rbind(between_df %>% filter(str_detect(Item1, 'G1')) %>% filter(str_detect(Item2, 'G1')), 
                                 between_df %>% filter(str_detect(Item1, 'G2')) %>% filter(str_detect(Item2, 'G2')), 
                                 between_df %>% filter(str_detect(Item1, 'G3')) %>% filter(str_detect(Item2, 'G3')), 
                                 between_df %>% filter(str_detect(Item1, 'G4')) %>% filter(str_detect(Item2, 'G4')), 
                                 between_df %>% filter(str_detect(Item1, 'G5')) %>% filter(str_detect(Item2, 'G5')), 
                                 between_df %>% filter(str_detect(Item1, 'G6')) %>% filter(str_detect(Item2, 'G6'))) %>%
    as_tibble() %>% 
    select(Label, Distance) %>% group_by(Label) %>% summarize(Resistance = mean(Distance))
  return(simi_between_group_df)
}
#Bacteria
simi_between_16s_df <- extract_fun(tax_16s_dist)
simi_between_16s.gen_df <- extract_fun(tax_gen_16s_dist)
simi_between_16s.spec_df <- extract_fun(tax_spec_16s_dist)
simi_between_16s.neu_df <- extract_fun(tax_neu_16s_dist)
#Fungi
simi_between_its_df <- extract_fun(tax_its_dist)
simi_between_its.gen_df <- extract_fun(tax_gen_its_dist)
simi_between_its.spec_df <- extract_fun(tax_spec_its_dist)
simi_between_its.neu_df <- extract_fun(tax_neu_its_dist)
#Protist
simi_between_pro_df <- extract_fun(tax_pro_dist)
simi_between_pro.gen_df <- extract_fun(tax_gen_pro_dist)
simi_between_pro.spec_df <- extract_fun(tax_spec_pro_dist)
simi_between_pro.neu_df <- extract_fun(tax_neu_pro_dist)
#Animal
simi_between_anim_df <- extract_fun(tax_anim_dist)
simi_between_anim.gen_df <- extract_fun(tax_gen_anim_dist)
simi_between_anim.spec_df <- extract_fun(tax_spec_anim_dist)
simi_between_anim.neu_df <- extract_fun(tax_neu_anim_dist)

resitance_df <- rbind(data.frame(Taxa = rep("Bacteria", 6), Group = rep("Overall species"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_16s_df),
                      data.frame(Taxa = rep("Bacteria", 6), Group = rep("Generalists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_16s.gen_df),
                      data.frame(Taxa = rep("Bacteria", 6), Group = rep("Specialists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_16s.spec_df),
                      data.frame(Taxa = rep("Bacteria", 6), Group = rep("Opportunists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_16s.neu_df),
                      data.frame(Taxa = rep("Fungi", 6), Group = rep("Overall species"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_its_df),
                      data.frame(Taxa = rep("Fungi", 6), Group = rep("Generalists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_its.gen_df),
                      data.frame(Taxa = rep("Fungi", 6), Group = rep("Specialists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_its.spec_df),
                      data.frame(Taxa = rep("Fungi", 6), Group = rep("Opportunists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_its.neu_df),
                      data.frame(Taxa = rep("Protist", 6), Group = rep("Overall species"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_pro_df),
                      data.frame(Taxa = rep("Protist", 6), Group = rep("Generalists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_pro.gen_df),
                      data.frame(Taxa = rep("Protist", 6), Group = rep("Specialists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_pro.spec_df),
                      data.frame(Taxa = rep("Protist", 6), Group = rep("Opportunists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_pro.neu_df),
                      data.frame(Taxa = rep("Animal", 6), Group = rep("Overall species"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_anim_df),
                      data.frame(Taxa = rep("Animal", 6), Group = rep("Generalists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_anim.gen_df),
                      data.frame(Taxa = rep("Animal", 6), Group = rep("Specialists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_anim.spec_df),
                      data.frame(Taxa = rep("Animal", 6), Group = rep("Opportunists"),
                                 Gully_id = c("EB", "HH", "HSX", "ML", "RS", "SLH"), 
                                 simi_between_anim.neu_df)
)
resitance_df
```

Plot
```{r, fig.align='center', fig.height=5.5, fig.width=2.5}
resistance_stats <- resitance_df %>% group_by(Taxa, Group) %>%
  get_summary_stats(Resistance, type = "common") %>% #or using type = "mean_sd"
  mutate(Taxa = factor(Taxa, levels = c("Bacteria", "Fungi", "Protist", "Animal"))) %>%
  mutate(Group = factor(Group, levels = rev(c("Overall species", "Generalists", "Specialists", "Opportunists"))))
resistance_comparison <- resistance_stats %>%
  ggplot(aes(x = Group, y = mean, colour = "#e95f5c")) +
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = mean - se, ymax = mean + se, colour = "#e95f5c"), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  labs(x = NULL, y = "Resistance") +
  # scale_color_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  facet_wrap(~Taxa, scales = "free_y", ncol = 1) +
  scale_y_continuous(expand = c(0, 0), limit = c(0, 1)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "bottom") +
  main_theme + theme(legend.position = "none")
resistance_comparison
```

```{r, fig.align = 'center', fig.height = 5.5, fig.width = 6.5}
compo_venn_eff.size_resis_plot <- cowplot::plot_grid(venn_plots, single_dist_comparison, resistance_comparison, ncol = 3, align = "h")
# ggsave(file.path(save.dir, './figs/null_model/process_comparison.pdf'), null_plot, width = 6, height = 3.5)
compo_venn_eff.size_resis_plot
```

