---
title: "Diversity and Distance analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set working and saving directory
```{r, message=FALSE}
setwd('e:/thermokarst_gully/')
save.dir <- file.path(getwd(),"result")
```
Loading packages
```{r}
pacman::p_load(phyloseq, ape, vegan, Biostrings, microbiome, tidyverse, rstatix)
```

Composition
```{r}
# Extrac the OTU table with the relative abundance higher than 1%
subextract_fun <- function(phylo, taxa_level, perc) {
  subphylo <- tax_glom(phylo, taxa_level, NArm = F)
  subphylo.rel  = transform_sample_counts(subphylo, function(x) x / sum(x))
  ra.tab <- otu_table(subphylo.rel)
  subtaxa_tab <- tax_table(subphylo.rel)[, taxa_level]
  otu_final_tab <- data.frame(subtaxa_tab, ra.tab) %>%
    mutate(MRA = rowMeans(select(., colnames(ra.tab)))) %>%
    filter(MRA >= perc) %>%
    arrange(desc(MRA)) %>%
    select(., -c('MRA')) %>%
    # bind_rows(summarise_all(., ~if(is.numeric(.)) 1-sum(.) else "Others"))
  return(otu_final_tab)
}

otu_rel_abun_1perc_phyla_16s <- subextract_fun(phylo_16s, "Phylum", 0.01)
# extract the taxa at class level
otu_rel_abun_1perc_class_16s <- subextract_fun(phylo_16s, "Class", 0.01)

#ecombine the class within proteobacteria and other phyla into one abundance table
otu_rel_abun_1perc_16s <- rbind(otu_rel_abun_1perc_class_16s %>%
                         filter(Class %in% c("Alphaproteobacteria", "Betaproteobacteria",
                                             "Gammaproteobacteria", "Deltaproteobacteria")) %>%
                         dplyr::rename(Phylum = Class), otu_rel_abun_1perc_phyla_16s %>%
                         filter(!Phylum %in% c("Proteobacteria", "Unassigned")))

# extrac the Fungal phyla with the relative abundance higher than 1%
otu_rel_abun_1perc_its <- subextract_fun(phylo_its, "Phylum", 0.01) %>%
  filter(!Phylum %in% c("Unassigned"))

# extrac the Protistan phyla with the relative abundance higher than 1%
otu_rel_abun_1perc_pro <- subextract_fun(phylo_protist, "Phylum", 0.01) %>%
  filter(!Phylum %in% c("Unassigned"))

# extrac the Animal phyla with the relative abundance higher than 1%
otu_rel_abun_1perc_anim <- subextract_fun(phylo_animal, "Class", 0.01) %>%
  filter(!Class %in% c("Unassigned"))

#Combine all data
compo_df <- rbind(otu_rel_abun_1perc_16s %>%
  pivot_longer(-Phylum, names_to = "Sample_id", values_to = "Rela_abun") %>%
    mutate(Group = rep("Bacteria", nrow(.))),
  otu_rel_abun_1perc_its %>%
    pivot_longer(-Phylum, names_to = "Sample_id", values_to = "Rela_abun") %>%
    mutate(Group = rep("Fungi", nrow(.))),
  otu_rel_abun_1perc_pro %>%
    pivot_longer(-Phylum, names_to = "Sample_id", values_to = "Rela_abun") %>%
    mutate(Group = rep("Protist", nrow(.))),
  otu_rel_abun_1perc_anim %>%
    pivot_longer(-Class, names_to = "Sample_id", values_to = "Rela_abun") %>%
    dplyr::rename(c(Phylum = Class)) %>% 
    mutate(Group = rep("Animal", nrow(.)))) %>%
  dplyr::rename(c(Taxa = Phylum)) %>%
  mutate(Type = ifelse(grepl("_C", Sample_id), "Control",
                       ifelse(grepl("_T", Sample_id), "Collapsed", ""))) %>%
  select(-Sample_id) %>%
  group_by(Type, Group, Taxa) %>%
  summarise(across(everything(), sum)) %>%
  mutate(Type_group = case_when(Type == "Control" ~ "a",
                                Type == "Collapsed" ~ "b"))

compo_df[1:5, ]
```

Arrange the data for sankey plot 
```{r}
sankey <- data.table::rbindlist(list(compo_df[c("Type", "Group", "Rela_abun")],
                                     compo_df[c("Group", "Taxa", "Rela_abun")]))

names(sankey) <- c('source', 'target', 'value')
knitr::kable(sankey[1:5, ])

# Make a connection data frame
links <- sankey 
# From these flows we need to create a node data frame: it lists every entities involved in the flow
nodes <- data.frame(
  name = c("Control", "Collapsed", "Bacteria", "Fungi", "Protist", "Animal", 
           otu_rel_abun_1perc_16s$Phylum,
           otu_rel_abun_1perc_its$Phylum, otu_rel_abun_1perc_pro$Phylum,
           otu_rel_abun_1perc_anim$Class)
)

# With networkD3, connection must be provided using id, not using real name like in the links dataframe.. So we need to reformat it.
links$IDsource <- match(links$source, nodes$name)-1 
links$IDtarget <- match(links$target, nodes$name)-1

# Set color of connections
# Add a 'group' column to each connection:
links_datset <- mutate(links,
                       group = rep(c("a", "b"), 54)
                       )
# Add a 'group' column to each node. Here I decide to put all of them in the same group to make them grey
nodes$group <- as.factor(c("my_unique_group"))
links_datset[1:5, ]
```
Sanke plot
```{r}
library(networkD3)
# Give a color for each group
# my_color <- 'd3.scaleOrdinal() .domain(["a", "b", "my_unique_group"]) .range(["#4970a0", "#f8696b", "#FCCDE5"])'

# Make the Network. I call my colour scale with the colourScale argument
sankey_plot <- sankeyNetwork(Links = links_datset, Nodes = nodes, 
                   Source = "IDsource", Target = "IDtarget", 
                   Value = "value", NodeID = "name", iterations = 0, 
                   fontFamily = 'Arial', fontSize = 15, 
                   nodeWidth = 20, nodePadding = 10, height = 700, width = 600,
                   sinksRight = F)
# save the widget
# recommend htmlwidgets
# library(htmlwidgets)
# saveWidget(p, file = paste0(sankey.plots.folder, "/sankey_plot24_6_10.html"))
# library(webshot)
# install phantom:
# webshot::install_phantomjs()
# Make a webshot in pdf : high quality but can not choose printed zone
# webshot(paste0(sankey.plots.folder, "/sankey_plot.html"), delay = 0.2)
sankey_plot

# 
# pacman::p_load(ggthemes, networkD3, data.table)
# sankeyNetwork(Links = links, Nodes = nodes, 
#               Source = "IDsource", Target = "IDtarget", 
#               Value = "value", NodeID = "name", iterations = 0, 
#               fontFamily = 'Arial', fontSize = 15, 
#               nodeWidth = 20, nodePadding = 10, height = 700, width = 700,
#               sinksRight = F)

```

## Test the overall difference in the taxonomic and functional composition.
```{r}
library(vegan)
#determine the dissimilarity matrix based on the bray-curties distance
tax_16s_dist <-vegdist(t(otu_16s), "bray")
tax_its_dist <-vegdist(t(otu_its), "bray")
tax_pro_dist <-vegdist(t(otu_pro), "bray")
tax_anim_dist <-vegdist(t(otu_anim), "bray")
fun_dist <-vegdist(t(ko_count_table), "bray")
#permanova test the difference in compositional variance
adonis2(tax_16s_dist ~ Group, data = metadata)
adonis2(tax_its_dist ~ Group, data = metadata)
adonis2(tax_pro_dist ~ Group, data = metadata)
adonis2(tax_anim_dist ~ Group, data = metadata)
adonis2(fun_dist ~ Group, data = metadata)
```


Visualization for the overall difference by PCoA plot.
```{r, fig.align='center', fig.width=8.9, fig.height=2.5}
#16S
ord.16s <-  cmdscale(tax_16s_dist,  k = 2, eig = T, add = T)
pcoa_16s_plot <- data.frame(Group = metadata$Group, scores(ord.16s)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.16s$eig[1] / sum(ord.16s$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.16s$eig[2] / sum(ord.16s$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none", 
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
#ITS
ord.its <-  cmdscale(tax_its_dist,  k = 2, eig = T, add = T)
pcoa_its_plot <- data.frame(Group = metadata$Group, scores(ord.its)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.its$eig[1] / sum(ord.its$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.its$eig[2] / sum(ord.its$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none", 
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
#protist
ord.pro <-  cmdscale(tax_pro_dist,  k = 2, eig = T, add = T)
pcoa_pro_plot <- data.frame(Group = metadata$Group, scores(ord.pro)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.pro$eig[1] / sum(ord.pro$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.pro$eig[2] / sum(ord.pro$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none", 
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
#Animal
ord.anim <-  cmdscale(tax_anim_dist,  k = 2, eig = T, add = T)
pcoa_anim_plot <- data.frame(Group = metadata$Group, scores(ord.anim)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.anim$eig[1] / sum(ord.anim$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.anim$eig[2] / sum(ord.anim$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none", 
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
#Function
ord.fun <-  cmdscale(fun_dist,  k = 2, eig = T, add = T)
pcoa_fun_plot <- data.frame(Group = metadata$Group, scores(ord.fun)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  # stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  labs(x=paste("PCoA1 (", format(100 * ord.fun$eig[1] / sum(ord.fun$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.fun$eig[2] / sum(ord.fun$eig), digits = 3), "%)", sep = "")) +
  theme(legend.position = "none",
        axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
cowplot::plot_grid(pcoa_16s_plot, pcoa_its_plot, pcoa_pro_plot, pcoa_anim_plot, pcoa_fun_plot, ncol = 5)
```






## Explore the biotic homogenization
```{r, message=FALSE}
# difference in taxonomic variance among Group
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
Similar_16s_data <- sapply(vars, function(x) usedist::dist_subset(tax_16s_dist, grep(x, metadata$Sample_name, value = T))) %>%
  data.frame() %>%  gather("tem_group", "distance") %>% 
  cbind(Gully_id = c(rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20)), 
        Group = c(rep(c('Control', 'Collapsed'), each = 10, times = 6))) %>%
  select(., -c('tem_group')) %>% 
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'))) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed')))

Similar_its_data <- sapply(vars, function(x) usedist::dist_subset(tax_its_dist, grep(x, metadata$Sample_name, value = T))) %>%
  data.frame() %>%  gather("tem_group", "distance") %>% 
  cbind(Gully_id = c(rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20)), 
        Group = c(rep(c('Control', 'Collapsed'), each = 10, times = 6))) %>%
  select(., -c('tem_group')) %>% 
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'))) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed')))

Similar_pro_data <- sapply(vars, function(x) usedist::dist_subset(tax_pro_dist, grep(x, metadata$Sample_name, value = T))) %>%
  data.frame() %>%  gather("tem_group", "distance") %>% 
  cbind(Gully_id = c(rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20)), 
        Group = c(rep(c('Control', 'Collapsed'), each = 10, times = 6))) %>%
  select(., -c('tem_group')) %>% 
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'))) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed')))

Similar_anim_data <- sapply(vars, function(x) usedist::dist_subset(tax_anim_dist, grep(x, metadata$Sample_name, value = T))) %>%
  data.frame() %>%  gather("tem_group", "distance") %>% 
  cbind(Gully_id = c(rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20)), 
        Group = c(rep(c('Control', 'Collapsed'), each = 10, times = 6))) %>%
  select(., -c('tem_group')) %>% 
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'))) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed')))

Similar_fun_data <- sapply(vars, function(x) usedist::dist_subset(fun_dist, grep(x, metadata$Sample_name, value = T))) %>%
  data.frame() %>%  gather("tem_group", "distance") %>% 
  cbind(Gully_id = c(rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20)), 
        Group = c(rep(c('Control', 'Collapsed'), each = 10, times = 6))) %>%
  select(., -c('tem_group')) %>% 
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'))) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed')))

similar_data <- data.frame(Group = Similar_16s_data$Group,
                            Gully_id = Similar_16s_data$Gully_id, 
                            distance_16s = Similar_16s_data$distance,
                            distance_its = Similar_its_data$distance,
                            distance_pro = Similar_pro_data$distance, 
                            distance_anima = Similar_anim_data$distance,
                            distance_fun = Similar_fun_data$distance)

#Descriptive statistics for all diversity indexes
distance_index <- c("distance_16s", "distance_its", "distance_pro", "distance_anima", "distance_fun")
distance_stats <- similar_data %>% group_by(Group) %>%
  get_summary_stats(distance_index, type = "common") %>% #or using type = "mean_sd"
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  arrange(variable, Group)
distance_stats
```
### Test the difference in beta distance between control and collapsed groups using liner models
```{r}
# determine the effect size of the permafrost thawing for the diversity indexes
dist_scale <- similar_data %>% 
  select(c("Group", "Gully_id", distance_index)) %>%
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
dist_S1 <- sapply(3:ncol(dist_scale), function(j) {
    message("Now j=", j, " in ", ncol(dist_scale), ". ", date())
    if (length(unique(dist_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(dist_scale[, j] ~ Group + (1 | Gully_id), data = dist_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(dist_S1)<-colnames(dist_scale)[-c(1:2)]
data.frame(dist_S1)
```

```{r, fig.align='center', fig.width=3.5, fig.height=2.5}
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
single_dist_comparison <- dist_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = rev(distance_index))) %>%
  mutate(colour = case_when(GroupCollapsed.mean <= 0 & Group.P <= 0.05 ~ "Negative",
                            GroupCollapsed.mean > 0 & Group.P <= 0.05 ~ "Positvie",
                            Group.P > 0.05 ~ "Neutral")) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, colour = colour)) +
  geom_hline(aes(yintercept = 0), size = 0.7,  colour = "gray2")+
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_color_manual(values=c("grey", "#e95f5c")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "bottom") +
  main_theme +
  theme(legend.position = "none",
        strip.background = element_rect(fill = c("#FFF6E1")),
        # axis.text.y = element_blank()
        )

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir.multifunc, "./single_div_comparison.pdf"),
#        single_div_comparison, width = 2.7, height = 5, units = "in")
single_dist_comparison
```

```{r, fig.align='center', fig.width=6.5, fig.height=7}
cowplot::plot_grid(pcoa_16s_plot, pcoa_its_plot, pcoa_pro_plot, pcoa_anim_plot, pcoa_fun_plot, single_dist_comparison, ncol = 2, align = "hv")
```

Linear mixed models test the effect of permafrost thawing on microbial similarity
```{r, message=FALSE}
library(lme4)
library(lmerTest)
lmm_dist_modes <- lapply(distance_index, function(x) {
  lmer(substitute(i ~ Group + (1|Gully_id), list(i = as.name(x))), data = similar_data)})
df <- NULL
for(i in 1:length(distance_index)) {
  tmp <- summary.model(lmm_dist_modes[[i]])
  if (is.null(df)){
    df <- tmp
  } else {
    df <- rbind(df, tmp)
  }
}

dist_result_lmm <-data.frame(dist_index = distance_index, 
                             group1 = rep("Control", length(distance_index)),
                             group2 = rep("Collapsed", length(distance_index)), df)
dist_result_lmm
```


```{r, fig.align='center', fig.height=2, fig.width=7}
# Add the p-value manually
sig.dist.labs <- data.frame(Group = distance_stats$Group, 
                            dist_index = distance_stats$variable,
                            y.pos = distance_stats$max *1.05,
                            sig.lab = c("b", "a", "b", "a", "a", "a", "b", "a", "b", "a"))
# Create a box plot
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 7),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 7),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.title = element_text(colour = 'black', size = 7),
        legend.text = element_text(colour = 'black', size = 6),
        legend.key.size = unit(0.5, 'cm'))
library(gghalves)
dist_plot <- similar_data %>% select(c("Group", distance_index)) %>%
  gather(dist_index, value, -c("Group")) %>% 
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  mutate(dist_index = factor(dist_index, levels = distance_index)) %>%
  ggplot(aes(Group, value, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  geom_text(data = sig.dist.labs, aes(y = y.pos, label = sig.lab), 
                  position = position_dodge(width = .75), size = 3.5) +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~dist_index, scales = "free_y", ncol = 5) +
  main_theme +
  theme(legend.position = "none")

# save the plot
dist_plot
```

