---
title: "Diversity and Distance analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set working and saving directory
```{r, message=FALSE}
setwd('e:/thermokarst_gully/')
save.dir <- file.path(getwd(),"result")
```
Loading packages
```{r}
pacman::p_load(phyloseq, ape, vegan, Biostrings, microbiome, tidyverse, rstatix)
```
Data input
```{r, message=FALSE}
#Read in data
source("script/read_data_all.R")
div.file <- "E:/thermokarst_gully/data/multifun/multifun_richness_table_all.txt"
div_table <-  read.table(div.file, header = TRUE, sep = "\t", row.names = 1,
                  as.is = TRUE, stringsAsFactors = FALSE, comment.char = "",
                  check.names = FALSE)
```

## Descriptive statistics for all diversity indexes
```{r}
div_index <- c("Bacterial richness",	"Fungal richness",	"Protist richness",	"Animal richness",	
               "Soil multidiversity", "Gene numbers")
div_stats <- div_table %>% group_by(Group) %>%
  get_summary_stats(div_index, type = "common") %>% #or using type = "mean_sd"
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  arrange(variable, Group)
div_stats
```
## Linear mixed models test the effect of permafrost thawing on microbial diversity
```{r, message=FALSE}
library(lme4)
library(lmerTest)
lmm_div_modes <- lapply(div_index, function(x) {
  lmer(substitute(i ~ Group + (1|Gully_id), list(i = as.name(x))), data = div_table)})
summary.model <- function(model){
  F.value <- anova(model)$'F value'
  p.value <- anova(model)$'Pr(>F)'
  p.stars <- function(p.values) {
    unclass(symnum(p.values, corr = FALSE, 
                   na = FALSE, cutpoints = c(0,0.001, 0.01, 0.05, 0.1, 1),
                   symbols = c("***", "**", "*", ".", " ")))}
  sig <- p.stars(p.value)
  results<-data.frame(F.value, p.value, sig)
  return(results)
}
df <- NULL
for(i in 1:length(div_index)) {
  tmp <- summary.model(lmm_div_modes[[i]])
  if (is.null(df)){
    df <- tmp
  } else {
    df <- rbind(df, tmp)
  }
}

div_result_lmm <-data.frame(diversity = div_index, group1 = rep("Uncollapsed", length(div_index)),
                            group2 = rep("Collapsed", length(div_index)), df)
div_result_lmm
```
```{r, fig.align='center', fig.height=4, fig.width=7}
# Add the significant symbols manually
sig.div.labs <- tibble(diversity = factor(div_index, levels = div_index),
                x1 = rep(0.5, length(div_index)),
                y1 = rep(0.95, length(div_index)),
                sig.labels = div_result_lmm$sig,
                x2 = rep(0.1, length(div_index)),
                y2 = rep(1, length(div_index)),
                panel.labels = letters[as.numeric(diversity)])

# Create a box plot
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 7),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 7),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.title = element_text(colour = 'black', size = 7),
        legend.text = element_text(colour = 'black', size = 6),
        legend.key.size = unit(0.5, 'cm'))
library(gghalves)
div_plot <- div_table %>% select(c("Group", div_index)) %>%
  gather(diversity, value, -c("Group")) %>% 
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  mutate(diversity = factor(diversity, levels = div_index)) %>%
  ggplot(aes(Group, value, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  ggpp::geom_text_npc(data = sig.div.labs, aes(npcx = x1, npcy = y1, label = sig.labels), inherit.aes = F) +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~diversity, scales = "free_y", ncol = 3) +
  main_theme +
  theme(legend.position = "none")

# save the plot
ggsave(file.path(save.dir, "./figs/diversity/div_four_groups_plot.pdf"),
       div_plot, width = 4.5, height = 3, units = "in")
div_plot
```

### Test the difference in subtaxa between uncollapsed and collapsed groups using liner models
```{r}
# determine the effect size of the permafrost thawing for the diversity indexes
tax_index <- c("Actinobacteria",	"Alphaproteobacteria",	"Acidobacteria",	"Firmicutes",	"Betaproteobacteria", "Planctomycetes",	"Gammaproteobacteria",	"Bacteroidetes",	"Deltaproteobacteria", "Ascomycota",	"Basidiomycota",	"Mortierellomycota", "Cercozoa",	"Conosa",	"Ciliophora",	"Chlorophyta",	"Apicomplexa",	"Ochrophyta",	"Pseudofungi",	"Lobosa",	"Sagenista",	"Opalozoa",	"Nematoda",	"Rotifera",	"Arthropoda",	"Annelida",	"Tardigrada", "Carbohydrate metabolism", "Nitrogen metabolism",	"Phosphorus metabolism",	"Sulfur metabolism")

div_scale <- div_table %>% 
  select(c("Group", "Gully_id", tax_index)) %>%
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
div_S1 <- sapply(3:ncol(div_scale), function(j) {
    message("Now j=", j, " in ", ncol(div_scale), ". ", date())
    if (length(unique(div_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(div_scale[, j] ~ Group + (1 | Gully_id), data = div_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(div_S1)<-colnames(div_scale)[-c(1:2)]
data.frame(div_S1)
```


```{r, fig.align='center', fig.width=3.5, fig.height=7.8}
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
single_div_comparison <- div_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = rev(tax_index))) %>%
  mutate(colour = case_when(GroupCollapsed.mean <= 0 & Group.P <= 0.05 ~ "Negative",
                            GroupCollapsed.mean > 0 & Group.P <= 0.05 ~ "Positvie",
                            Group.P > 0.05 ~ "Neutral")) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, colour = colour)) +
  geom_hline(aes(yintercept = 0), size = 0.375,  colour = "gray2")+
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_color_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0.5, xmax = 4.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 4.5, xmax = 9.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 9.5, xmax = 19.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 19.5, xmax = 22.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#5fb236") +
  annotate("rect", xmin = 22.5, xmax = 31.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#1ca9c9") +
  main_theme +
  theme(legend.position = "none",
        strip.background = element_rect(fill = c("#FFF6E1")),
        # axis.text.y = element_blank()
        )

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
ggsave(file.path(save.dir, "./figs/diversity/single_div_comparison.pdf"),
       single_div_comparison, width = 2.5, height = 5.25, units = "in")
single_div_comparison
```


```{r}
# read data
ko_description <- 'E:/thermokarst_gully/data/metagenome/KO_description.txt'
ko_9_levels_file <- 'E:/thermokarst_gully/data/metagenome/ko_9_levels.txt'

ko_9_levels <- read.table(ko_9_levels_file, header = TRUE, sep = "\t", 
                  as.is = TRUE, stringsAsFactors = FALSE, comment.char = "",
                  check.names = FALSE)

# Filter the L2 == "Metabolism" rows and merge the KO table
ko_metabolism_df <- ko_9_levels %>% filter(L1 == "Metabolism") %>% inner_join(ko_tpm_table %>% rownames_to_column("KO"), c("KO" = "KO")) %>% select(L2, all_of(metadata$Sample_name))

ko_metabolism_df %>% group_by(L2) %>% colSums(. != 0)


ko_metabolism_df %>%
  group_by(L2) %>%
  summarise(across(metadata$Sample_name, ~ sum(.x != 0))) %>%
  write.csv(., file = "E:/thermokarst_gully/data/multifun/kegg_L1_KO.csv")
```

