---
title: "Determine the drivers underlying the microbial assemblages"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set working and saving directory
```{r, message=FALSE}
setwd('e:/thermokarst_gully/')
save.dir <- file.path(getwd(),"result")
```
Loading packages and read data
```{r}
pacman::p_load(phyloseq, ape, vegan, Biostrings, microbiome, tidyverse, rstatix)
source("script/read_data_all.R")
div_table <- read.table(file = "E:/thermokarst_gully/data/multifun/multifun_richness_table_all.csv",
             sep = ',',  header = T, row.names = 1, stringsAsFactors = F)
```

Test the difference in the environmental variables between the uncollapsed and collapsed sites
```{r}
env_vars <- c("Plant_richness", "AGB", "BGB", "pH", "Soil_moisture", "Clay_Silt", "WHC", "SOC",
              "NH4_N", "NO3_N", "AP")
env_stats <- metadata %>% group_by(Group) %>%
  get_summary_stats(env_vars, type = "common") %>% #or using type = "mean_sd"
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  arrange(variable, Group)
env_stats
```

```{r}
library(lme4)
library(lmerTest)
lmm_env_modes <- lapply(env_vars, function(x) {
  lmer(substitute(i ~ Group + (1|Gully_id), list(i = as.name(x))), data = metadata)})
summary.model <- function(model){
  F.value <- anova(model)$'F value'
  p.value <- anova(model)$'Pr(>F)'
  p.stars <- function(p.values) {
    unclass(symnum(p.values, corr = FALSE, 
                   na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 1),
                   symbols = c("***", "**", "*", "ns")))}
  sig <- p.stars(p.value)
  results<-data.frame(F.value, p.value, sig)
  return(results)
}
df <- NULL
for(i in 1:length(env_vars)) {
  tmp <- summary.model(lmm_env_modes[[i]])
  if (is.null(df)){
    df <- tmp
  } else {
    df <- rbind(df, tmp)
  }
}

env_result_lmm <-data.frame(Variables = env_vars, group1 = rep("Uncollapsed", length(env_vars)),
                            group2 = rep("Collapsed", length(env_vars)), df)
env_result_lmm
```

```{r, fig.align='center', fig.height=5, fig.width=6.5}
# Add the p-value manually
env_stats <- data.frame(env_stats,
                       sig.lab = c("a", "a", "a", "b", "a", "b", 
                                   "a", "a", "a", "b", "a", "b",
                                   "a", "b", "a", "b", "a", "a",
                                   "b", "a", "a", "a"))
# Create a bar plot
env_plot <- env_stats %>% 
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  mutate(variable = factor(variable, levels = env_vars)) %>%
  ggplot(aes(Group, mean, fill = Group)) +
  geom_bar(position = position_dodge(0.8), stat = 'identity', 
           colour = 'black', width = 0.6, )+
  geom_errorbar(aes(ymin = mean, ymax = mean + se), width = .2) +
  # geom_text(aes(y = 1.05*(mean + se), label = sig.lab), 
  #           position = position_dodge(width = .75), size = 2.5, vjust = 0.01) + # add lowercase letters to indicate the significance
  scale_y_continuous(expand = expansion(mult = c(0, 0.2))) +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~variable, scales = "free", ncol = 4, strip.position = "left", labeller = facet_labeller) +
  theme_classic() + 
  theme(legend.position = "none",
        strip.placement = "outside",
        strip.background = element_rect(colour = NA),
        strip.text = element_text(colour = 'black', size = 6, margin = margin()),
        axis.title = element_text(color = 'black',size = 6),
        axis.text = element_text(colour = 'black', size = 5),
        axis.line = element_line(size = 0.4),
        axis.ticks = element_line(color = "black", linewidth = 0.4),
        legend.title = element_text(colour = 'black', size = 6),
        legend.text = element_text(colour = 'black', size = 5),
        legend.key.size = unit(0.5, 'cm'))

lines <- tibble(variable = factor(c("Plant_richness", "AGB", "BGB", "pH", "Soil_moisture", "Clay_Silt", "WHC", "SOC", "NH4_N", "NO3_N", "AP"), levels = env_vars),
                x = c(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1),
                xend = c(2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2),
                y = c(17, 270, 550, 8, 220, 72, 260, 185, 19, 40, 2.4),
                yend = c(17, 270, 550, 8, 220, 72, 260, 185, 19, 40, 2.4)
)
stars <- tibble(variable = factor(c("Plant_richness", "AGB", "BGB", "pH", "Soil_moisture", "Clay_Silt", "WHC", "SOC", "NH4_N", "NO3_N", "AP"), levels = env_vars),
                x = c(1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5),
                y = c(18, 280, 565, 8.5, 230, 76, 275, 195, 20, 42, 2.5),
                labels = env_result_lmm$sig
)

env_plot <- env_plot +
  # geom_segment(data = lines, aes(x = x, xend = xend, y = y, yend = yend), inherit.aes = FALSE) + # add a sigment to denote the comparison between groups
  geom_text(data = stars, aes(x = x, y = y, label = labels), inherit.aes = F)

# save the plot
ggsave(file.path(save.dir, "./figs/env_effect/env_comparison_plot.pdf"),
       env_plot, width = 145, height = 90, units = "mm")
env_plot
```


```{r}
env_vars <- c("Plant_richness", "pH", "Soil_moisture", "Clay_Silt", "WHC", "AGB", "BGB", "SOC",
              "NH4_N", "NO3_N", "AP")
div_vars <- c("Ric_16s", "Ric_its", "Ric_pro", "Ric_anim", "Ric_kegg")

df_all <- data.frame(metadata[, c("Group", "Gully_id", env_vars)],
                 div_table[, div_vars])
```

```{r, message=FALSE}
#linner mixed model for matrix to matrix
library(lme4)
lmm.mat.cal <- function(y, x, input_df){
  y <- as.matrix(y)
  x <- as.matrix(x)
  df <- NULL
  for(i in colnames(y)){
    for(j in colnames(x)){
      a <- y[, i, drop = F]
      b <- x[, j, drop = F]
      mode <- lmer(a ~ b + (1 | Gully_id), data = input_df, na.action = na.omit)
      coeff <- summary(mode)$coefficients[2, 1]
      r.square <- MuMIn::r.squaredGLMM(mode)[1]
      p.value <- car::Anova(mode, type = 2)[, 3]
      tmp <- c(i, j, coeff, r.square, p.value)
      if(is.null(df)){
        df <- tmp  
      }
      else{
        df <- rbind(df, tmp)
      }    
    }
  }
  df <- data.frame(row.names = NULL, df) %>% mutate_at(c(3:5), as.numeric)
  colnames(df) <- c("Diversity", "Env", "coefficents", "r.square", "Pvalue")
  df$Pvalue <- as.numeric(as.character(df$Pvalue))
  df$AdjPvalue <- rep(0, dim(df)[1])
  #You can adjust the p-values for multiple comparison using Benjamini & Hochberg (1995):
  for(i in unique(df$Env)){
      sel <- df$Env == i
      df$AdjPvalue[sel] <- p.adjust(df$Pvalue[sel], method = "BH")
      }
  #Now we generate the labels for signifant values
  df$Sig <- cut(df$AdjPvalue, breaks = c(-Inf, 0.001, 0.01, 0.05, Inf), label = c("***", "**", "*", ""))
  df$Env <- factor(df$Env, ordered = T, levels = rev(colnames(x)))
  df$Diversity <- factor(df$Diversity, ordered = T, levels = colnames(y))
  return(df)
}

diver_tab <- df_all[ ,div_vars] %>% scale()
env_tab <- df_all[ ,env_vars] %>% scale()
lmm.matrix <- lmm.mat.cal(diver_tab, env_tab, df_all)

df_uncollapsed <- df_all %>% filter(Group == "Uncollapsed")
diver_uncollapsed_tab <- df_uncollapsed %>% select(div_vars) %>% scale()
env_uncollapsed_tab <- df_uncollapsed %>% select(env_vars) %>% scale()

df_collapsed <- df_all %>% filter(Group == "Collapsed")
diver_collapsed_tab <- df_collapsed %>% select(div_vars) %>% scale()
env_collapsed_tab <- df_collapsed %>% select(env_vars) %>% scale()


lmm.matrix_uncollapsed <- lmm.mat.cal(diver_uncollapsed_tab, env_uncollapsed_tab, df_uncollapsed)
# lmm.matrix$Correlation[lmm.matrix$AdjPvalue >= 0.05] <- 0
lmm.matrix_collapsed <- lmm.mat.cal(diver_collapsed_tab, env_collapsed_tab, df_collapsed)
lmm.matrix_all <- rbind(data.frame(Group = rep("Uncollapsed", nrow(lmm.matrix_uncollapsed)),
                                   lmm.matrix_uncollapsed),
                        data.frame(Group = rep("Collapsed", nrow(lmm.matrix_collapsed)),
                                   lmm.matrix_collapsed)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  mutate(Diversity = factor(Diversity, levels = div_vars))
lmm.matrix_all                      
```

```{r, fig.align='center', fig.width=5.5, fig.height=2.5}
#plot
heatmap_lmm_plot <- ggplot(aes(x = Env, y = Diversity, fill = coefficents), data = lmm.matrix_all )+
  geom_tile() + scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C")+
  geom_text(aes(label = Sig), color="black", size = 3.5)+
  labs(y = NULL, x = 'Variables', fill = 'Standardized coefficients') +
  scale_y_discrete(position = "right") +
  facet_wrap(~Group, ncol = 2) +
  theme(panel.background = element_rect(fill = 'white', colour ='black'),
        #panel.grid=element_blank(), 
        axis.title = element_text(color = 'black', size = 7),
        axis.ticks.length = unit(0.2, "lines"), axis.ticks = element_line(color = 'black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6, angle = 45, hjust = 1),
        legend.title = element_text(size = 7),
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.25, "cm"),
        legend.background = element_rect(colour = "white"))
heatmap_lmm_plot
```

```{r}
partial.mantel.fun <- function(phylo) {
  env.table <- data.frame(sample_data(phylo))
  #env.table <- env.table[complete.cases(env.table), ]
  otu_table <- as.matrix(t(otu_table(phylo)))
  otu_table_hel <- decostand(otu_table, 'hellinger')
  otu_table_hel_dist <- vegdist(otu_table_hel, 'bray',upper=F)
  df <- NULL
  vars <- c("Plant_richness", "AGB", "BGB", "pH", "Soil_moisture", "Clay_Silt", "WHC", "SOC", "NH4_N", "NO3_N", "AP")
  for (x in vars) {
    y.dist <- vegdist(scale(env.table[,x]), 'euclidean', na.rm = T)
    z.dist <- vegdist(scale(env.table[ , setdiff(vars, x)]), 'euclidean', na.rm = T)
    mode <- mantel.partial(otu_table_hel_dist, y.dist, z.dist, 
                           method = "pearson", permutations = 999, na.rm = T)
    r <- mode$statistic
    p <- mode$signif
    tmp <- data.frame(env = x, r = r, p.value = p)
    if(is.null(df))
      df <- tmp
    else
      df <- rbind(df ,tmp)
  }
  return(df)
}

phylo_16s_uncollapsed <- subset_samples(phylo_16s, Group == 'Uncollapsed')
phylo_16s_uncollapsed <- prune_taxa(taxa_sums(phylo_16s_uncollapsed)>=1, phylo_16s_uncollapsed)

phylo_its_uncollapsed <- subset_samples(phylo_its, Group == 'Uncollapsed')
phylo_its_uncollapsed <- prune_taxa(taxa_sums(phylo_its_uncollapsed)>=1, phylo_its_uncollapsed)

phylo_pro_uncollapsed <- subset_samples(phylo_protist, Group == 'Uncollapsed')
phylo_pro_uncollapsed <- prune_taxa(taxa_sums(phylo_pro_uncollapsed)>=1, phylo_pro_uncollapsed)

phylo_anim_uncollapsed <- subset_samples(phylo_animal, Group == 'Uncollapsed')
phylo_anim_uncollapsed <- prune_taxa(taxa_sums(phylo_anim_uncollapsed)>=1, phylo_anim_uncollapsed)

phylo_16s_collapsed <- subset_samples(phylo_16s, Group == 'Collapsed')
phylo_16s_collapsed <- prune_taxa(taxa_sums(phylo_16s_collapsed)>=1, phylo_16s_collapsed)

phylo_its_collapsed <- subset_samples(phylo_its, Group == 'Collapsed')
phylo_its_collapsed <- prune_taxa(taxa_sums(phylo_its_collapsed)>=1, phylo_its_collapsed)

phylo_pro_collapsed <- subset_samples(phylo_protist, Group == 'Collapsed')
phylo_pro_collapsed <- prune_taxa(taxa_sums(phylo_pro_collapsed)>=1, phylo_pro_collapsed)

phylo_anim_collapsed <- subset_samples(phylo_animal, Group == 'Collapsed')
phylo_anim_collapsed <- prune_taxa(taxa_sums(phylo_anim_collapsed)>=1, phylo_anim_collapsed)


set.seed(123)
par.mant.16s.uncollapsed <- partial.mantel.fun(phylo_16s_uncollapsed)
set.seed(123)
par.mant.16s.collapsed <- partial.mantel.fun(phylo_16s_collapsed)

set.seed(123)
par.mant.its.uncollapsed <- partial.mantel.fun(phylo_its_uncollapsed)
set.seed(123)
par.mant.its.collapsed <- partial.mantel.fun(phylo_its_collapsed)

set.seed(123)
par.mant.pro.uncollapsed <- partial.mantel.fun(phylo_pro_uncollapsed)
set.seed(123)
par.mant.pro.collapsed <- partial.mantel.fun(phylo_pro_collapsed)

set.seed(123)
par.mant.anim.uncollapsed <- partial.mantel.fun(phylo_anim_uncollapsed)
set.seed(123)
par.mant.anim.collapsed <- partial.mantel.fun(phylo_anim_collapsed)
```

```{r, fig.align='center', fig.width=6.5, fig.height=5}
library(linkET)
par.man.tibble.uncollapsed <- tibble(spec = c(rep('Bacteria', nrow(par.mant.16s.uncollapsed)),
                                          rep('Fungi', nrow(par.mant.its.uncollapsed)),
                                          rep('Protist', nrow(par.mant.pro.uncollapsed)),
                                          rep('Animal', nrow(par.mant.anim.uncollapsed))),
                         rbind(par.mant.16s.uncollapsed, par.mant.its.uncollapsed, 
                              par.mant.pro.uncollapsed, par.mant.anim.uncollapsed))

par.man.tibble.collapsed <- tibble(spec = c(rep('Bacteria', nrow(par.mant.16s.collapsed)),
                                          rep('Fungi', nrow(par.mant.its.collapsed)),
                                          rep('Protist', nrow(par.mant.pro.collapsed)),
                                          rep('Animal', nrow(par.mant.anim.collapsed))),
                         rbind(par.mant.16s.collapsed, par.mant.its.collapsed, 
                              par.mant.pro.collapsed, par.mant.anim.collapsed))


vars <- c("Plant_richness", "AGB", "BGB", "pH", "Soil_moisture", "Clay_Silt", "WHC", "SOC", "NH4_N", "NO3_N", "AP")
env.table <- metadata[ , vars]
mantel_uncollapsed <- par.man.tibble.uncollapsed %>% 
  mutate(r = cut(r, breaks = c(-Inf, 0.2, 0.4, Inf), 
                 labels = c("<0.20", "0.20-0.4", ">0.40"),
                 right = FALSE),
         p.value = cut(p.value, breaks = c(-Inf, 0.001, 0.01, 0.05, Inf),
                       labels = c("<0.001", "0.001-0.01", "0.01-0.05", ">0.05"),
                       right = T))
set_corrplot_style()
p_uncollapsed <-  qcorrplot(correlate(env.table, engine = "Hmisc", method = "spearman"), type = "lower", diag = FALSE) +
  geom_square() +
  geom_couple(aes(colour = p.value, size = r), 
              data = mantel_uncollapsed, 
              curvature = nice_curvature()) +
  scale_size_manual(values = c(0.3, 1, 1.75)) +
  scale_colour_manual(values = c("#e95f5c", "#79ceb8", '#3C5488FF', 'grey')) +
  # scale_x_discrete(labels = rep_str) +
  # scale_y_discrete(labels = rep_str) +
  guides(size = guide_legend(title = "Mantel's r",
                             override.aes = list(colour = "grey35"), 
                             order = 2),
         colour = guide_legend(title = "Mantel's p", 
                               override.aes = list(size = 3), 
                               order = 1),
         fill = guide_colorbar(title = expression(paste("Spearman's ", rho), order = 3))) +
  theme(panel.grid=element_blank(), 
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6, hjust = 1),
        legend.title = element_text(size = 6), 
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.8, "lines"))

ggsave(file.path(save.dir, './figs/env_effect/partail_matel_plot_for_uncollapsed.pdf'), p_uncollapsed, width = 5, height = 2.5, units = "in")
p_uncollapsed
```

```{r, fig.align='center', fig.width=6.5, fig.height=5}
mantel_collapsed <- par.man.tibble.collapsed %>% 
  mutate(r = cut(r, breaks = c(-Inf, 0.2, 0.4, Inf), 
                 labels = c("<0.20", "0.20-0.4", ">0.40"),
                 right = FALSE),
         p.value = cut(p.value, breaks = c(-Inf, 0.001, 0.01, 0.05, Inf),
                       labels = c("<0.001", "0.001-0.01", "0.01-0.05", ">0.05"),
                       right = T))
set_corrplot_style()
p_collapsed <-  qcorrplot(correlate(env.table, engine = "Hmisc", method = "spearman"), type = "lower", diag = FALSE) +
  geom_square() +
  geom_couple(aes(colour = p.value, size = r), 
              data = mantel_collapsed, 
              curvature = nice_curvature()) +
  scale_size_manual(values = c(0.3, 1, 1.75)) +
  scale_colour_manual(values = c("#e95f5c", "#79ceb8", '#3C5488FF', 'grey')) +
  # scale_x_discrete(labels = rep_str) +
  # scale_y_discrete(labels = rep_str) +
  guides(size = guide_legend(title = "Mantel's r",
                             override.aes = list(colour = "grey35"), 
                             order = 2),
         colour = guide_legend(title = "Mantel's p", 
                               override.aes = list(size = 3), 
                               order = 1),
         fill = guide_colorbar(title = expression(paste("Spearman's ", rho), order = 3))) +
  theme(panel.grid=element_blank(), 
        axis.text.y = element_text(colour = 'black', size = 5),
        axis.text.x = element_text(colour = 'black', size = 5, hjust = 1),
        legend.title = element_text(size = 5), 
        legend.text = element_text(size = 5),
        legend.key.size = unit(0.8, "lines"))

ggsave(file.path(save.dir, './figs/env_effect/partail_matel_plot_for_collapsed.pdf'), p_collapsed, width = 5, height = 2.5, units = "in")
p_collapsed
```


Mantel test for functional genes
```{r, fig.align='center', fig.width=6.5, fig.height=5}
partial.mantel.gene.fun <- function(gene.tab, group) {
  gene_table <- as.matrix(t(gene.tab))
  gene_table_hel <- decostand(gene_table, 'hellinger')
  gene_table_hel_dist <- vegdist(gene_table_hel, 'bray', upper = F)
  vars <- c("Plant_richness", "AGB", "BGB", "pH", "Soil_moisture", "Clay_Silt", "WHC", "SOC", "NH4_N", "NO3_N", "AP")
  env.tab <- metadata %>% filter(Group == group) %>% select(vars) %>% data.frame()
  df <- NULL
  for (x in vars) {
    y.dist <- vegdist(scale(env.tab[, x]), 'euclidean', na.rm = T)
    z.dist <- vegdist(scale(env.tab[ , setdiff(vars, x)]), 'euclidean', na.rm = T)
    mode <- mantel.partial(gene_table_hel_dist, y.dist, z.dist,
                           method = "pearson", permutations = 999, na.rm = T)
    r <- mode$statistic
    p <- mode$signif
    tmp <- data.frame(env = x, r = r, p.value = p)
    if(is.null(df))
      df <- tmp
    else
      df <- rbind(df, tmp)
  }
  return(df)
}

gene.uncollapsed.tab <- ko_count_table[, grepl("_C", colnames(ko_count_table))]
gene.collapsed.tab <- ko_count_table[, grepl("_T", colnames(ko_count_table))]
par.man.gene.uncollapsed <- partial.mantel.gene.fun(gene.uncollapsed.tab, "Uncollapsed")
par.man.gene.collapsed <- partial.mantel.gene.fun(gene.collapsed.tab, "Collapsed")


par.man.tibble.gene <- tibble(spec = c(rep('Uncollapsed', nrow(par.man.gene.uncollapsed)),
                                          rep('collapsed', nrow(par.man.gene.collapsed))),
                         rbind(par.man.gene.uncollapsed, par.man.gene.collapsed))

vars <- c("Plant_richness", "AGB", "BGB", "pH", "Soil_moisture", "Clay_Silt", "WHC", "SOC", "NH4_N", "NO3_N", "AP")
env.table <- metadata[ , vars]
mantel_gene <- par.man.tibble.gene %>% 
  mutate(r = cut(r, breaks = c(-Inf, 0.2, 0.4, Inf), 
                 labels = c("<0.20", "0.20-0.4", ">0.40"),
                 right = FALSE),
         p.value = cut(p.value, breaks = c(-Inf, 0.001, 0.01, 0.05, Inf),
                       labels = c("<0.001", "0.001-0.01", "0.01-0.05", ">0.05"),
                       right = T))
set_corrplot_style()
p_gene <-  qcorrplot(correlate(env.table, engine = "Hmisc", method = "spearman"), type = "lower", diag = FALSE) +
  geom_square() +
  geom_couple(aes(colour = p.value, size = r), 
              data = mantel_gene, 
              curvature = nice_curvature()) +
  scale_size_manual(values = c(0.5, 1.5, 3)) +
  scale_colour_manual(values = c("#e95f5c", "#79ceb8", 'grey')) +
  # scale_x_discrete(labels = rep_str) +
  # scale_y_discrete(labels = rep_str) +
  guides(size = guide_legend(title = "Mantel's r",
                             override.aes = list(colour = "grey35"), 
                             order = 2),
         colour = guide_legend(title = "Mantel's p", 
                               override.aes = list(size = 3), 
                               order = 1),
         fill = guide_colorbar(title = expression(paste("Spearman's ", rho), order = 3))) +
  theme(panel.grid=element_blank(), 
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6, hjust = 1),
        legend.title = element_text(size = 6), 
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.8, "lines"))

# ggsave(file.path(save.dir, './figs/env_effect/partail_matel_plot_for_revision12.pdf'), p, width = 8.9, height = 4.5, units = "in")
p_gene
```

## variation partition analysis
```{r}

```





Using the Null model to infer the ecological processes underlying the community assembly.
```{r}
save.dir = "E:/thermokarst_gully/result/"
# extract the community and tree files from the phylo projects
comm_16s <- t(otu_table(phylo_16s))
tree_16s <- phy_tree(phylo_16s)

comm_its <- t(otu_table(phylo_its))
tree_its <- phy_tree(phylo_its)

comm_pro <- t(otu_table(phylo_protist))
tree_pro <- phy_tree(phylo_protist)

comm_anim <- t(otu_table(phylo_animal))
tree_anim <- phy_tree(phylo_animal)
```


## Read the null model tables
```{r}
icamp_16s <- read.csv(file.path(save.dir, './tables/null_model/16s_all/iCAMP.process.CbMPDiCBraya.csv'), row.names = 1, header = T, stringsAsFactors = F)

icamp_its <- read.csv(file.path(save.dir, './tables/null_model/its_all/iCAMP.process.CbMPDiCBraya.csv'), row.names = 1, header = T, stringsAsFactors = F)

icamp_pro <- read.csv(file.path(save.dir, './tables/null_model/pro_all/iCAMP.process.CbMPDiCBraya.csv'), row.names = 1, header = T, stringsAsFactors = F)

icamp_anim <- read.csv(file.path(save.dir, './tables/null_model/anim_all/iCAMP.process.CbMPDiCBraya.csv'), row.names = 1, header = T, stringsAsFactors = F)

null.df.arrange <- function(df) {
  rbind(df %>%
  filter(grepl("_C", sample1) & grepl("_C", sample2)) %>%
  pivot_longer(cols = -c(1, 2), names_to = "process", values_to = "rel_impor") %>%
  select(-c(1,2)) %>%
  mutate(Group = rep("Uncollapsed", nrow(.))),
  df %>%
    filter(grepl("_T", sample1) & grepl("_T", sample2)) %>%
  pivot_longer(cols = -c(1, 2), names_to = "process", values_to = "rel_impor") %>%
  select(-c(1,2)) %>%
  mutate(Group = rep("Collapsed", nrow(.))))
}
null.df.16s <- null.df.arrange(icamp_16s)
null.df.its <- null.df.arrange(icamp_its)
null.df.pro <- null.df.arrange(icamp_pro)
null.df.anim <- null.df.arrange(icamp_anim)
rep_str1 = c("Heterogeneous.Selection" = "Heterogeneous Selection",
            "Homogeneous.Selection" = "Homogeneous Selection",
            "Dispersal.Limitation" = "Dispersal Limitation",
            "Homogenizing.Dispersal" = "Homogenizing Dispersal",
            "Drift.and.Others" = "Drift and Others")
null.df <- rbind(data.frame(Taxa = rep("Bacteria", nrow(null.df.16s)), null.df.16s),
                 data.frame(Taxa = rep("Fungi", nrow(null.df.its)), null.df.its),
                 data.frame(Taxa = rep("Protist", nrow(null.df.pro)), null.df.pro),
                 data.frame(Taxa = rep("Animal", nrow(null.df.anim)), null.df.anim)) %>%
  mutate(process = stringr::str_replace_all(process, rep_str1)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  mutate(Taxa = factor(Taxa, levels = c("Bacteria", "Fungi", "Protist", "Animal"))) %>%
  mutate(process = factor(process, levels = c("Heterogeneous Selection", "Homogeneous Selection",
                                              "Dispersal Limitation", "Homogenizing Dispersal",
                                              "Drift and Others")))
```

```{r, fig.align='center', fig.width=8.9, fig.height=8.9}
# plot
rep_str2 <- list("Dispersal Limitation" = "DL",
            "Drift and Others" = "DR",
            "Homogeneous Selection" = "HoS",
            "Homogenizing Dispersal" = "HD",
            "Heterogeneous Selection" = "HeS"
)

# write a function to change the facet labels
facet_labeller <- function(variable,value){
  return(rep_str2[value])
}
#load the packages
pacman::p_load(ggplot2, ggpubr, gghalves)
my_comparisons_process <- list( c("Heterogeneous Selection", "Homogeneous Selection"), 
                              c('Heterogeneous Selection', 'Dispersal Limitation'), 
                              c('Heterogeneous Selection', 'Homogenizing Dispersal'),
                              c('Heterogeneous Selection', 'Drift and Others'),
                              c('Homogeneous Selection', 'Dispersal Limitation'),
                              c('Homogeneous Selection', 'Homogenizing Dispersal'),
                              c('Homogeneous Selection', 'Drift and Others'),
                              c('Dispersal Limitation', 'Homogenizing Dispersal'),
                              c('Dispersal Limitation', 'Drift and Others'),
                              c('Homogenizing Dispersal', 'Drift and Others'))
process_within_group <- null.df %>%
  ggplot(aes(process, rel_impor * 100, fill = process)) +
  geom_half_violin(position = position_nudge(x = 0.2), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.3, size = 0.75, outlier.color = NA) +
  # geom_jitter(aes(fill = process), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_x_discrete(labels = rep_str2) + 
  facet_grid(Taxa ~ Group, scales = 'free') +
  stat_compare_means(comparisons = my_comparisons_process, 
                     method = "wilcox.test", p.adjust.method = "BH", 
                     bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00) +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00", "#1F77B4", "#AEC7E8", "#FF7F0E", "#FFBB78")) +
  theme_bw() +
  theme(axis.title = element_text(colour = "black", size = 14),
        axis.text.x = element_text(colour = "black", size = 12),
        axis.text.y = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        legend.key.size = unit(1,"line"),
        panel.grid = element_blank(),
        panel.spacing = unit(0, "lines"),
        legend.position = "none")
process_within_group
```
为了出图
```{r, fig.align='center', fig.width=4.5, fig.height=4.5}
process_within_group_1 <- null.df %>%
  ggplot(aes(process, rel_impor * 100, fill = process)) +
  # geom_half_violin(position = position_nudge(x = 0.2), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.3, size = 0.75, lwd = 0.2, outlier.color = NA) +
  # geom_jitter(aes(fill = process), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_x_discrete(labels = rep_str2) + 
  facet_grid(Taxa ~ Group, scales = 'free') +
  labs(x = 'Ecological processes', y = 'Relative importance (%)') +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00", "#1F77B4", "#AEC7E8", "#FF7F0E", "#FFBB78")) +
  theme_bw() +
  theme(axis.title = element_text(colour = "black", size = 6),
        axis.text.x = element_text(colour = "black", size = 6),
        axis.text.y = element_text(colour = "black", size = 6),
        strip.text = element_text(colour = "black", size = 6),
        axis.ticks.length = unit(0.1, "lines"),
        legend.key.size = unit(1,"line"),
        panel.grid = element_blank(),
        panel.spacing = unit(0, "lines"),
        legend.position = "none")
process_within_group_1
```

```{r, fig.align='center', fig.width=8.9, fig.height=8.9}
#boxplot
library(ggpubr)
my_comparisons_group <- list(c('Uncollapsed', 'Collapsed'))
process_among_group <- null.df %>%
  ggplot(aes(Group, rel_impor * 100)) +
  geom_half_violin(position = position_nudge(x = 0.2), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.3, size = 0.75, outlier.color = NA) +
  # geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  scale_x_discrete(labels = rep_str2) + 
  facet_grid(Taxa ~ process, scales = 'free') +
  stat_compare_means(comparisons = my_comparisons_group, 
                     method = "wilcox.test", p.adjust.method = "BH",
                     bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00) +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00", "#1F77B4", "#AEC7E8", "#FF7F0E", "#FFBB78")) +
  theme_bw() +
  theme(axis.title = element_text(colour = "black", size = 14),
        axis.text.x = element_text(colour = "black", size = 12, 
                                   angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        legend.key.size = unit(1,"line"),
        panel.grid = element_blank(),
        panel.spacing = unit(0, "lines"),
        legend.position = "none")
process_among_group
```


### Determine the effect size of the permafrost collapsed on the ecological processes using wilcox test
```{r}
null.df %>%
  group_by(Taxa, process, Group) %>%
  get_summary_stats(rel_impor, type = "mean_sd") %>%
  print(n = 40)

wilcox_eff_df <- null.df %>%
  group_by(Taxa, process) %>% 
  wilcox_effsize(rel_impor ~ Group, ref.group = "Uncollapsed", ci = T) %>%
  arrange(factor(process, levels = c("Heterogeneous Selection", "Homogeneous Selection",
                                              "Dispersal Limitation", "Homogenizing Dispersal",
                                              "Drift and Others"))) %>%
  arrange(factor(Taxa, levels = c("Bacteria", "Fungi", "Protist", "Animal")))

#add a coefficients and significance to indicate the positive and negative variations according to the increasing/decreasing relative importance determined by the wilcox test
wilcox_eff_df <- wilcox_eff_df %>% 
  mutate(coeff = c(1, -1, 1, -1, -1, 1, -1, 1, -1, -1, 1, 1, 1, -1, -1, 1, 1, 1, -1, -1)) %>%
  mutate(sig = c("**", "***", "***", "***", "***",
                 "***", "***", "***", "***", "***",
                 "", "***", "***", "***", "***",
                 "***", "***", "***", "", "***"))
wilcox_eff_df
```

```{r, fig.align='center', fig.width=2.5, fig.height=6}
wilcox_eff_plot <- wilcox_eff_df %>%
  select(Taxa, process, effsize, conf.low, conf.high, coeff, sig) %>%
  mutate(effsize = effsize * coeff) %>%
  mutate(conf.low = conf.low * coeff) %>%
  mutate(conf.high = conf.high * coeff) %>%
  as.data.frame() %>%
  ggplot(aes(x = process, y = effsize, colour = process)) +
  geom_hline(aes(yintercept = 0), size = 0.7,  colour = "gray2")+
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = conf.low, ymax = conf.high), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.6) +
  geom_text(aes(label = sig, x = process, y = (effsize/abs(effsize))*(abs(conf.high)*1.35)),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Wilcox effect size") +
  facet_grid(Taxa ~., ) +
  scale_color_manual(values=c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00", "#1F77B4")) +
  scale_x_discrete(position = "bottom", labels = rep_str2) + 
  scale_y_continuous(expand = c(0, 0), limit = c(-1, 1)) +
  theme_bw() + coord_flip() +
  theme(legend.position = "none", 
        panel.grid=element_blank(), 
        axis.title = element_text(color = 'black', size = 6),
        axis.ticks.length = unit(0.1, "lines"), axis.ticks = element_line(color = 'black', size = 0.3),
        axis.line = element_line(colour = "black", size = 0.1), 
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6, hjust = 1),
        panel.spacing = unit(0, "lines"),
        strip.text = element_text(colour = 'black', size = 6),
        legend.title = element_text(size = 6),
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.2, 'cm'),
        legend.background = element_rect(colour = "white"))
wilcox_eff_plot
```

```{r}
null_plot <- cowplot::plot_grid(process_within_group_1, wilcox_eff_plot, ncol = 2, align = "h")
ggsave(file.path(save.dir, './figs/null_model/process_comparison.pdf'), null_plot, width = 6, height = 3.5)
null_plot
```


```{r}
save.dir = "E:/thermokarst_gully/result/"
if(!dir.exists(save.dir)){dir.create(save.dir)}
# prune the phylo object into uncollapsed and collapsed groups
comm_16s_uncollapsed <- t(otu_table(phylo_16s_uncollapsed))
tree_16s_uncollapsed <- phy_tree(phylo_16s_uncollapsed)

comm_its_uncollapsed <- t(otu_table(phylo_its_uncollapsed))
tree_its_uncollapsed <- phy_tree(phylo_its_uncollapsed)

comm_pro_uncollapsed <- t(otu_table(phylo_pro_uncollapsed))
tree_pro_uncollapsed <- phy_tree(phylo_pro_uncollapsed)

comm_anim_uncollapsed <- t(otu_table(phylo_anim_uncollapsed))
tree_anim_uncollapsed <- phy_tree(phylo_anim_uncollapsed)


comm_16s_collapsed <- t(otu_table(phylo_16s_collapsed))
tree_16s_collapsed <- phy_tree(phylo_16s_collapsed)

comm_its_collapsed <- t(otu_table(phylo_its_collapsed))
tree_its_collapsed <- phy_tree(phylo_its_collapsed)

comm_pro_collapsed <- t(otu_table(phylo_pro_collapsed))
tree_pro_collapsed <- phy_tree(phylo_pro_collapsed)

comm_anim_collapsed <- t(otu_table(phylo_anim_collapsed))
tree_anim_collapsed <- phy_tree(phylo_anim_collapsed)
```

```{r,message=FALSE}
setwd(save.dir)
library(iCAMP)
set.seed(123)
icamp_16s_uncollapsed <- icamp.big(comm = comm_16s_uncollapsed, tree = tree_16s_uncollapsed, pd.wd = paste0(save.dir,"tables/null_model/16s_uncollapsed"), 
                       ses.cut = 1.96, rc.cut = 0.95, bin.size.limit = 24, 
                       rand = 1000, nworker = 8)
head(icamp_16s_uncollapsed$CbMPDiCBraya)
# write.csv(icamp_16s_uncollapsed$CbMPDiCBraya,
#           file.path(save.dir, './tables/null_model/16s_uncollapsed/iCAMP.process.CbMPDiCBraya.csv'))

set.seed(123)
icamp_its_uncollapsed <- icamp.big(comm = comm_its_uncollapsed, tree = tree_its_uncollapsed, pd.wd = paste0(save.dir,"tables/null_model/its_uncollapsed"), 
                       ses.cut = 1.96, rc.cut = 0.95, bin.size.limit = 24, 
                       rand = 1000, nworker = 8)
head(icamp_its_uncollapsed$CbMPDiCBraya)
# write.csv(icamp_its_uncollapsed$CbMPDiCBraya,
          # file.path(save.dir, './tables/null_model/its_uncollapsed/iCAMP.process.CbMPDiCBraya.csv'))


set.seed(123)
icamp_pro_uncollapsed <- icamp.big(comm = comm_pro_uncollapsed, tree = tree_pro_uncollapsed, pd.wd = paste0(save.dir,"tables/null_model/pro_uncollapsed"), 
                       ses.cut = 1.96, rc.cut = 0.95, bin.size.limit = 24, 
                       rand = 1000, nworker = 8)
head(icamp_pro_uncollapsed$CbMPDiCBraya)
# write.csv(icamp_pro_uncollapsed$CbMPDiCBraya,
#           file.path(save.dir, './tables/null_model/pro_uncollapsed/iCAMP.process.CbMPDiCBraya.csv'))

set.seed(123)
icamp_anim_uncollapsed <- icamp.big(comm = comm_anim_uncollapsed, tree = tree_anim_uncollapsed, pd.wd = paste0(save.dir,"tables/null_model/anim_uncollapsed"), 
                       ses.cut = 1.96, rc.cut = 0.95, bin.size.limit = 24, 
                       rand = 1000, nworker = 8)
head(icamp_anim_uncollapsed$CbMPDiCBraya)
# write.csv(icamp_anim_uncollapsed$CbMPDiCBraya,
#           file.path(save.dir, './tables/null_model/anim_uncollapsed/iCAMP.process.CbMPDiCBraya.csv'))

set.seed(123)
icamp_16s_collapsed <- icamp.big(comm = comm_16s_collapsed, tree = tree_16s_collapsed, pd.wd = paste0(save.dir,"tables/null_model/16s_collapsed"), 
                       ses.cut = 1.96, rc.cut = 0.95, bin.size.limit = 24, 
                       rand = 1000, nworker = 8)
head(icamp_16s_collapsed$CbMPDiCBraya)
# write.csv(icamp_16s_collapsed$CbMPDiCBraya,
#           file.path(save.dir, './tables/null_model/16s_collapsed/iCAMP.process.CbMPDiCBraya.csv'))

set.seed(123)
icamp_its_collapsed <- icamp.big(comm = comm_its_collapsed, tree = tree_its_collapsed, pd.wd = paste0(save.dir,"tables/null_model/its_collapsed"), 
                       ses.cut = 1.96, rc.cut = 0.95, bin.size.limit = 24, 
                       rand = 1000, nworker = 8)
head(icamp_its_collapsed$CbMPDiCBraya)
# write.csv(icamp_its_collapsed$CbMPDiCBraya,
#           file.path(save.dir, './tables/null_model/its_collapsed/iCAMP.process.CbMPDiCBraya.csv'))

set.seed(123)
icamp_pro_collapsed <- icamp.big(comm = comm_pro_collapsed, tree = tree_pro_collapsed, pd.wd = paste0(save.dir,"tables/null_model/pro_collapsed"), 
                       ses.cut = 1.96, rc.cut = 0.95, bin.size.limit = 24, 
                       rand = 1000, nworker = 8)
head(icamp_pro_collapsed$CbMPDiCBraya)
# write.csv(icamp_pro_collapsed$CbMPDiCBraya,
#           file.path(save.dir, './tables/null_model/pro_collapsed/iCAMP.process.CbMPDiCBraya.csv'))

set.seed(123)
icamp_anim_collapsed <- icamp.big(comm = comm_anim_collapsed, tree = tree_anim_collapsed, pd.wd = paste0(save.dir,"tables/null_model/anim_collapsed"), 
                       ses.cut = 1.96, rc.cut = 0.95, bin.size.limit = 24, 
                       rand = 1000, nworker = 8)
head(icamp_anim_collapsed$CbMPDiCBraya)
# write.csv(icamp_anim_collapsed$CbMPDiCBraya,
#           file.path(save.dir, './tables/null_model/anim_collapsed/iCAMP.process.CbMPDiCBraya.csv'))
```

