---
title: "Functional gene analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 300)
```

## Data input
Set work directory
```{r}
setwd('e:/thermokarst_gully/')
wd_16s <- file.path(getwd(),"data/16S/rdp")
# if (!dir.exists(wd_16s)) {
#   dir.create(wd_16s)
# }
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```

Loading packages
```{r}
library(tidyverse)
library(ggrepel)
library(EnhancedVolcano)
library(edgeR)
library(DESeq2)
```

Data input
```{r}
source("script/read_data_rdp.R")
```

## 1. edgeR analysis. reference:https://www.reneshbedre.com/blog/edger-tutorial.html
```{r}
#The field in the class definition file that defines the classes of the data.
data_classes <- "Group"

# Load Expression Data
RNASeq <- ko_count_table

# Load group information
RNASeq[1:5, 1:5]
classDefinitions_RNASeq <- metadata
classDefinitions_RNASeq[1:5, 1:4]

# Filter Data (RNA-seq read counts are converted to CPM values 
# and genes with CPM > 1 in at least 50 of the samples are 
# retained for further study, a gene mush have at least 50 measurements 
# with more than 1 CPM in one of the classes to be included in the analysis)
cpms <- cpm(RNASeq)
keep <- rowSums(cpms > 1) >= 1
counts <- RNASeq[keep,]

# Normalization and Dispersion
# create data structure to hold counts and group information for each sample.
d <- DGEList(counts = counts, group =  factor(classDefinitions_RNASeq$Group))
# Filter out the genes with low counts
# #filterByExpr function to remove the low count genes. 
# #This function keeps the genes with a worthwhile counts (at least 10 read counts) in a minimal number of samples.
# keep <- filterByExpr(y = d)
# d <- d[keep, , keep.lib.sizes=FALSE]
# Normalization and effective library size
d <- calcNormFactors(d)
# Model fitting and estimating dispersions
d <- estimateCommonDisp(d)
d <- estimateTagwiseDisp(d)

# Testing for differential gene expression
et <- exactTest(object = d)
# topTags() function is useful to extract the table with adjusted p values (FDR).
top_degs <- topTags(object = et, n = "Inf")
top_degs
#Get a summary DGE table (returns significant genes with absolute log fold change at least 1 and adjusted p value < 0.05)
summary(decideTests(object = et, lfc = 1))
# write.csv(as.data.frame(top_degs), file = file.path(save.dir, "tables/kegg_Collapsed_vs_control_edge.csv"))
```

DGE Visualization
```{r}
# Create a volcano plot
## set the plot theme
main_theme = theme_linedraw() + 
  theme(panel.grid=element_blank(), 
        strip.text = element_text(colour = 'black', size = 12),
        strip.background = element_rect(colour = 'grey', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color = 'black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour = 'black', size = 12),
        axis.text.x = element_text(colour = 'black', size = 12),
        legend.position = c(0.15, 0.9),
        legend.title = element_blank(),
        legend.text = element_text(size = 12),
        legend.key = element_blank(),
        legend.background = element_rect(colour = "white"))
data <- top_degs %>% as.data.frame() %>% 
  mutate(Expression = case_when(logFC >= log(2) & FDR <= 0.05 ~ "Up-regulated",
                           logFC <= -log(2) & FDR <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged"))
ggplot(data, aes(logFC, -log(FDR,10))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("log"[2]*"FC")) + 
  ylab(expression("-log"[10]*"FDR")) +
  scale_color_manual(values = c("dodgerblue3", "gray50", "firebrick3")) +
  guides(colour = guide_legend(override.aes = list(size=1.5))) +
  main_theme +
  theme(legend.background = element_blank(),
        legend.key = element_blank())
```

```{r}
data <- top_degs %>% as.data.frame() %>%
  mutate(
    Significance = case_when(
      abs(logFC) >= log(2) & FDR <= 0.05 & FDR > 0.01 ~ "FDR 0.05", 
      abs(logFC) >= log(2) & FDR <= 0.01 & FDR > 0.001 ~ "FDR 0.01",
      abs(logFC) >= log(2) & FDR <= 0.001 ~ "FDR 0.001", 
      TRUE ~ "Unchanged"))

ggplot(data, aes(logFC, -log(FDR,10))) +
  geom_point(aes(color = Significance), size = 2/5) +
  xlab(expression("log"[2]*"FC")) + 
  ylab(expression("-log"[10]*"FDR")) +
  scale_color_viridis_d() +
  guides(colour = guide_legend(override.aes = list(size=1.5))) +
  main_theme +
  theme(legend.background = element_blank(),
        legend.key = element_blank())
```


## 2. DESeq2 analysis
```{r}
group_df <- metadata
rownames(group_df) <- metadata$Sample_name # make sure the colnames of count table match the rownames of group_df.
group_df$Group <- factor(group_df$Group, levels = c("Control", "Collapsed"))
KEGG_dds <- DESeqDataSetFromMatrix(countData = round(ko_count_table + 1), 
                                   colData = group_df, design = ~ Group)
KEGG_deseq <- DESeq(KEGG_dds)
KEGG_res <- results(KEGG_deseq, contrast = c("Group", 'Collapsed', 'Control'))
KEGG_res$padj[is.na(KEGG_res$padj)] = 1
KEGG_significant = rownames(KEGG_res)[(KEGG_res$padj < 0.05) & (KEGG_res$log2FoldChange > 1)]
length(KEGG_significant)
write.csv(as.data.frame(KEGG_res), file = file.path(save.dir, "tables/kegg_Collapsed_vs_control_deseq.csv"))
# Volcano plot
data <- KEGG_res %>% as.data.frame() %>% 
  mutate(
    Expression = case_when(log2FoldChange >= log(2) & padj <= 0.05 ~ "Up-regulated",
                           log2FoldChange <= -log(2) & padj <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged")
  )
ggplot(data, aes(log2FoldChange, -log(padj,10))) +
  geom_point(aes(color = Expression), size = 2/5) +
  xlab(expression("log"[2]*"FC")) + 
  ylab(expression("-log"[10]*"padj")) +
  scale_color_manual(values = c("dodgerblue3", "gray50", "firebrick3")) +
  guides(colour = guide_legend(override.aes = list(size=1.5))) 


EnhancedVolcano(KEGG_res,
                lab = rownames(KEGG_res),
                pCutoff = 0.05,
                FCcutoff = 1,
                col = c("grey", "grey30", "grey30", "#23A671"),
                x = 'log2FoldChange',
                title = NULL,
                subtitle = NULL,
                caption = NULL,
                y = 'padj')
```