---
title: "Functional gene analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 300)
```

## Data input
Set work directory
```{r}
setwd('e:/thermokarst_gully/')
wd_16s <- file.path(getwd(),"data/16S/rdp")
# if (!dir.exists(wd_16s)) {
#   dir.create(wd_16s)
# }
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```

Loading packages
```{r}
pacman::p_load(tidyverse, ggrepel, EnhancedVolcano, edgeR, DESeq2)
```

Data input
```{r}
source("script/read_data_rdp.R")
```

## Differential analysis of functional gene profiles
### 1. edgeR analysis.
Reference:https://www.reneshbedre.com/blog/edger-tutorial.html
```{r}
#The field in the class definition file that defines the classes of the data.
data_classes <- "Group"

# Load Expression Data
RNASeq <- ko_count_table

# Load group information
RNASeq[1:5, 1:5]
classDefinitions_RNASeq <- metadata
classDefinitions_RNASeq[1:5, 1:4]

# Filter Data (RNA-seq read counts are converted to CPM values 
# and genes with CPM > 1 in at least 50 of the samples are 
# retained for further study, a gene mush have at least 50 measurements 
# with more than 1 CPM in one of the classes to be included in the analysis)
cpms <- cpm(RNASeq)
keep <- rowSums(cpms > 1) >= 1
counts <- RNASeq[keep,]

# Normalization and Dispersion
# create data structure to hold counts and group information for each sample.
d <- DGEList(counts = counts, group =  factor(classDefinitions_RNASeq$Group))
# Filter out the genes with low counts
# #filterByExpr function to remove the low count genes. 
# #This function keeps the genes with a worthwhile counts (at least 10 read counts) in a minimal number of samples.
# keep <- filterByExpr(y = d)
# d <- d[keep, , keep.lib.sizes=FALSE]
# Normalization and effective library size
d <- calcNormFactors(d)
# Model fitting and estimating dispersions
d <- estimateCommonDisp(d)
d <- estimateTagwiseDisp(d)

# Testing for differential gene expression
et <- exactTest(object = d)
# topTags() function is useful to extract the table with adjusted p values (FDR).
top_degs <- topTags(object = et, n = "Inf")
top_degs
#Get a summary DGE table (returns significant genes with absolute log fold change at least 1 and adjusted p value < 0.05)
summary(decideTests(object = et, lfc = 1))
# write.csv(as.data.frame(top_degs), file = file.path(save.dir, "tables/kegg_Collapsed_vs_control_edge.csv"))
```

DGE Visualization
```{r, fig.align='center', fig.width=5.5, fig.height=4}
# Create a volcano plot
## set the plot theme
main_theme = theme_linedraw() + 
  theme(panel.grid=element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 10),
        strip.background = element_rect(colour = 'grey', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 12),
        axis.ticks = element_line(color = "black", linewidth = 1),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour = 'black', size = 10),
        axis.text.x = element_text(colour = 'black', size = 10),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        legend.key = element_blank(),
        legend.background = element_blank())
data_edger <- top_degs %>% as.data.frame() %>% 
  mutate(Expression = case_when(logFC >= log(2) & FDR <= 0.05 ~ "Up-regulated",
                           logFC <= -log(2) & FDR <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged"))
volca_plot_edger <- ggplot(data_edger, aes(logFC, -log(FDR,10))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("log"[2]*"FC")) + 
  ylab(expression("-log"[10]*"FDR")) +
  scale_color_manual(values = c("#79ceb8", "gray70", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  guides(colour = guide_legend(override.aes = list(size=1.5))) +
  main_theme +
  theme(legend.background = element_blank(),
        legend.key = element_blank(),
        legend.position = c(0.15, 0.85))
volca_plot_edger
```

### 2. DESeq2 analysis
```{r, fig.align='center', fig.width=5.5, fig.height=4}
group_df <- metadata
rownames(group_df) <- metadata$Sample_name # make sure the colnames of count table match the rownames of group_df.
group_df$Group <- factor(group_df$Group, levels = c("Control", "Collapsed"))
KEGG_dds <- DESeqDataSetFromMatrix(countData = round(ko_count_table + 1), 
                                   colData = group_df, design = ~ Group)
KEGG_deseq <- DESeq(KEGG_dds)
KEGG_res <- results(KEGG_deseq, contrast = c("Group", 'Collapsed', 'Control'))
KEGG_res$padj[is.na(KEGG_res$padj)] = 1
KEGG_significant = rownames(KEGG_res)[(KEGG_res$padj < 0.05) & (KEGG_res$log2FoldChange > 1)]
# write.csv(as.data.frame(KEGG_res), file = file.path(save.dir, "tables/kegg_Collapsed_vs_control_deseq.csv"))
# Volcano plot
data_deseq2 <- KEGG_res %>% as.data.frame() %>% 
  mutate(
    Expression = case_when(log2FoldChange >= log(2) & padj <= 0.05 ~ "Up-regulated",
                           log2FoldChange <= -log(2) & padj <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged")
  )
volca_plot_deseq2 <- ggplot(data_deseq2, aes(log2FoldChange, -log(padj,10))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("log"[2]*"FC")) + 
  ylab(expression("-log"[10]*"padj")) +
  scale_color_manual(values = c("#79ceb8", "gray70", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  guides(colour = guide_legend(override.aes = list(size = 1.5))) +
  main_theme +
  theme(legend.background = element_blank(),
        legend.key = element_blank(),
        legend.position = c(0.15, 0.85))
volca_plot_deseq2
```

## Extract the Up-regulated and Down-regulated genes within data_edger data
```{r}
up_KOs <- data_edger %>% filter(Expression %in% "Up-regulated") %>%
  rownames(.)
down_KOs <- data_edger %>% filter(Expression %in% "Down-regulated") %>%
  rownames(.)

ko_9_levels <- read.delim(file.path(wd_fun, "/ko_9_levels.txt"), header = T, sep = "\t")

up_df <- data_edger %>% 
  mutate(KO = rownames(.)) %>%
  filter(Expression %in% "Up-regulated") %>%
  inner_join(ko_9_levels, ky = "KO") %>%
  select("KO", "Description", "L2", "L3", "PathwayID") %>%
  left_join(cbind(KO = rownames(ko_tpm_table), ko_tpm_table), by = "KO")

down_df <- data_edger %>% 
  mutate(KO = rownames(.)) %>%
  filter(Expression %in% "Down-regulated") %>%
  inner_join(ko_9_levels, ky = "KO") %>%
  select("KO", "Description", "L2", "L3", "PathwayID") %>%
  left_join(cbind(KO = rownames(ko_tpm_table), ko_tpm_table), by = "KO")

```

```{r}


