---
title: "Functional gene analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 300)
```

## Data input
Set work directory
```{r}
setwd('e:/thermokarst_gully/')
wd_16s <- file.path(getwd(),"data/16S/rdp")
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```

Loading packages
```{r}
pacman::p_load(tidyverse, ggrepel, EnhancedVolcano, edgeR, ggplot2)
```

Data input
```{r}
source("script/read_data_all.R")
```

Permutational multivariate analysis of variance using distance matrices (adonis) to test the difference of functional composition
```{r}
library(vegan)
# determine the dissimilarity matrix based on the bray-curties distance
fun_dist <-vegdist(t(ko_tpm_table), "bray" )
# permanova, ANOSIM and MRPP analysis
adonis2(fun_dist ~ Group, data = metadata)
mrpp(fun_dist, metadata$Group, perm = 999)
anosim(fun_dist, metadata$Group, perm = 999)
```

PCoA plot
```{r}
# taxa
ord.fun <-  cmdscale(fun_dist,  k = 2, eig = T, add = T)
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 6),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 6),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.position = "none")
pcoa_fun_p1 <- data.frame(Group = metadata$Group, scores(ord.fun)) %>%
  mutate(Group = factor(Group, levels = c('Uncollapsed', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_color_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  # scale_x_continuous(expand = c(0.03, 0.03)) +
  # scale_y_continuous(expand = c(0.03, 0.03)) +
  labs(x = paste("PCoA1 (", format(100 * ord.fun$eig[1] / sum(ord.fun$eig), digits = 3), "%)", sep = ""),
       y = paste("PCoA2 (", format(100 * ord.fun$eig[2] / sum(ord.fun$eig), digits = 3), "%)", sep = "")) +
  main_theme +
  theme(legend.background = element_blank(),
        legend.title = element_text(size = 6),
        legend.text = element_text(size = 6),
        legend.key = element_blank(),
        legend.position = c(0.85, 0.85),
        legend.key.size = unit(0.4, 'cm'))
# creat a directory for kegg-gene results
if (!dir.exists(file.path(save.dir, "figs/kegg/"))) {
  dir.create(file.path(save.dir, "figs/kegg/"))
}
# ggsave(file.path(save.dir, "./figs/kegg/PCoA_fun_bray.pdf"),
#        pcoa_fun_plot, width = 3.5, height = 2.5, units = "in")
pcoa_fun_p1
```
## Determine the funtional biotic homogenization
```{r}
library(vegan)
#determine the dissimilarity matrix based on the bray-curties distance
fun_dist <-vegdist(t(ko_count_table), "bray")
#permanova test the difference in compositional variance
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
adonis2(fun_dist ~ Group, data = metadata)
similar_fun_data <- lapply(vars, function(x) usedist::dist_subset(fun_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Uncollapsed', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Uncollapsed', 'Collapsed')))
```

```{r, fig.align='center', fig.height=3.5, fig.width=3.5}
library(lme4)
library(lmerTest)
# determine the effect size of the permafrost thawing for the diversity indexes
dist_scale <- similar_fun_data %>% 
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))


fm1 <- lmer(distance ~ Group + (1 | Gully_id), data = dist_scale)
presult <- car::Anova(fm1, type = 2)
coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
names(coefs) <- paste0(names(coefs), ".mean")
SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
names(SEvalues) <- paste0(names(SEvalues), ".se")
tvalues <- coef(summary(fm1))[, "t value"]  ##t values
names(tvalues) <- paste0(names(tvalues), ".t")
chisqP <- c(presult[, 1], presult[, 3])
names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))
result <- data.frame(variables = "Function", GroupCollapsed.mean = coefs[2], GroupCollapsed.se = SEvalues[2], sig = "***")


fun_dist_comparison <-  ggplot(data = result, aes(x = variables, y = GroupCollapsed.mean, colour = "#e95f5c")) +
  geom_hline(aes(yintercept = 0), size = 0.7,  colour = "gray2")+
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "bottom") +
  theme(legend.position = "none", 
        panel.grid=element_blank(), 
        axis.title = element_text(color = 'black', size = 6),
        axis.ticks.length = unit(0.1, "lines"), axis.ticks = element_line(color = 'black', size = 0.3),
        axis.line = element_line(colour = "black", size = 0.1), 
        axis.text.y = element_blank(),
        axis.text.x = element_text(colour = 'black', size = 6, hjust = 1),
        legend.title = element_text(size = 7),
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.2, 'cm'),
        legend.background = element_rect(colour = "white"))

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir.multifunc, "./single_div_comparison.pdf"),
#        single_div_comparison, width = 2.7, height = 5, units = "in")
fun_dist_comparison <- ggplotGrob(fun_dist_comparison)
pcoa_fun_plot <- pcoa_fun_p1 + 
  annotation_custom(fun_dist_comparison, 
                    xmin = 0.06, xmax = 0.14, 
                    ymin = -0.075, ymax = -0.03)
pcoa_fun_plot
```

## Differential analysis of functional gene profiles
### 1. edgeR analysis.
Reference:https://www.reneshbedre.com/blog/edger-tutorial.html
```{r}
library(edgeR)
#The field in the class definition file that defines the classes of the data.
data_classes <- "Group"

# Load Expression Data
RNASeq <- ko_count_table

# Load group information
RNASeq[1:5, 1:5]
classDefinitions_RNASeq <- metadata
classDefinitions_RNASeq[1:5, 1:4]

# Filter Data (RNA-seq read counts are converted to CPM values 
# and genes with CPM > 1 are retained for further study)
cpms <- cpm(RNASeq)
keep <- rowSums(cpms > 1) >= 1
counts <- RNASeq[keep,]

# Normalization and Dispersion
# create data structure to hold counts and group information for each sample.
d <- DGEList(counts = counts, group =  factor(classDefinitions_RNASeq$Group))
# Filter out the genes with low counts
# #filterByExpr function to remove the low count genes. 
# #This function keeps the genes with a worthwhile counts (at least 10 read counts) in a minimal number of samples.
# keep <- filterByExpr(y = d)
# d <- d[keep, , keep.lib.sizes=FALSE]
# Normalization and effective library size
d <- calcNormFactors(d)
# Model fitting and estimating dispersions
d <- estimateCommonDisp(d)
d <- estimateTagwiseDisp(d)

# Testing for differential gene expression
et <- exactTest(object = d, pair = c("Uncollapsed","Collapsed"), prior.count=0)
# topTags() function is useful to extract the table with adjusted p values (FDR).
top_degs <- topTags(object = et, n = "Inf")
top_degs
#Get a summary DGE table (returns significant genes with absolute log fold change at least 1 and adjusted p value < 0.05)
summary(decideTests(object = et, lfc = 1))
# if (!dir.exists(file.path(save.dir, "tables/kegg/"))) {
#   dir.create(file.path(save.dir, "tables/kegg/"))
# }
# write.csv(as.data.frame(top_degs), file = file.path(save.dir, "tables/kegg/kegg_Collapsed_vs_control_edge.csv"))
```


```{r, fig.align='center', fig.width=7.5, fig.height=6}
library(cowplot)
fun_plot <- ggdraw() +
  draw_plot(single_fun_div_comparison, x = 0, y = 0, width = 2/5, height = 1) +
  draw_plot(pcoa_fun_p1, x = 2/5, y = 3.5/5, width = 1.3/5, height = 1.5/5) +
  draw_plot(volca_plot_edger, x = 3.5/5, y = 3.5/5, width = 1.3/5, height = 1.5/5) +
  draw_plot(kegg_heatmap, x = 1.2/5, y = 0, width = 3.7/5, height = 3.5/5) +
  draw_plot_label(label = c("a", "b", 'c', "d"), size = 8,
                  x = c(0, 2/5, 3.5/5, 2/5), y = c(1, 1, 1, 3.5/5))


# ggsave(file.path(save.dir, "/figs/kegg/fun_plot.pdf"),
#        fun_plot, width = 7.5, height = 6, units = "in")
fun_plot
```

## Explore the losted genes after permafrost thawing
```{r}
# library
library(ggvenn)
x <- list(
  Control = ko_tpm_table %>% data.frame() %>%
    mutate(rowsum = rowSums(select(., grep('_C', metadata$Sample_name, value = T)))) %>%
    filter(rowsum > 0) %>%
    rownames(), 
  Collapsed = ko_tpm_table %>% data.frame() %>%
    mutate(rowsum = rowSums(select(., grep('_T', metadata$Sample_name, value = T)))) %>%
    filter(rowsum > 0) %>%
    rownames()
  )
# plot
fun.venn <- ggvenn(
  x, 
  fill_color = c("#f8766d", "#a3a500", "#00b0f6"),
  stroke_color = NA,
  set_name_size = 4,
  text_size = 4,
  show_percentage = F
)
fun.venn
```


```{r}
#loading packages
library(tidyverse)
#reading data
KO_L2_file <- 'E:/thermokarst_gully/data/metagenome/KEGG.PathwayL2.raw.txt'

KO_L2_df <- read.table(KO_L2_file, header = TRUE, sep = "\t", 
                  as.is = TRUE, stringsAsFactors = FALSE, comment.char = "",
                  check.names = FALSE)

# Using dplyr to process the data
aggre_data <- KO_L2_df %>%
  # Add a column for row means (excluding the L2 column)
  mutate(RowMean = rowMeans(dplyr::select(., -PathwayL2))) %>%
  # Arrange rows by RowMean in descending order
  arrange(desc(RowMean)) %>%
  # Add a new column to classify rows as "Top 10" or "Other"
  mutate(Pathway = if_else(dplyr::row_number(.) <= 15, as.character(PathwayL2), "Other")) %>%
  # Group by "Pathway" and summarize
  group_by(Pathway) %>%
  dplyr::summarize(., across(starts_with("G"), sum)) %>% # Sum numerical columns 
  # Retain row mean for ordering purposes
  mutate(RowMean = rowMeans(select(., -Pathway))) %>%
  # Arrange by RowMean in descending order, placing "Other" last
  mutate(PathwayOrder = if_else(Pathway == "Other", -Inf, RowMean)) %>%
  arrange(desc(PathwayOrder)) %>%
  # factor(Pathway = factor(Pathway, levels = Pathway)) %>% # Order the Pathway
  select(-PathwayOrder, -RowMean)  # Drop intermediate columns

# Pull out the pathway for ordering purposes
pathway_order <- aggre_data %>% pull(Pathway)

##Stack plot using ggplot2 
pathway_plot <- aggre_data %>% 
  pivot_longer(-Pathway, names_to = 'Sample_id', values_to = 'TPM') %>%
  mutate(Pathway = factor(Pathway, ordered = T, levels = pathway_order),
         Group = case_when(grepl("_C", Sample_id) ~ "Uncollapsed",
                           grepl("_T", Sample_id) ~ "Collapsed"),
         Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  mutate(sample_id = rep(1:60, 16)) %>%
  group_by(Group, Sample_id) %>%
  mutate(prop = (TPM * 100 / sum(TPM))) %>%
  ggplot(aes(x = sample_id, y = prop, fill = Pathway)) +
  geom_area() +
  labs(x = 'Sample', y = 'Proportion (%)') +
  scale_fill_manual(values = brewer.accent(18)) +
  scale_y_continuous(expand = c(0, 0)) +
  facet_grid(~ Group, scales = 'free_x', space = 'free_x') +
  main_theme +
  theme(legend.position = "none",
        legend.key.size = unit(0.5, "lines"),
        panel.spacing = unit(0, "lines"))
pathway_plot
```

```{r, fig.align='center', fig.width=8, fig.height=5}
library(cowplot)
path_plot <- ggdraw() +
  draw_plot(fun.venn, x = 0, y = 0.5, width = 1/3, height = 1/2) +
  draw_plot(pcoa_fun_p1, x = 0, y = 0, width = 1/3, height = 1/2) +
  draw_plot(pathway_plot, x = 1/3, y = 0, width = 2/3, height = 1) +
  draw_plot_label(label = c("a", "b", 'c'), size = 8,
                  x = c(0, 0, 1/3), y = c(1, 0.5, 1))
# ggsave(file.path(save.dir, "/figs/kegg/path_plot1.pdf"),
#        path_plot, width = 5.5, height = 4, units = "in")
path_plot
```


Effect size plot
```{r, fig.align='center', fig.width=3.5, fig.height=4.5}
# determine the effect size of the permafrost thawing on each phylum
pathway_scale <- KO_L2_df %>% filter(PathwayL2 != "") %>% 
  column_to_rownames("PathwayL2") %>% t() %>% data.frame(check.names = F) %>%
  rownames_to_column("Sample_id") %>% 
  mutate(Gully_id = sapply(stringr::str_split(Sample_id, "_",  n = 2), `[`, 1)) %>%
  mutate(Group = case_when(grepl("_C", Sample_id) ~ "Uncollapsed",
                           grepl("_T", Sample_id) ~ "Collapsed")) %>%
  mutate(across(where(is.numeric), scale)) %>%
  relocate(where(is.character)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  select(-Sample_id)

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
pathway_S1 <- sapply(3:ncol(pathway_scale), function(j) {
    if (length(unique(pathway_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(pathway_scale[, j] ~ Group + (1 | Gully_id), data = pathway_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(pathway_S1)<-colnames(pathway_scale)[-c(1:2)]
pathway_S1[1:6, 1:6]
```


```{r, fig.align='center', fig.width=3.5, fig.height=7}
ko_9_levels_file <- 'E:/thermokarst_gully/data/metagenome/ko_9_levels.txt'

ko_9_levels <- read.table(ko_9_levels_file, header = TRUE, sep = "\t", 
                  as.is = TRUE, stringsAsFactors = FALSE, comment.char = "",
                  check.names = FALSE)

pathway_index <- unique(ko_9_levels[, c("L1", "L2")]) %>% pull(L2)
pathway_index <- pathway_index[pathway_index != ""]

p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
diff_in_pathway_plot <- pathway_S1 %>%
  t() %>%
  data.frame(check.names = F) %>%
  tibble::rownames_to_column(., "variables") %>%
  # filter(variables %in% pathway_index) %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  # mutate(variables = factor(variables, levels = rev(pathway_index))) %>%
  mutate(colour = case_when(GroupCollapsed.mean <= 0 & Group.P <= 0.05 ~ "Negative",
                            GroupCollapsed.mean > 0 & Group.P <= 0.05 ~ "Positvie",
                            Group.P > 0.05 ~ "Neutral")) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, color = colour)) +
  geom_hline(aes(yintercept =0), size=0.7,  colour="gray2")+
  geom_point(size = 1.5) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se + 0.2)),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") + 
  scale_colour_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0, xmax = 7.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#ff6666") +
  annotate("rect", xmin = 7.5, xmax = 15.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#f19837") +
  annotate("rect", xmin = 15.5, xmax = 20.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#e56eee") +
  annotate("rect", xmin = 20.5, xmax = 23.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#5fb236") +
  annotate("rect", xmin = 23.5, xmax = 28.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#035096") +
  annotate("rect", xmin = 28.5, xmax = 38, ymin = -2, ymax = 2, alpha = 0.2, fill = "#8b008b") +
  main_theme +
  theme(strip.background = element_rect(fill = c("#FFF6E1")), legend.position = "none")

# ggsave(file.path(save.dir, "/figs/kegg/diff_in_pathway_plot.pdf"),
#        diff_in_pathway_plot, width = 3, height =6, units = "in")
diff_in_pathway_plot
```

## Select the genes involved in C, N, S, and Fe cyclings
#### Arrange the log fold change table
```{r}
# loading packages
pacman::p_load(phyloseq, picante, microbiome, readxl, tidyverse, ggpubr, ggplot2)

# Read data
sel_ko <- read_csv(file.path(wd_fun, 'CN_ko_input.csv'), col_names = T)

C_N_ko_count_tab <- ko_count_table[rownames(ko_count_table) %in% sel_ko$KO, ]
C_N_ko_count_tab <- cbind(KO = rownames(C_N_ko_count_tab), C_N_ko_count_tab)

C_N_ko_tpm_tab <- ko_tpm_table[rownames(ko_tpm_table) %in% sel_ko$KO, ]
C_N_ko_tpm_tab <- cbind(KO = rownames(C_N_ko_tpm_tab), C_N_ko_tpm_tab)
```

## heatmap
```{r, fig.align='center', fig.width=8, fig.height=12}
C_N_path1 <- C_N_ko_count_tab %>% 
  inner_join(sel_ko, c("KO" = "KO")) %>% 
  select(-c(1)) %>%
  group_by(Category, Subcategory, Enzyme_protein_encoded) %>%
  summarise(across(everything(), sum)) %>% data.frame() %>%
  mutate(Subcategory = factor(Subcategory, levels = unique(sel_ko$Subcategory), ordered = T))
C_N_path2 <- C_N_path1 %>% select(c(3:63)) %>% 
  tibble::column_to_rownames('Enzyme_protein_encoded')
```

Test the difference in functional genes involving in carbon, nitrogen, sulfur, and other elements using linear mixed models
```{r}
# determine the effect size of the permafrost thawing on each phylum
count_scale <- data.frame(Gully_id = metadata$Gully_id, Group = metadata$Group, t(C_N_path2), check.names=FALSE) %>% 
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
count_S1 <- sapply(3:ncol(count_scale), function(j) {
    if (length(unique(count_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(count_scale[, j] ~ Group + (1 | Gully_id), data = count_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(count_S1)<-colnames(count_scale)[-c(1:2)]
count_S1[1:6, 1:6]
```

```{r}
C_names <- c("Alpha-amylase", "Glucoamylase", "Pullulanase", "Isopullulanase",
             "Arabinofuranosidase", "Beta_mannanase", "Xylanase", "Xylose isomerase",
             "Beta-glucosidase",  "Endoglucanase", "Exoglucanase",
             "Acetylglucosaminidase", "Endochitinase", "Exochitinase", "Pectinase", 
             "Aryl-aldehyde oxidase", "Isocitrate lyase", "Limonene-1, 2-epoxide hydrolase", 
             "Malate synthase", "Vanillate demethylase", 
             "Glyoxal oxidase", "Phenol oxidase (tyrosinase)",
             "Methyl coenzyme M reductase", "Particulate methane monooxygenase", "Soluable methane monooxygenase",
             "Pyruvate oxidation", "Pyruvate <=> acetyl-CoA + formate", "Acetogenesis", 
             "Lactate utilization", "Acetate to acetyl-CoA", "RuBisCo")
N_names <- c("amoA",	"amoB",	"amoC",	"hao",	"nxrA",	"nxrB", "narG",	"narH",	"narI", "napA",	
             "napB", "nirK", "nirS",	"norB",	"norC",	"nosZ", "HzsC", "HzsB", "HzsA", "hdh",
             "nrfA",	"nrfH", "nirB",	"nirD",	
             "nasA",	"nasB", "narB",	"NR", "NIT-6", "nirA",	"nifD", "nifK", "nifH", "nrtA",	"nrtB", "nrtC",	
             "nrtD", "nmo", "gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC", "glnA")
S_names <- c("sat", "cysC", "cysD", "cysNC", "csyH", "cysJ", "cysN", "aprA", "aprB", "dsrA", "dsrB",
             "asrA", "asrB", "asrC", "sir", "sor", "sreA", "sreB", "sreC", "hydA", "hydD", "hydB",
             "hydG", "psrA", "psrB", "psrC", "sqr", "fccA", "fccB", "soxD", "soxX", "soxA", "soxB",
             "soxC", "soxY", "soxZ", "ttrA", "ttrB", "ttrC", "phsA", "phsB", "phsC")
Other_names <- c("Cyc1", "Cyc2", "MtrA", "MtrB", "MtrC", "arsC (grx)", "arsC (trx)", 
                 "aioA", "arsM", "ygfM", "xdhD", "YgfK")
```

Effect size plot for carbon metabolisms
```{r, fig.align='center', fig.width=5.5, fig.height=2.5}
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
eff.size_carbon_plot <- count_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  filter(variables %in% C_names) %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = C_names)) %>%
  mutate(colour = case_when(GroupCollapsed.mean <= 0 & Group.P <= 0.05 ~ "Negative",
                            GroupCollapsed.mean > 0 & Group.P <= 0.05 ~ "Positvie",
                            Group.P > 0.05 ~ "Neutral")) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, fill = colour)) +
  geom_hline(aes(yintercept = 0), size = 0.375,  colour = "gray2") +
  geom_bar(stat="identity", width = 0.6) +
  # geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), 
                cex = 0.5) +
  geom_text(aes(label = sig, x = variables, 
                y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_fill_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  scale_x_discrete(position = "bottom") +
  annotate("rect", xmin = 0.5, xmax = 4.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#E31A1C") +
  annotate("rect", xmin = 4.5, xmax = 8.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#893f45") +
  annotate("rect", xmin = 8.5, xmax = 11.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 11.5, xmax = 14.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#5fb236") +
  annotate("rect", xmin = 14.5, xmax = 15.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#1F78B4") +
  annotate("rect", xmin = 15.5, xmax = 20.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#6A3D9A") +
  annotate("rect", xmin = 20.5, xmax = 22.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 22.5, xmax = 25.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#00bfff") +
  annotate("rect", xmin = 25.5, xmax = 30.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 30.5, xmax = 31.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#8b008b") +
  main_theme +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        strip.background = element_rect(fill = c("#FFF6E1")),
        # axis.text.y = element_blank()
        )
my_col <- c("#E31A1C", "#893f45", "#1F78B4", "#6A3D9A", "#b31b1b", "#5d3954", "#008b8b", "#8b008b", "#e75480", "#872657", "#00bfff",  "#c08081", "#355e3b", "#29ab87", "#f56991", "#32cd32", "#035096", "#bb3385", "#009E73")

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir, "/figs/kegg/eff.size_carbon_plot.pdf"),
#        eff.size_carbon_plot, width = 5.5, height = 2.5, units = "in")
eff.size_carbon_plot
```
Effect size plot for nitrogen metabolisms
```{r, fig.align='center', fig.width=3.5, fig.height=4.5}
eff.size_nitrogen_plot <- count_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  filter(variables %in% N_names) %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = N_names)) %>%
  mutate(colour = case_when(GroupCollapsed.mean <= 0 & Group.P <= 0.05 ~ "Negative",
                            GroupCollapsed.mean > 0 & Group.P <= 0.05 ~ "Positvie",
                            Group.P > 0.05 ~ "Neutral")) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, colour = colour)) +
  geom_hline(aes(yintercept = 0), size = 0.375,  colour = "gray2")+
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, 
                y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_color_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0.5, xmax = 4.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 4.5, xmax = 8.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 8.5, xmax = 11.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 11.5, xmax = 14.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#5fb236") +
  annotate("rect", xmin = 14.5, xmax = 15.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#9457eb") +
  annotate("rect", xmin = 15.5, xmax = 20.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 20.5, xmax = 22.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 22.5, xmax = 25.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 25.5, xmax = 27.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#5fb236") +
  annotate("rect", xmin = 27.5, xmax = 28.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#9457eb") +
  annotate("rect", xmin = 28.5, xmax = 29.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 29.5, xmax = 30.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  main_theme +
  theme(legend.position = "none",
        strip.background = element_rect(fill = c("#FFF6E1")),
        # axis.text.y = element_blank()
        )
my_col <- c("#E31A1C", "#893f45", "#1F78B4", "#6A3D9A", "#b31b1b", "#5d3954", "#008b8b", "#8b008b", "#e75480", "#872657", "#00bfff",  "#c08081", "#355e3b", "#29ab87", "#f56991", "#32cd32", "#035096", "#bb3385", "#009E73")

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir, "/figs/env/diff_in_env_plot.pdf"),
#        diff_in_env_plot, width = 2.5, height = 3.5, units = "in")
eff.size_nitrogen_plot
```
Effect size plot for sulfur metabolisms
```{r, fig.align='center', fig.width=3.5, fig.height=4.5}
eff.size_sulfur_plot <- count_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  filter(variables %in% S_names) %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = S_names)) %>%
  mutate(colour = case_when(GroupCollapsed.mean <= 0 & Group.P <= 0.05 ~ "Negative",
                            GroupCollapsed.mean > 0 & Group.P <= 0.05 ~ "Positvie",
                            Group.P > 0.05 ~ "Neutral")) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, colour = colour)) +
  geom_hline(aes(yintercept = 0), size = 0.375,  colour = "gray2")+
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, 
                y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_color_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0.5, xmax = 4.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 4.5, xmax = 8.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 8.5, xmax = 11.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 11.5, xmax = 14.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#5fb236") +
  annotate("rect", xmin = 14.5, xmax = 15.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#9457eb") +
  annotate("rect", xmin = 15.5, xmax = 20.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 20.5, xmax = 22.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 22.5, xmax = 25.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 25.5, xmax = 27.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#5fb236") +
  annotate("rect", xmin = 27.5, xmax = 28.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#9457eb") +
  annotate("rect", xmin = 28.5, xmax = 29.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 29.5, xmax = 30.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  main_theme +
  theme(legend.position = "none",
        strip.background = element_rect(fill = c("#FFF6E1")),
        # axis.text.y = element_blank()
        )
my_col <- c("#E31A1C", "#893f45", "#1F78B4", "#6A3D9A", "#b31b1b", "#5d3954", "#008b8b", "#8b008b", "#e75480", "#872657", "#00bfff",  "#c08081", "#355e3b", "#29ab87", "#f56991", "#32cd32", "#035096", "#bb3385", "#009E73")

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir, "/figs/env/diff_in_env_plot.pdf"),
#        diff_in_env_plot, width = 2.5, height = 3.5, units = "in")
eff.size_sulfur_plot
```
Effect size plot for other elements metabolisms
```{r, fig.align='center', fig.width=3.5, fig.height=4.5}
eff.size_other_plot <- count_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  filter(variables %in% Other_names) %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = Other_names)) %>%
  mutate(colour = case_when(GroupCollapsed.mean <= 0 & Group.P <= 0.05 ~ "Negative",
                            GroupCollapsed.mean > 0 & Group.P <= 0.05 ~ "Positvie",
                            Group.P > 0.05 ~ "Neutral")) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, colour = colour)) +
  geom_hline(aes(yintercept = 0), size = 0.375,  colour = "gray2")+
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, 
                y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_color_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0.5, xmax = 4.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ffdb00") +
  annotate("rect", xmin = 4.5, xmax = 8.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 8.5, xmax = 11.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 11.5, xmax = 14.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#5fb236") +
  main_theme +
  theme(legend.position = "none",
        strip.background = element_rect(fill = c("#FFF6E1")),
        # axis.text.y = element_blank()
        )
my_col <- c("#E31A1C", "#893f45", "#1F78B4", "#6A3D9A", "#b31b1b", "#5d3954", "#008b8b", "#8b008b", "#e75480", "#872657", "#00bfff",  "#c08081", "#355e3b", "#29ab87", "#f56991", "#32cd32", "#035096", "#bb3385", "#009E73")

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir, "/figs/env/diff_in_env_plot.pdf"),
#        diff_in_env_plot, width = 2.5, height = 3.5, units = "in")
eff.size_other_plot
```










Explore the functional resistance using Orwin & Wardle (2004) indices. 
Firstly, explore the response of measured soil functions to permafrost collapse
```{r}
# determine the effect size of the permafrost thawing on each phylum
env_scale <- metadata %>% 
  select(-c("Sample_id", "Sample_name")) %>%
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
env_S1 <- sapply(3:ncol(env_scale), function(j) {
    message("Now j=", j, " in ", ncol(env_scale), ". ", date())
    if (length(unique(env_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(env_scale[, j] ~ Group + (1 | Gully_id), data = env_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(env_S1)<-colnames(env_scale)[-c(1:2)]
data.frame(env_S1)
```
Effect size plot
```{r, fig.align='center', fig.width=3.5, fig.height=4.5}
fun_index <- c("Carbohydrate metabolism", "Nitrogen metabolism",	"Phosphorus metabolism",	"Sulfur metabolism",
               "Plant_shannon",	"Plant_richness",	"AGB",	"BGB",	"BG",	"NAG",	"LAP",	"PPO",	"H2O2",	
               "WHC",	"Soil_respiration",	"NH4_N",	"NO3_N",	"AP",	"SOC",	"plant_pathogen_control")
# rep_str = c("Plant_richness" = "Plant Richness",
#             "Plant_shannon" = "Plant Shannon",
#             "Soil_moisture" = "Soil moisture",
#             "H2O2" = expression(paste("H"[2], "O"[2])))
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
diff_in_env_plot <- env_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  filter(variables %in% fun_index) %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = fun_index)) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, color = variables)) +
  geom_hline(aes(yintercept =0), size=0.7,  colour="gray2")+
  geom_point(size = 1.5) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size")+
  # scale_color_manual(values=c("#D55E00","#446DA9","#446DA9","#446DA9","#446DA9","#D55E00","#D55E00")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0, xmax = 4.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#ff6666") +
  annotate("rect", xmin = 4.5, xmax = 9.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#f19837") +
  annotate("rect", xmin = 9.5, xmax = 10.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#e56eee") +
  annotate("rect", xmin = 10.5, xmax = 11.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#5fb236") +
  annotate("rect", xmin = 11.5, xmax = 15.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#035096") +
  annotate("rect", xmin = 15.5, xmax = 16.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#00bfff") +
  main_theme +
  theme(strip.background = element_rect(fill = c("#FFF6E1")), legend.position = "none")

if (!dir.exists(file.path(save.dir, "figs/env/"))) {
  dir.create(file.path(save.dir, "figs/env/"))
}
ggsave(file.path(save.dir, "/figs/env/diff_in_env_plot.pdf"),
       diff_in_env_plot, width = 2.5, height = 3.5, units = "in")
diff_in_env_plot
```
Secondly, calculate the functional resistance for each measured soil function. Resistance (t0) = 1- (2x|D0|)/(C0 + |D0|)
```{r}
func_df <- metadata[, -c(1:2)] %>% group_by(Gully_id, Group) %>%
  select(where(is.numeric)) %>%
  summarise(across(everything(), mean))

library(dplyr)

calculate_resistance <- function(df) {
  df %>%
    group_by(Gully_id) %>%
    summarise(across(where(is.numeric), ~ {
      # Separate the values for uncollapsed and collapsed
      C0 <- .[Group == "Uncollapsed"]
      D0 <- .[Group == "Collapsed"] - C0
      
      # Calculate resistance
      1 - (2 * abs(D0)) / (C0 + abs(D0))
    }))
}

# Determine the resistance
resistance_results <- calculate_resistance(func_df)
resistance_results
```
```{r, fig.align='center', fig.width=3.5, fig.height=4.5}

resistance_fun_stats <- resistance_results %>% as.data.frame() %>% 
  pivot_longer(cols = -c("Gully_id"), names_to = "Function", values_to = "Resistance") %>%
  group_by(Function) %>%
  get_summary_stats(Resistance, type = "common") %>% #or using type = "mean_sd" 
  filter(Function %in% fun_index)

#Plot
resistance_func_comparison <- resistance_fun_stats %>%
  mutate(Function = factor(Function, levels = fun_index)) %>%
  ggplot(aes(x = Function, y = mean, colour = "#e95f5c")) +
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = mean - se, ymax = mean + se, colour = "#e95f5c"), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  labs(x = NULL, y = "Resistance") +
  # scale_color_manual(values=c("#79ceb8", "grey", "#e95f5c")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-1, 1.5)) +
  coord_flip() + scale_x_discrete(position = "bottom") +
  annotate("rect", xmin = 0, xmax = 4.5, ymin = -1, ymax = 1.5, alpha = 0.2, fill = "#ff6666") +
  annotate("rect", xmin = 4.5, xmax = 9.5, ymin = -1, ymax = 1.5, alpha = 0.2, fill = "#f19837") +
  annotate("rect", xmin = 9.5, xmax = 10.5, ymin = -1, ymax = 1.5, alpha = 0.2, fill = "#e56eee") +
  annotate("rect", xmin = 10.5, xmax = 11.5, ymin = -1, ymax = 1.5, alpha = 0.2, fill = "#5fb236") +
  annotate("rect", xmin = 11.5, xmax = 15.5, ymin = -1, ymax = 1.5, alpha = 0.2, fill = "#035096") +
  annotate("rect", xmin = 15.5, xmax = 16.5, ymin = -1, ymax = 1.5, alpha = 0.2, fill = "#00bfff") +
  main_theme + theme(legend.position = "none")
resistance_func_comparison
```

