---
title: "Functional gene analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 300)
```

## Data input
Set work directory
```{r}
setwd('e:/thermokarst_gully/')
wd_16s <- file.path(getwd(),"data/16S/rdp")
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```

Loading packages
```{r}
pacman::p_load(tidyverse, ggrepel, EnhancedVolcano, edgeR, DESeq2)
```

Data input
```{r}
source("script/read_data_all.R")
```

Permutational multivariate analysis of variance using distance matrices (adonis) to test the difference of functional composition
```{r}
library(vegan)
# determine the dissimilarity matrix based on the bray-curties distance
fun_dist <-vegdist(t(ko_tpm_table), "bray" )
# permanova, ANOSIM and MRPP analysis
adonis2(fun_dist ~ Group, data = metadata)
mrpp(fun_dist, metadata$Group, perm = 999)
anosim(fun_dist, metadata$Group, perm = 999)
```

PCoA plot
```{r}
# taxa
ord.fun <-  cmdscale(fun_dist,  k = 2, eig = T, add = T)
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 6),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 6),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.position = "none")
pcoa_fun_p1 <- data.frame(Group = metadata$Group, scores(ord.fun)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_color_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  # scale_x_continuous(expand = c(0.03, 0.03)) +
  # scale_y_continuous(expand = c(0.03, 0.03)) +
  labs(x = paste("PCoA1 (", format(100 * ord.fun$eig[1] / sum(ord.fun$eig), digits = 3), "%)", sep = ""),
       y = paste("PCoA2 (", format(100 * ord.fun$eig[2] / sum(ord.fun$eig), digits = 3), "%)", sep = "")) +
  main_theme +
  theme(legend.background = element_blank(),
        legend.title = element_text(size = 6),
        legend.text = element_text(size = 6),
        legend.key = element_blank(),
        legend.position = c(0.85, 0.85),
        legend.key.size = unit(0.4, 'cm'))
# creat a directory for kegg-gene results
if (!dir.exists(file.path(save.dir, "figs/kegg/"))) {
  dir.create(file.path(save.dir, "figs/kegg/"))
}
# ggsave(file.path(save.dir, "./figs/kegg/PCoA_fun_bray.pdf"),
#        pcoa_fun_plot, width = 3.5, height = 2.5, units = "in")
pcoa_fun_p1
```
## Determine the funtional biotic homogenization
```{r}
library(vegan)
#determine the dissimilarity matrix based on the bray-curties distance
fun_dist <-vegdist(t(ko_count_table), "bray")
#permanova test the difference in compositional variance
vars <- c('G1_C', 'G1_T', 'G2_C', 'G2_T', 'G3_C', 'G3_T', 'G4_C', 'G4_T', 'G5_C', 'G5_T', 'G6_C', 'G6_T')
adonis2(fun_dist ~ Group, data = metadata)
similar_fun_data <- lapply(vars, function(x) usedist::dist_subset(fun_dist, grep(x, metadata$Sample_name, value = TRUE))) %>%
  do.call(cbind, .) %>%
  data.frame() %>%
  gather("tem_group", "distance") %>%
  cbind(Gully_id = rep(c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH'), each = 20),
        Group = rep(c('Control', 'Collapsed'), each = 10, times = 6)) %>%
  select(-tem_group) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')),
         Group = factor(Group, levels = c('Control', 'Collapsed')))
```

```{r, fig.align='center', fig.height=3.5, fig.width=3.5}
library(lme4)
library(lmerTest)
# determine the effect size of the permafrost thawing for the diversity indexes
dist_scale <- similar_fun_data %>% 
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))


fm1 <- lmer(distance ~ Group + (1 | Gully_id), data = dist_scale)
presult <- car::Anova(fm1, type = 2)
coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
names(coefs) <- paste0(names(coefs), ".mean")
SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
names(SEvalues) <- paste0(names(SEvalues), ".se")
tvalues <- coef(summary(fm1))[, "t value"]  ##t values
names(tvalues) <- paste0(names(tvalues), ".t")
chisqP <- c(presult[, 1], presult[, 3])
names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))
result <- data.frame(variables = "Function", GroupCollapsed.mean = coefs[2], GroupCollapsed.se = SEvalues[2], sig = "***")


fun_dist_comparison <-  ggplot(data = result, aes(x = variables, y = GroupCollapsed.mean, colour = "#e95f5c")) +
  geom_hline(aes(yintercept = 0), size = 0.7,  colour = "gray2")+
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size") +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "bottom") +
  theme(legend.position = "none", 
        panel.grid=element_blank(), 
        axis.title = element_text(color = 'black', size = 6),
        axis.ticks.length = unit(0.1, "lines"), axis.ticks = element_line(color = 'black', size = 0.3),
        axis.line = element_line(colour = "black", size = 0.1), 
        axis.text.y = element_blank(),
        axis.text.x = element_text(colour = 'black', size = 6, hjust = 1),
        legend.title = element_text(size = 7),
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.2, 'cm'),
        legend.background = element_rect(colour = "white"))

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir.multifunc, "./single_div_comparison.pdf"),
#        single_div_comparison, width = 2.7, height = 5, units = "in")
fun_dist_comparison <- ggplotGrob(fun_dist_comparison)
pcoa_fun_plot <- pcoa_fun_p1 + 
  annotation_custom(fun_dist_comparison, 
                    xmin = 0.06, xmax = 0.14, 
                    ymin = -0.075, ymax = -0.03)
pcoa_fun_plot
```

## Differential analysis of functional gene profiles
### 1. edgeR analysis.
Reference:https://www.reneshbedre.com/blog/edger-tutorial.html
```{r}
library(edgeR)
#The field in the class definition file that defines the classes of the data.
data_classes <- "Group"

# Load Expression Data
RNASeq <- ko_count_table

# Load group information
RNASeq[1:5, 1:5]
classDefinitions_RNASeq <- metadata
classDefinitions_RNASeq[1:5, 1:4]

# Filter Data (RNA-seq read counts are converted to CPM values 
# and genes with CPM > 1 in at least 50 of the samples are 
# retained for further study, a gene mush have at least 50 measurements 
# with more than 1 CPM in one of the classes to be included in the analysis)
cpms <- cpm(RNASeq)
keep <- rowSums(cpms > 1) >= 1
counts <- RNASeq[keep,]

# Normalization and Dispersion
# create data structure to hold counts and group information for each sample.
d <- DGEList(counts = counts, group =  factor(classDefinitions_RNASeq$Group))
# Filter out the genes with low counts
# #filterByExpr function to remove the low count genes. 
# #This function keeps the genes with a worthwhile counts (at least 10 read counts) in a minimal number of samples.
# keep <- filterByExpr(y = d)
# d <- d[keep, , keep.lib.sizes=FALSE]
# Normalization and effective library size
d <- calcNormFactors(d)
# Model fitting and estimating dispersions
d <- estimateCommonDisp(d)
d <- estimateTagwiseDisp(d)

# Testing for differential gene expression
et <- exactTest(object = d, pair = c("Control","Collapsed"))
# topTags() function is useful to extract the table with adjusted p values (FDR).
top_degs <- topTags(object = et, n = "Inf")
top_degs
#Get a summary DGE table (returns significant genes with absolute log fold change at least 1 and adjusted p value < 0.05)
summary(decideTests(object = et, lfc = log(2)))
if (!dir.exists(file.path(save.dir, "tables/kegg/"))) {
  dir.create(file.path(save.dir, "tables/kegg/"))
}
write.csv(as.data.frame(top_degs), file = file.path(save.dir, "tables/kegg/kegg_Collapsed_vs_control_edge.csv"))
```

DGE Visualization
```{r, fig.align='center', fig.width=5.5, fig.height=4}
# Create a volcano plot
## set the plot theme
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 6),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 6),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6))

data_edger <- top_degs %>% as.data.frame() %>% 
  mutate(Expression = case_when(logFC >= log(2) & FDR <= 0.05 ~ "Up-regulated",
                           logFC <= -log(2) & FDR <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged"))
volca_plot_edger <- ggplot(data_edger, aes(logFC, -log(FDR,10))) +
  geom_point(aes(color = Expression), size = 1) +
  xlab(expression("log"[2]*"FC")) + 
  ylab(expression("-log"[10]*"FDR")) +
  scale_color_manual(values = c("#79ceb8", "gray70", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  guides(colour = guide_legend(override.aes = list(size=1.5))) +
  main_theme +
  theme(legend.background = element_blank(),
        legend.title = element_text(size = 6),
        legend.text = element_text(size = 6),
        legend.key = element_blank(),
        legend.position = c(0.80, 0.85),
        legend.key.size = unit(0.4, 'cm'))

# ggsave(file.path(save.dir, "./figs/kegg/volca_plot_edger.pdf"),
#        volca_plot_edger, width = 3.5, height = 2.5, units = "in")
volca_plot_edger
```

### 2. DESeq2 analysis
```{r, fig.align='center', fig.width=5.5, fig.height=6}
group_df <- metadata
rownames(group_df) <- metadata$Sample_name # make sure the colnames of count table match the rownames of group_df.
group_df$Group <- factor(group_df$Group, levels = c("Control", "Collapsed"))
KEGG_dds <- DESeqDataSetFromMatrix(countData = round(ko_count_table + 1), 
                                   colData = group_df, design = ~ Group)
KEGG_deseq <- DESeq(KEGG_dds)
KEGG_res <- results(KEGG_deseq, contrast = c("Group", 'Collapsed', 'Control'),
                    independentFiltering=TRUE, alpha=0.01, pAdjustMethod="BH", parallel=TRUE)
KEGG_res <- lfcShrink(KEGG_deseq, contrast = c("Group", 'Collapsed', 'Control'), res = KEGG_res, type = 'normal')
KEGG_res$padj[is.na(KEGG_res$padj)] = 1
KEGG_significant = rownames(KEGG_res)[(KEGG_res$padj < 0.05) & (KEGG_res$log2FoldChange > 1)]
write.csv(as.data.frame(KEGG_res), file = file.path(save.dir, "tables/kegg/kegg_Collapsed_vs_control_deseq.csv"))
# Volcano plot
volca_plot_deseq2 <- EnhancedVolcano(KEGG_res,
    lab = rownames(KEGG_res),
    title = NULL,
    x = 'log2FoldChange',
    y = 'pvalue',
    #pCutoff = 10e-32,
    FCcutoff = 0.585,
    pointSize = 3.0,
    labSize = 6.0)
volca_plot_deseq2
```

## Extract the Up-regulated and Down-regulated genes within data_edger data
```{r}
up_KOs <- data_edger %>% filter(Expression %in% "Up-regulated") %>%
  rownames(.)
down_KOs <- data_edger %>% filter(Expression %in% "Down-regulated") %>%
  rownames(.)

ko_9_levels <- read.delim(file.path(wd_fun, "/ko_9_levels.txt"), header = T, sep = "\t")

up_df <- data_edger %>% 
  mutate(KO = rownames(.)) %>%
  filter(Expression %in% "Up-regulated") %>%
  inner_join(ko_9_levels, by = "KO") %>%
  select("KO", "Description", "L2", "L3", "PathwayID") %>%
  left_join(cbind(KO = rownames(ko_tpm_table), ko_tpm_table), by = "KO")

down_df <- data_edger %>% 
  mutate(KO = rownames(.)) %>%
  filter(Expression %in% "Down-regulated") %>%
  inner_join(ko_9_levels, by = "KO") %>%
  select("KO", "Description", "L2", "L3", "PathwayID") %>%
  left_join(cbind(KO = rownames(ko_tpm_table), ko_tpm_table), by = "KO")

```
## 3. KEGG decoder
Determine the completion of pathways based only on the KOs enriched in the collapsed samples.
```{r}
kegg_decoder_fun <- function(group, kegg_sig) {
  samples <- grep(group, metadata$Sample_name, value = T)
  kegg_decoder_df = data.frame(Sample = c(), KEGG = c())
  for (kegg_id in kegg_sig) {
    for (sample in samples) {
      if (ko_count_table[rownames(ko_count_table) == kegg_id, sample] > 0) {
        kegg_decoder_df = rbind(kegg_decoder_df, data.frame(Sample = sample, KEGG = kegg_id))
      }
    }
  }
  return(kegg_decoder_df)
}
kegg_decoder_collapsed <- kegg_decoder_fun('_T', up_KOs)
kegg_decoder_control <- kegg_decoder_fun('_C', down_KOs)

kegg_decoder_collapsed$Sample <- sapply(kegg_decoder_collapsed$Sample, FUN = function(x){sub('_', '', x)})
kegg_decoder_control$Sample <- sapply(kegg_decoder_control$Sample, FUN = function(x){sub('_', '', x)})

# write.table(kegg_decoder_collapsed, file.path(save.dir, 'tables/kegg/kegg_decoder_collapsed_in.tsv'), sep='\t',  col.names=FALSE, row.names=FALSE , quote=FALSE)
# write.table(kegg_decoder_control, file.path(save.dir, 'tables/kegg/kegg_decoder_control_in.tsv'), sep='\t',  col.names=FALSE, row.names=FALSE , quote=FALSE)
# in the Data folder using the CryoBiome-kegg-decoder env in linux with following script: KEGG-decoder -i kegg_decoder_collapsed_in.tsv -o KEGG_decoder_collapsed_output.tsv
```

Heatmap showing the completion of pathways across collapsed samples using R
```{r, fig.align='center', fig.width=6.5, fig.height=4.5}
KEGG_decoder_collapsed_table <- read.table(file.path(save.dir, 'tables/kegg/KEGG_decoder_collapsed_output.tsv'), sep = "\t", header = TRUE)
KEGG_decoder_control_table <- read.table(file.path(save.dir, 'tables/kegg/KEGG_decoder_control_output.tsv'), sep = "\t", header = TRUE)

# renaming pathways for plotting
rep_str = c("Mixed.acid..Formate" = "Mixed acid: Formate",
            "Methanogenesis.via.dimethylamine" = "Methanogenesis via dimethylamine",
            "Methanogenesis.via.methylamine" = "Methanogenesis via methylamine",
            "NADP.reducing.hydrogenase" = "NADP-reducing hydrogenase",
            "hydrogen.quinone.oxidoreductase" = "Hydrogen:quinone oxidoreductase",
            "NiFe.hydrogenase" = "NiFe hydrogenase",
            "dissimilatory.sulfite.....sulfide" = "Dissimilatory sulfite <=> sulfide",
            "Methanogenesis.via.acetate" = "Methanogenesis via acetate",
            "nitrogen.fixation" = "Nitrogen fixation",
            "Coenzyme.B.Coenzyme.M.regeneration" = "Coenzyme B/Coenzyme M regeneration",
            "Methanogenesis.via.methanol" = "Methanogenesis via methanol",
            "Coenzyme.M.reduction.to.methane" = "Coenzyme M reduction to methane",
            "DMSO.reductase" = "DMSO reductase",
            "thiosulfate.polysulfide.reductase" = "Thiosulfate/polysulfide reductase",
            "Wood.Ljungdahl" = "Wood-Ljungdahl",
            "anoxygenic.type.I.reaction.center" = "Anoxygenic photosynthesis type-I reaction centers",
            "NAD.reducing.hydrogenase" = "NAD-reducing hydrogenase",
            "Methanogenesis.via.CO2" = "Methanogenesis via CO2",
            "X4.Hydroxybutyrate.3.hydroxypropionate" = "4-Hydroxybutyrate/3-hydroxypropionate",
            "thiamin.biosynthesis" = "Thiamin biosynthesis",
            "Competence.related.core.components" = "Competence-related core components",
            "Photosystem.I" = "Photosystem I",
            "sulfhydrogenase" = "Sulfhydrogenase",
            "Type.IV.Secretion" = "Type IV Secretion" ,
            "Type.III.Secretion" = "Type III Secretion", 
            "transporter..vitamin.B12" = "Transporter: vitamin B12",
            "CP.lyase.operon" = "CP-lyase operon", 
            "Methanogenesis.via.dimethylsulfide..methanethiol..methylpropanoate" = "Methanogenesis via DMS, MeSH, MMPA", # Methanethiol (MeSH) and dimethylsulfide (DMS) and (MMPA) 3-methylpropionate
            "CP.lyase.complex" = "CP-lyase complex",
            "C.P.lyase.cleavage.PhnJ" = "C-P lyase cleavage PhnJ",
            "starch.glycogen.degradation" = "Starch/glycogen degradation")
# select the columns with the colsums larger than zero and arrange the data
KEGG_decoder_control_table <- KEGG_decoder_control_table %>%
  select_if(~ !is.numeric(.) || sum(.) != 0) %>%
  dplyr::rename(Sample = Function) %>%
  mutate(Sample = str_replace(Sample, "C", "_C")) %>%
  gather(-Sample, key = "pathway", value = "Completion") %>% 
  as.data.frame() %>%
  mutate(pathway = stringr::str_replace_all(pathway, rep_str))

KEGG_decoder_collapsed_table <- KEGG_decoder_collapsed_table %>%
  select_if(~ !is.numeric(.) || sum(.) != 0) %>%
  dplyr::rename(Sample = Function) %>%
  mutate(Sample = str_replace(Sample, "T", "_T")) %>%
  gather(-Sample, key = "pathway", value = "Completion") %>% 
  as.data.frame() %>%
  mutate(pathway = stringr::str_replace_all(pathway, rep_str))

# row bind the control and collapsed kegg-decoder tables and reorder the levels of pathway.
KEGG_decoder_table <- rbind(KEGG_decoder_control_table, KEGG_decoder_collapsed_table) %>%
  mutate(Group = case_when(grepl("_C", Sample) ~ "Down-regulated", 
                           grepl("_T", Sample) ~ "Up-regulated")) %>% # creat a new variables named Group
  mutate(Group = factor(Group, levels = c("Down-regulated", "Up-regulated"))) %>%
  mutate(pathway = factor(pathway, levels = rev(c(rev(unique((KEGG_decoder_control_table$pathway)[order(KEGG_decoder_control_table$Completion)])), unique((KEGG_decoder_collapsed_table$pathway)[order(KEGG_decoder_collapsed_table$Completion)])[-c(3, 5)]))))

# heatmap
kegg_heatmap <- ggplot(KEGG_decoder_table) + 
  geom_tile(aes(x = as.factor(Sample), y = as.factor(pathway), fill = Completion)) +
  xlab("") + ylab("") + 
  guides(fill = guide_colourbar(barwidth = 0.5, barheight = 10)) +
  scale_fill_gradientn(colours = c("white", "#e95f5c"), values = c(0, 1)) + 
  facet_grid( ~ Group, scales = "free_x", space = "free_x") +
  main_theme +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 4),
        panel.spacing = unit(0, "lines"),
        legend.title = element_text(size = 6),
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.1, 'cm')
        )
ggsave(file.path(save.dir, "/figs/kegg/kegg_heatmap.pdf"),
       kegg_heatmap, width = 7.5, height = 5, units = "in")
kegg_heatmap
```
```{r, fig.align='center', fig.width=8.9, fig.height=6}
library(cowplot)
left_col <- plot_grid(pcoa_fun_plot, volca_plot_edger, labels = c('a', 'b'), label_size = 8, ncol = 1, align = "v")
func_plot <- plot_grid(left_col, kegg_heatmap, labels = c('', 'c'), label_size = 8,rel_widths = c(1, 2.2), ncol = 2)
ggsave(file.path(save.dir, "/figs/kegg/func_plot.pdf"),
       func_plot, width = 8, height = 5, units = "in")
func_plot
```

## Explore the losted genes after permafrost thawing
```{r}
# library
library(ggvenn)
x <- list(
  Control = ko_tpm_table %>% data.frame() %>%
    mutate(rowsum = rowSums(select(., grep('_C', metadata$Sample_name, value = T)))) %>%
    filter(rowsum > 0) %>%
    rownames(), 
  Collapsed = ko_tpm_table %>% data.frame() %>%
    mutate(rowsum = rowSums(select(., grep('_T', metadata$Sample_name, value = T)))) %>%
    filter(rowsum > 0) %>%
    rownames()
  )
# plot
ggvenn(
  x, 
  fill_color = c("#f8766d", "#a3a500", "#00b0f6"),
  stroke_color = NA,
  set_name_size = 4,
  text_size = 4,
  show_percentage = F
)
```

## Select the genes involved in C, N, S, and Fe cyclings
#### Arrange the log fold change table
```{r}
# loading packages
pacman::p_load(phyloseq, picante, microbiome, readxl, tidyverse, ggpubr, ggplot2)

# Read data
sel_ko <- read_csv(file.path(wd_fun, 'CN_ko_input.csv'), col_names = T)

C_N_ko_count_tab <- ko_count_table[rownames(ko_count_table) %in% sel_ko$KO, ]
C_N_ko_count_tab <- cbind(KO = rownames(C_N_ko_count_tab), C_N_ko_count_tab)

C_N_ko_tpm_tab <- ko_tpm_table[rownames(ko_tpm_table) %in% sel_ko$KO, ]
C_N_ko_tpm_tab <- cbind(KO = rownames(C_N_ko_tpm_tab), C_N_ko_tpm_tab)

logFC_table <- data_edger %>% 
  mutate(KO = rownames(.)) %>%
  inner_join(sel_ko, c("KO" = "KO")) %>% 
  mutate(sig = cut(FDR, breaks=c(-Inf, 0.001, 0.01, 0.05, Inf), label=c("***", "**", "*", "")))
```


```{r}
C_names <- c("Alpha-amylase", "Glucoamylase", "Pullulanase", "Isopullulanase",
             "Arabinofuranosidase", "Beta_mannanase", "Xylanase", "Xylose isomerase",
             "Beta-glucosidase",  "Endoglucanase", "Exoglucanase",
             "Acetylglucosaminidase", "Endochitinase", "Exochitinase",
             "Pectinase", 
             "Aryl-aldehyde oxidase", "Isocitrate lyase", "Limonene-1, 2-epoxide hydrolase", "Malate synthase", "Vanillate demethylase", 
             "Glyoxal oxidase", "Phenol oxidase (tyrosinase)",
             "Methyl coenzyme M reductase", "Particulate methane monooxygenase", "Soluable methane monooxygenase",
             "Carbon monoxide dehydrogenase", "Tetrahydrofolate formylase", "ATP citrate lyase", 
             "Propionyl-CoA carboxylase", "RuBisCo")
N_names <- c("amoA",	"amoB",	"amoC",	"hao",	"nxrA",	"nxrB", "narG",	"narH",	"narI", "napA",	
             "napB", "nirK", "nirS",	"norB",	"norC",	"nosZ", "nrfA",	"nrfH", "nirB",	"nirD",	
             "nasA",	"nasB", "narB",	"NR", "NIT-6", "nirA",	"nifD", "nifK", "nifH", "nrtA",	"nrtB", "nrtC",	
             "nrtD", "nmo", "gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC", "glnA")
S_names <- c("sat", "cysC", "cysD", "cysNC", "csyH", "cysJ", "cysN", "aprA", "aprB", "dsrA", "dsrB",
             "asrA", "asrB", "asrC", "sir", "sor", "sreA", "sreB", "sreC", "hydA", "hydD", "hydB",
             "hydG", "psrA", "psrB", "psrC", "sqr", "fccA", "fccB", "soxD", "soxX", "soxA", "soxB",
             "soxC", "soxY", "soxZ", "ttrA", "ttrB", "ttrC", "phsA", "phsB", "phsC")
Other_names <- c("Cyc1", "Cyc2", "MtrA", "MtrB", "MtrC", "arsC (grx)", "arsC (trx)", 
                 "aioA", "arsM", "ygfM", "xdhD", "YgfK")
```

## Plot for nitrogen
```{r, fig.align='center', fig.height=2.5, fig.width=8}
p_N_enrich <- logFC_table %>% filter(Enzyme_protein_encoded %in% N_names) %>%
  mutate(pathway = factor(Enzyme_protein_encoded, levels = rev(N_names), ordered = T)) %>%
  ggplot(aes(x = pathway, y = logFC, color = Expression))+
  geom_point(size = 2) + 
  scale_color_manual(values = c("#2C7BB6", "grey20", "#D7191C")) +
  labs(y = "logFC", x = 'Genes') +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 6, colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6))
ggsave(file.path(save.dir, "/figs/kegg/p_N_enrich.pdf"),
       p_N_enrich, width = 7, height = 3, units = "in")
p_N_enrich
```
```{r, fig.align='center', fig.width=8.9, fig.height=1.5}
N_enrich_heatmap <- logFC_table %>% filter(Enzyme_protein_encoded %in% N_names) %>%
  mutate(pathway = factor(Enzyme_protein_encoded, levels = rev(N_names), ordered = T)) %>% 
  ggplot(aes(x = pathway, y = Category, fill = logFC)) +
  geom_tile(height = 1, width = 1, size = 0.2, color = "black") + 
  scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C") +  # low="#2C7BB6", mid="white", high="#D7191C" or low = "#009E73", mid = "white", high = "#E69F00"
  geom_text(aes(label = sig), 
            color = "black", size = 1.5) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  labs(y = NULL, x = NULL, fill = expression(paste("log"[2], "FC"))) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 6, colour = 'black'),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        # legend.key.height= unit(0.35, 'cm'),
        # legend.key.width= unit(0.35, 'cm'),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6)) +
  coord_fixed(ratio=1)
N_enrich_heatmap
```

## Heatmap for sulfur
```{r, fig.align='center', fig.height=2.5, fig.width=8}
p_S_enrich <- logFC_table %>% filter(Enzyme_protein_encoded %in% S_names) %>%
  mutate(pathway = factor(Enzyme_protein_encoded, levels = rev(S_names), ordered = T)) %>%
  ggplot(aes(x = pathway, y = logFC, color = Expression))+
  geom_point(size = 2) + 
  scale_color_manual(values = c("#2C7BB6", "grey20", "#D7191C")) +
  labs(y = "logFC", x = 'Genes') +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 6, colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6))
p_S_enrich
```
```{r, fig.align='center', fig.width=8.9, fig.height=1.5}
S_enrich_heatmap <- logFC_table %>% filter(Enzyme_protein_encoded %in% S_names) %>%
  mutate(pathway = factor(Enzyme_protein_encoded, levels = rev(S_names), ordered = T)) %>% 
  ggplot(aes(x = pathway, y = Category, fill = logFC)) +
  geom_tile(height = 1, width = 1, size = 0.2, color = "black") + 
  scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C") +  # low="#2C7BB6", mid="white", high="#D7191C" or low = "#009E73", mid = "white", high = "#E69F00"
  geom_text(aes(label = sig), 
            color = "black", size = 1.5) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  labs(y = NULL, x = NULL, fill = expression(paste("log"[2], "FC"))) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 6, colour = 'black'),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        # legend.key.height= unit(0.35, 'cm'),
        # legend.key.width= unit(0.35, 'cm'),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6)) +
  coord_fixed(ratio=1)
S_enrich_heatmap
```
## Other biogeochemical processes
```{r, fig.align='center', fig.height=2.5, fig.width=4}
p_other_enrich <- logFC_table %>% filter(Enzyme_protein_encoded %in% Other_names) %>%
  mutate(pathway = factor(Enzyme_protein_encoded, levels = rev(Other_names), ordered = T)) %>%
  ggplot(aes(x = pathway, y = logFC, color = Expression))+
  geom_point(size = 2) + 
  scale_color_manual(values = c("#2C7BB6", "grey20", "#D7191C")) +
  labs(y = "logFC", x = 'Genes') +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 6, colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6))
p_other_enrich
```
```{r, fig.align='center', fig.width=8.9, fig.height=1.5}
Other_enrich_heatmap <- logFC_table %>% filter(Enzyme_protein_encoded %in% Other_names) %>%
  mutate(pathway = factor(Enzyme_protein_encoded, levels = rev(Other_names), ordered = T)) %>% 
  mutate(Category = rep("Others", nrow(.))) %>%
  ggplot(aes(x = pathway, y = Category, fill = logFC)) +
  geom_tile(height = 1, width = 1, size = 0.2, color = "black") + 
  scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C") +  # low="#2C7BB6", mid="white", high="#D7191C" or low = "#009E73", mid = "white", high = "#E69F00"
  geom_text(aes(label = sig), 
            color = "black", size = 1.5) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  labs(y = NULL, x = NULL, fill = expression(paste("log"[2], "FC"))) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 6, colour = 'black'),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        # legend.key.height= unit(0.35, 'cm'),
        # legend.key.width= unit(0.35, 'cm'),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6)) +
  coord_fixed(ratio=1)
Other_enrich_heatmap
```


```{r}
up_df %>% select(-c(1,2,3,5)) %>%
  group_by(L3) %>%
  summarise(across(everything(), ~ sum(.x, na.rm = TRUE)))
```


```{r}
C_N_ko_count_tab <- ko_count_table[rownames(ko_count_table) %in% sel_ko$KO, ]
C_N_ko_count_tab <- cbind(KO = rownames(C_N_ko_count_tab), C_N_ko_count_tab)

C_N_ko_tpm_tab <- ko_tpm_table[rownames(ko_tpm_table) %in% sel_ko$KO, ]
C_N_ko_tpm_tab <- cbind(KO = rownames(C_N_ko_tpm_tab), C_N_ko_tpm_tab)
```

## heatmap
```{r, fig.align='center', fig.width=8, fig.height=12}
C_N_path1 <- C_N_ko_count_tab %>% 
  inner_join(sel_ko, c("KO" = "KO")) %>% 
  select(-c(1)) %>%
  group_by(Category, Subcategory, Enzyme_protein_encoded) %>%
  summarise(across(everything(), sum)) %>% data.frame() %>%
  mutate(Subcategory = factor(Subcategory, levels = unique(sel_ko$Subcategory), ordered = T))
C_N_path2 <- C_N_path1 %>% select(c(3:63)) %>% 
  tibble::column_to_rownames('Enzyme_protein_encoded')
```
###################################################
## edgeR analysis
```{r}
# 1. alternately edgeR analysi.
library("edgeR")

#The field in the class definition file that defines the classes of the data.
data_classes <- "Group"

# Load Expression Data
RNASeq <- C_N_path2

# Load subtype information
RNASeq[1:5, 1:5]
classDefinitions_RNASeq <- meta_dat
classDefinitions_RNASeq[1:5, 1:3]

# Filter Data (RNA-seq read counts are converted to CPM values 
# and genes with CPM > 1 in at least 50 of the samples are 
# retained for further study, a gene mush have at least 50 measurements 
# with more than 1 CPM in one of the classes to be included in the analysis)
cpms <- cpm(RNASeq)
keep <- rowSums(cpms > 1) >= 1
counts <- RNASeq[keep,]

# Normalization and Dispersion
# create data structure to hold counts and subtype information for each sample.
d <- DGEList(counts = counts, group = classDefinitions_RNASeq$Group)
# Normalize the data
d <- calcNormFactors(d)

# create multidimensional scaling(MDS) plot. The command below will automatically
# generate the plot containing all samples where each subtype is a different color.
# Ideally there should be a good separation between the different classes.
# mds_filename <- file.path(save.dir, "figs/metagenome/mdsplot_allsamples_CN.png")
# png(filename = mds_filename)
# mds_output <- plotMDS(d, labels = NULL, pch = 1,
#                       col= c("darkgreen", "red", "orange")[factor(classDefinitions_RNASeq$layer)],
#                       xlim = c(-2.5,4), ylim = c(-2.5,4))
# legend("topright",
#        legend = levels(factor(classDefinitions_RNASeq$layer)),
#        pch = c(1), col = c("darkgreen","red", "orange"), title = "Class",
#        bty = 'n', cex = 0.75)
# dev.off()

#calculate dispersion
d <- estimateCommonDisp(d)
d <- estimateTagwiseDisp(d)
```


## Compare Collapsed to the Control group.
```{r}
classes <- factor(classDefinitions_RNASeq$Group)
modelDesign <- model.matrix(~ 0 + classes)
contrast_collap <- makeContrasts(
  collavsrest = "classesCollapsed-classesControl", levels = modelDesign)
fit_glm <- glmFit(d, modelDesign)
collavsrest <- glmLRT(fit_glm , contrast = contrast_collap)
tt_collavsrest <- topTags(collavsrest, n = nrow(d))
# Create g:Profiler input list
# tt <- tt_exact_test
#get the indices of scored dataset that have FDR < 0.05

select_collapsed_genes = tt_collavsrest$table %>%  filter(FDR < 0.05)
#output how many genes there are in the set that have FDR < 0.05
nrow(select_collapsed_genes)
```

## arrange the log fold change table
```{r}
C_names <- c("Alpha-amylase", "Glucoamylase", "Pullulanase", "Isopullulanase",
             "Arabinofuranosidase", "Beta_mannanase", "Xylanase", "Xylose isomerase",
             "Beta-glucosidase",  "Endoglucanase", "Exoglucanase",
             "Acetylglucosaminidase", "Endochitinase", "Exochitinase",
             "Pectinase", 
             "Aryl-aldehyde oxidase", "Isocitrate lyase", "Limonene-1, 2-epoxide hydrolase", "Malate synthase", "Vanillate demethylase", 
             "Glyoxal oxidase", "Phenol oxidase (tyrosinase)",
             "Methyl coenzyme M reductase", "Particulate methane monooxygenase", "Soluable methane monooxygenase",
             "Carbon monoxide dehydrogenase", "Tetrahydrofolate formylase", "ATP citrate lyase", 
             "Propionyl-CoA carboxylase", "RuBisCo")
carbon_logFC_table <- tt_collavsrest$table %>%
  rownames_to_column(var = "pathway") %>%
  mutate(Expression = case_when(logFC >= log(2) & FDR <= 0.05 ~ "Up-regulated",
                           logFC <= -log(2) & FDR <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged")) %>%
  mutate(sig = case_when(PValue <= 0.001 ~ "***",
                         PValue > 0.001 & PValue <= 0.01 ~ "**",
                         PValue > 0.01 & PValue <= 0.05 ~ "*"))
```

## Heatmap for carbon
```{r, fig.align='center', fig.width=8, fig.height=4.7}
p_C_enrich <- carbon_logFC_table %>% filter(pathway %in% C_names) %>%
  mutate(pathway = factor(pathway, levels = C_names, ordered = T)) %>%
  ggplot(aes(x = pathway, y = logFC, color = Expression))+
  geom_point(size = 2) + 
  scale_color_manual(values = c("#2C7BB6", "grey20", "#D7191C")) +
  labs(y = "logFC", x = 'Pathway') +
  theme(axis.title.x = element_blank(),
        axis.text = element_text(size = 6, colour = 'black'),
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6))
ggsave(file.path(save.dir, "/figs/kegg/p_C_enrich.pdf"),
       p_C_enrich, width = 6.5, height = 3, units = "in")
p_C_enrich
```
```{r, fig.align='center', fig.width=8.9, fig.height=2.5}
C_enrich_heatmap <- carbon_logFC_table %>% filter(pathway %in% C_names) %>%
  mutate(pathway = factor(pathway, levels = C_names, ordered = T)) %>% 
  mutate(Category = rep("Carbon", nrow(.))) %>%
  ggplot(aes(x = pathway, y = Category, fill = logFC)) +
  geom_tile(height = 1, width = 1, size = 0.2, color = "black") + 
  scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C") +  # low="#2C7BB6", mid="white", high="#D7191C" or low = "#009E73", mid = "white", high = "#E69F00"
  geom_text(aes(label = sig), 
            color = "black", size = 1.5) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  labs(y = NULL, x = NULL, fill = expression(paste("log"[2], "FC"))) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 6, colour = 'black'),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        # legend.key.height= unit(0.35, 'cm'),
        # legend.key.width= unit(0.35, 'cm'),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6)) +
  coord_fixed(ratio=1)

C_enrich_heatmap
```

Combine the plot
```{r, fig.align='center', fig.width=8.9, fig.height=10}
library(cowplot)
left_col <- plot_grid(pcoa_fun_plot, volca_plot_edger, labels = c('A', 'B'), label_size = 6, ncol = 1, align = "v")
top_plot <- plot_grid(left_col, kegg_heatmap, labels = c('', 'C'), label_size = 6,rel_widths = c(1, 2.2), ncol = 2)
funct_plot <- plot_grid(top_plot, C_enrich_heatmap, N_enrich_heatmap,
                        S_enrich_heatmap, Other_enrich_heatmap,
                        labels = c("", "D", "E", "F", "G"),
                        label_size = 6, rel_heights = c(4, 1.45, 1, 1, 0.65), ncol = 1)
ggsave(file.path(save.dir, "/figs/funct_plot.pdf"),
       funct_plot, width = 8.9, height = 10, units = "in")
funct_plot
```


## Difference analysis based on the CAZy results
```{r, fig.align='center', fig.width=6.5, fig.height=4}
# reading data
pacman::p_load(ggplot2, ggpubr, gghalves)
cazy_table <- read.delim(file.path(wd_fun, "./eggnog.CAZy.TPM.spf"), 
                           header = T, sep = "\t")
cazy_table <- cazy_table %>%
  select(-1) %>%
  separate(CAZy, 
           into = c("category", "num"), 
           sep = "(?<=[A-Za-z])(?=[0-9])"
           ) %>%
  select(-2) %>%
  na.omit() %>%
  group_by(category) %>%
  summarise(across(, mean, na.rm = TRUE)) %>%
  pivot_longer(cols = -category, names_to = "Sample", values_to = "Abundance") %>%
  mutate(Group = case_when(grepl("_C", Sample) ~ "Control", 
                           grepl("_T", Sample) ~ "Collapsed")) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed")))

ggplot(cazy_table, aes(Group, Abundance, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~category, scales = "free_y") +
  main_theme
```

