---
title: "Metagenome_analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 300)
```

## set work directory

```{r}
setwd('e:/thermokarst_gully/')
wd_16s <- file.path(getwd(),"data/16S/rdp")
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```

## loading packages

```{r}
pacman::p_load(phyloseq, picante, microbiome, readxl, tidyverse, ggpubr, ggplot2)
```

## reading input files
```{r}
source("script/read_data_all.R")
# tree.file <- file.path(wd_fun, 'binning_70/result/gtdb_tree/bacteria/tax.unrooted.tree')
abundance_tab.file <- file.path(wd_fun, "/MAGs/bin_abundance_coverm.txt")
tax.file <- file.path(wd_fun, "/MAGs/annotation.txt")
# bin genome information
bin_infor_file <- file.path(wd_fun, "MAGs/Widb.csv")
metabolic_output_file <- file.path(wd_fun, "MAGs/METABOLIC_result.csv")
# tree <- read.tree(tree.file)
abundance_tab <- read.table(abundance_tab.file, header = TRUE, sep = "\t", row.names = 1,
                            as.is = TRUE, stringsAsFactors = FALSE, comment.char = "",
                            check.names = FALSE)[-1, ]
tax_bin <- read.table(tax.file, header = TRUE, sep = "\t", as.is = TRUE, 
                      stringsAsFactors = FALSE, comment.char = "", check.names = FALSE)
metabolic_tab <- read.csv(metabolic_output_file) %>%
  select(c('Category', 'Function', grep('Hmm.presence', colnames(.))))
```

## Metabolic functional MAGs

```{r}
target_cat <- c("Thermophilic specific", "Ethanol fermentation", "Complex carbon degradation", "Fermentation", "C1 metabolism", "Carbon fixation", "Methane metabolism", "Amino acid utilization", "Fatty acid degradation", "Aromatics degradation",  "Nitrogen cycling", "Sulfur cycling", "Hydrogenases", "Iron cycling", "Manganese cycling")

C_N_metapathway <- c("Amino acid utilization", "Amylolytic enzymes", 
                     "Cellulose degrading", "Endohemicellulases",
                     "Hemicullulose debranching", "Chitin degrading", 
                     "Phenol => Benzoyl-CoA", "acyl-CoA dehydrogenase",
                     "Other oligosaccharide degrading",
                     "Acetogenesis", "Lactate utilization", "Pyruvate oxidation",  
                     "Pyruvate <=> acetyl-CoA + formate", "Alcohol utilization",
                     "Acetate to acetyl-CoA", "Methane production",
                     "Methane oxidation - Partculate methane monooxygenase",
                     "Methane oxidation - Soluble methane monoxygenase", 
                     "Ammonia oxidation", "Nitrate reduction", "Nitric oxide reduction",
                     "Nitrite oxidation", "Nitrite reduction to ammonia", "Nitrite reduction",
                     "Nitrous oxide reduction", "N2 fixation")
other_metapathway <- c("Sulfate reduction", "Sulfide oxidation", "Sulfite reduction",
                       "Sulfur oxidation", "Sulfur reduction", "Thiosulfate disproportionation", 
                       "Thiosulfate oxidation", "Iron oxidation", "Iron reduction",
                       "Arsenite oxidation", "Arsenate reduction", "Selenate reduction", 
                       "FeFe hydrogenase", "Ni-Fe Hydrogenase")
dat1 <- metabolic_tab %>% 
  filter(Category %in% target_cat) %>%
  mutate(across(starts_with("s_"), ~ifelse(. == "Present", 1, 0))) %>%
  group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  filter(rowSums(across(where(is.numeric)))!=0) %>% 
  mutate(across(starts_with("s_"), ~ifelse(. >= 1, 1, 0))) %>%
  pivot_longer(cols = -c(Category, Function), names_to = "ID", values_to = 'presence_or_absent') %>%
  mutate(ID = gsub(".Hmm.presence", "", ID))
# write the table for itol annotation
# write.csv(dat3, "E:/permafrost/data/metagenome/metabolic/METABOLIC_70_layer/METABOLIC_70_all/itol_DIY_manual/annotation.csv")
```

## Determine the total number of MAGs for each metabolic pathway

```{r, fig.width=10, fig.height=8}
num.mags.df <- dat1 %>% select(c(1, 2, 4)) %>%
  group_by(Category, Function) %>% 
  summarise(across(everything(), sum)) %>%
  mutate(Category = factor(Category, levels = target_cat))
  # filter(Function %in% c(C_N_metapathway, other_metapathway)) %>% 
  # mutate(Function = factor(Function, levels = c(C_N_metapathway, other_metapathway)))
knitr::kable(arrange(num.mags.df, desc(presence_or_absent)))
```

```{r, fig.width=10, fig.height=12}
my_colors <- c("#e4d00a", "#FB8072", "#FDB462", "#80B1D3", "#BEBADA", "#B3DE69", "#FCCDE5", "#A6CEE3",
               "#8DD3C7", "#B2DF8A", "#33A02C", "#FB9A99", "#E31A1C", "#893f45", "#FF7F00", "#1F78B4",
               "#6A3D9A", "#b31b1b", "#5d3954", "#008b8b", "#8b008b", "#e75480", "#872657", "#8fbc8f",
               "#00bfff", "#d71868", "#ffff00", "#c08081", "#cc6666", "#e49b0f", "#c90016", "#355e3b",
               "#29ab87", "#9457eb", "#f56991", "#32cd32", "#035096", "#bb3385", "#009E73", "#000000")
               
set.seed(1234)  # Set seed
my_col <- sample(my_colors) # Sample vector            
values = c("#6A3D9A", "#FCCDE5", "#1F78B4", "#A6CEE3", "#B3DE69", "#E69F00", "#56B4E9", "#009E73",
           "#F0E442", "#0072B2", "#D55E00", "#FB9A99", "#E31A1C", "#FDBF6F", "#8DD3C7", "#FFFFB3", 
           "#BEBADA", "#FB8072", "#B2DF8A", "#33A02C", "#FF7F00", "#000000")

values = c("#FFFFB3", "#FB8072", "#FDB462", "#80B1D3", "#BEBADA", "#B3DE69", "#FCCDE5", "#A6CEE3",
           "#8DD3C7", "#B2DF8A", "#33A02C", "#FB9A99", "#E31A1C", "#FDBF6F", "#FF7F00", "#1F78B4",
           "#6A3D9A", "#FCCDE5", "#1F78B4", "#A6CEE3", "#B3DE69", "#E69F00", "#56B4E9", "#009E73")

num.mags.plot <- num.mags.df %>% 
  mutate(Function = factor(Function, levels = num.mags.df$Function)) %>%
  ggplot(aes(x = Function, y = presence_or_absent, fill = Function)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rep(my_col, 10)) +
  scale_y_continuous(limits = c(0, 350), expand = c(0, 0)) +
  labs(x = NULL, y = 'Number of MAGs') +
  theme_classic() +
  theme(axis.text = element_text(colour = 'black'),
        legend.position = "none") +
  coord_flip()
num.mags.plot
# plot.name <- paste0(wd_fun, "/metabolic/results/number_of_MAGs_Plot.PDF")
# print(num.mags.plot)
# cairo_pdf(filename = plot.name, width = 10, height = 8, onefile = TRUE)
# num.mags.plot
# dev.off()
```

## Determine the metabolic potential of each taxa (phylum/class)
```{r, fig.align='center', fig.width=8.9, fig.height=10}
tax_bin1 <- rbind(tax_bin %>% 
  filter(Class %in% c("c__Alphaproteobacteria", "c__Betaproteobacteria",
                      "c__Gammaproteobacteria", "c__Deltaproteobacteria")) %>%
    select(ID, Phylum = Class), tax_bin %>% select(ID, Phylum) %>%
  filter(!Phylum %in% c("p__Proteobacteria"))) %>%
  mutate(Phylum = sapply(strsplit(Phylum, "__"), "[", 2))

# count the number of bins of each taxa
num.bins.df <- as.data.frame(table(tax_bin1$Phylum)) %>%
  select(Phylum = Var1, Freq)

# merge the tax table and the metabolic table
dat2 <- dat1 %>%
  mutate(ID = str_replace(ID, ".Hmm.presence", "")) %>%
  inner_join(tax_bin1, by = "ID") %>%
  as.data.frame() %>% # "Adding missing grouping variables" message 
  select(Category, Function, Phylum, presence_or_absent) %>%
  group_by(Category, Function, Phylum) %>%
  dplyr::summarize(presence = sum(presence_or_absent)) %>%
  left_join(num.bins.df, by = "Phylum") %>%
  mutate(pro = presence/Freq) %>%
  ungroup()

# plot heatmap with no clusters
pacman::p_load(pheatmap)
category <- unique(dat2$Category)

my_col <- c("#E31A1C", "#893f45", "#1F78B4", "#6A3D9A", "#b31b1b", "#5d3954", "#008b8b", "#8b008b", "#e75480", "#872657", "#00bfff",  "#c08081", "#355e3b", "#29ab87", "#f56991", "#32cd32", "#035096", "#bb3385", "#009E73")

plot_list = list()
for (i in 1:13) {
    p <- pheatmap(dat2 %>% filter(Category == category[i]) %>%
           dplyr::select(Function, Phylum, pro) %>% 
           pivot_wider(names_from = Phylum, values_from = pro) %>%
           column_to_rownames(var = "Function"),
         cluster_rows = F,
         cluster_cols = F,
         # show_rownames = FALSE,
         show_colnames = FALSE,
         border_color = 'Black',
         cellwidth = 10, cellheight = 10,
         fontsize_col = 8, fontsize_row = 8,
         color = c(colorRampPalette(c("white", my_col[i]))(3)),
         breaks = c(0, 0.2, 0.499, 1),
         legend = F, silent = T
         )
    plot_list[[i]] = p
}
p14 <- pheatmap(dat2 %>% filter(Category == category[14]) %>%
           dplyr::select(Function, Phylum, pro) %>% 
           pivot_wider(names_from = Phylum, values_from = pro) %>%
           column_to_rownames(var = "Function"),
         cluster_rows = F,
         cluster_cols = F,
         border_color = 'Black',
         cellwidth = 10, cellheight = 10,
         fontsize_col = 8, fontsize_row = 8,
         color = c(colorRampPalette(c("white", my_col[14]))(3)),
         breaks = c(0, 0.2, 0.499, 1),
         legend = F, silent = T
         )
heatmaps <- cowplot::plot_grid(plot_list[[1]][[4]], plot_list[[2]][[4]], plot_list[[3]][[4]], plot_list[[4]][[4]], 
                               plot_list[[5]][[4]], plot_list[[6]][[4]], plot_list[[7]][[4]], plot_list[[8]][[4]],
                               plot_list[[9]][[4]], plot_list[[10]][[4]], plot_list[[11]][[4]], plot_list[[12]][[4]],
                               plot_list[[13]][[4]], p14[[4]], ncol = 1, align = "v",
                               rel_heights = c(8, 3, 5, 3, 6, 2, 1, 5, 2, 3, 1, 3, 8, 13))

# ggsave(file.path(save.dir, "./figs/MAGs/heatmap_tax_fun_cor.pdf"),
#        heatmaps, width = 8.9, height = 10, units = "in")
heatmaps
```
Alternatively, use the gaps_row to split the panels
```{r, fig.align='center', fig.width=10, fig.height=10}
dat3 <- dat2 %>% dplyr::select(Function, Phylum, pro) %>% 
  pivot_wider(names_from = Phylum, values_from = pro) %>%
  column_to_rownames(var = "Function")

# 1) reorder the matrix based in the annotation
dat3_ordered <- dat3[num.mags.df$Function, ]

# 2) annotatio
sub_anno <- num.mags.df %>% select(Category, Function) %>% column_to_rownames(var = "Function") 

# 3) plot heatmap with no clusters
pheatmap(dat3_ordered, 
         cluster_rows = F,
         cluster_cols = F,
         gaps_row = c(8, 11, 16, 19, 25, 27, 28, 33, 35, 38, 39, 42, 50),
         # gaps_col = 6,
         annotation_row = sub_anno,
         border_color = 'Black',
         cellwidth = 10, cellheight = 10,
         fontsize_col = 8, fontsize_row = 8,
         color = c("white","#A3D1D1", "#3D7878"),
         breaks = c(0,0.2, 0.499, 1),
         legend = F
         # distances 0 to 0.2 are black, 0.2 to 0.5 are light blue, > 0.5 are black
         )
```


```{r, fig.align='center', fig.width=9.2, fig.height=8}
# input the data
MW_score_input_table <- file.path(wd_fun, "MAGs/MW-score_result.txt")
MW_score_table <- read.delim(MW_score_input_table, header = T, row.names = 1, sep = "\t")

# load libraries
pacman::p_load(ggplot2, ggplotify, pheatmap, patchwork, aplot)

# plot
mw_score_df <- data.frame(MW_score_table[, 1])
p1 <- as.ggplot(pheatmap(mw_score_df, 
         cluster_rows = F,
         cluster_cols = F,
         show_rownames = F,
         number_color = "black",
         gaps_row = c(19, 30, 35),
         border_color = 'Black', # scale = "column",
         cellwidth = 18, cellheight = 10,
         fontsize_col = 8, fontsize_row = 8,
               display_numbers =  round(mw_score_df, 1), 
               fontsize_number = 6,
               color = colorRampPalette(c("white", "#F8696B"), bias=0.5)(10),
         legend = F, silent = T
         ))


cper_df <- data.frame(MW_score_table[, -1])
p2 <- as.ggplot(pheatmap(cper_df, 
               cluster_rows = F,
               cluster_cols = F,
               number_color = "black",
               gaps_row = c(19, 30, 35),
               border_color = 'Black', scale = "row",
               cellwidth = 18, cellheight = 10,
               fontsize_col = 8, fontsize_row = 8,
         display_numbers = round(cper_df, 1), fontsize_number = 6,
               color = colorRampPalette(c("#5A8AC6", "white", "#F8696B"), bias=0.9)(100),
         legend = F, silent = T
         ))

# Decorate a 'ggplot' with Associated Information
# heatmap_MW_score <- p2 %>% insert_left(p1, width = 0.05)
heatmap_MW_score <- cowplot::plot_grid(p1, p2, ncol = 2, align = "h", rel_widths = c(0.05, 1))
# ggsave(file.path(save.dir, "./figs/MAGs/heatmap_MW_score.pdf"),
#        heatmap_MW_score, width = 9.2, height = 8, units = "in")
heatmap_MW_score
```
## Metabolic difference among different layers
```{r}
# Set work directory
sankey.plots.folder <- file.path(save.dir, "MAGs/metabolic/sankey") # Name of new output folder
if (!dir.exists(sankey.plots.folder)) {
  dir.create(sankey.plots.folder)
}
# Input files
ctrl_input_table <- file.path(wd_fun, "metabolic/control/METABOLIC_Figures_Input/Metabolic_Sankey_diagram_input1.txt")
collap_input_table <- file.path(wd_fun, "metabolic/collapsed/METABOLIC_Figures_Input/Metabolic_Sankey_diagram_input1.txt")
# Load the energy flow diagram input ---------------------------
table_ctrl <- read.table(ctrl_input_table, header = T, sep = "\t")
table_collap <- read.table(collap_input_table, header = T, sep = "\t")
```

Loading packages

```{r}
pacman::p_load(ggthemes, networkD3, data.table)
```

Generate a snakey plot with reactions as node, and edges are MAGs, and are colored by taxonomic group.
```{r}
# Rename the categories ---------------------------
table_ctrl$Category <- ifelse(grepl("C-S", table_ctrl$Reaction), "Carbon", 
                               ifelse(grepl("N-S", table_ctrl$Reaction), "Nitrogen",
                                      ifelse(grepl("S-S", table_ctrl$Reaction), "Sulfur",
                                             ifelse(grepl("O-S", table_ctrl$Reaction), "Others",""))))
table_ctrl$Category <- as.factor(table_ctrl$Category)

table_collap$Category <- ifelse(grepl("C-S", table_collap$Reaction), "Carbon", 
                               ifelse(grepl("N-S", table_collap$Reaction), "Nitrogen",
                                      ifelse(grepl("S-S", table_collap$Reaction), "Sulfur",
                                             ifelse(grepl("O-S", table_collap$Reaction), "Others",""))))
table_collap$Category <- as.factor(table_collap$Category)
```

```{r}
# reformate the data
table_all <- rbind(cbind(Group = rep("Uncollapsed", nrow(table_ctrl)), table_ctrl),
                   cbind(Group = rep("Collapsed", nrow(table_collap)), table_collap)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed")))

taxa_names <- table_all %>% select(Taxa, Freq) %>%
  group_by(Taxa) %>%
  summarise(sum = sum(Freq)) %>%
  arrange(desc(sum)) %>%
  pull(Taxa)
reaction_orders <- c(sort(grep('C-', unique(table_all$Reaction), value = T), decreasing = F),
                     sort(grep('N-', unique(table_all$Reaction), value = T), decreasing = F),
                     sort(grep('S-S', unique(table_all$Reaction), value = T), decreasing = F),
                     sort(grep('O-', unique(table_all$Reaction), value = T), decreasing = F))

table_all <- table_all %>%
  mutate(Taxa = factor(Taxa, levels = taxa_names)) %>%
  mutate(Category = factor(Category, levels = c('Carbon', 'Nitrogen', 
                                                'Sulfur', 'Others'))) %>%
  mutate(Reaction = factor(Reaction, levels = reaction_orders))

sankey <- data.table::rbindlist(list(table_all[c("Group", "Taxa", "Freq")],
                         table_all[c("Taxa", "Reaction", "Freq")]))
names(sankey) <- c('source', 'target', 'value')
knitr::kable(sankey[1:5, ])
```

```{r}
# Make a connection data frame
links <- sankey 
# From these flows we need to create a node data frame: it lists every entities involved in the flow
nodes <- data.frame(
  name = c(as.character(links$source), as.character(links$target)) %>% 
    unique()
)
nodes <- data.frame(
  name = c("Uncollapsed", "Collapsed", taxa_names,
           sort(grep('C-', unique(table_all$Reaction), value = T), decreasing = F),
           sort(grep('N-', unique(table_all$Reaction), value = T), decreasing = F),
           sort(grep('S-S', unique(table_all$Reaction), value = T), decreasing = F),
           sort(grep('O-', unique(table_all$Reaction), value = T), decreasing = F))
)
# With networkD3, connection must be provided using id, not using real name like in the links dataframe.. So we need to reformat it.
links$IDsource <- match(links$source, nodes$name)-1 
links$IDtarget <- match(links$target, nodes$name)-1
```

```{r}
# Set color of connections
# Add a 'group' column to each connection:
ass_link_col_fun <- function(df) {
  if (df["source"] == taxa_names[1] | df["target"] == taxa_names[1]) {
    group = 'a'
  } else if (df["source"] == taxa_names[2] | df["target"] == taxa_names[2]) {
    group = 'b'
  } else if (df["source"] == taxa_names[3] | df["target"] == taxa_names[3]) {
    group = 'c'
  } else if (df["source"] == taxa_names[4] | df["target"] == taxa_names[4]) {
    group = 'd'
  } else if (df["source"] == taxa_names[5] | df["target"] == taxa_names[5]) {
    group = 'e'
  } else if (df["source"] == taxa_names[6] | df["target"] == taxa_names[6]) {
    group = 'f'
  } else if (df["source"] == taxa_names[7] | df["target"] == taxa_names[7]) {
    group = 'g'
  } else if (df["source"] == taxa_names[8] | df["target"] == taxa_names[8]) {
    group = 'h'
  } else if (df["source"] == taxa_names[9] | df["target"] == taxa_names[9]) {
    group = 'i'
  } else if (df["source"] == taxa_names[10] | df["target"] == taxa_names[10]) {
    group = 'j'
  } else if (df["source"] == taxa_names[11] | df["target"] == taxa_names[11]) {
    group = 'k'
  } else if (df["source"] == taxa_names[12] | df["target"] == taxa_names[12]) {
    group = 'l'
  } else if (df["source"] == taxa_names[13] | df["target"] == taxa_names[13]) {
    group = 'm'
  } else if (df["source"] == taxa_names[14] | df["target"] == taxa_names[14]) {
    group = 'n'
  } else if (df["source"] == taxa_names[15] | df["target"] == taxa_names[15]) {
    group = 'o'
  } else if (df["source"] == taxa_names[16] | df["target"] == taxa_names[16]) {
    group = 'p'
  } else if (df["source"] == taxa_names[17] | df["target"] == taxa_names[17]) {
    group = 'q'
  } else if (df["source"] == taxa_names[18] | df["target"] == taxa_names[18]) {
    group = 'r'
  } else if (df["source"] == taxa_names[19] | df["target"] == taxa_names[19]) {
    group = 's'
  } else if (df["source"] == taxa_names[20] | df["target"] == taxa_names[20]) {
    group = 't'
  } else if (df["source"] == taxa_names[21] | df["target"] == taxa_names[21]) {
    group = 'u'
  } else if (df["source"] == taxa_names[21] | df["target"] == taxa_names[21]) {
    group = 'u'
  }
  return(group)
}
links$group <- apply(links, 1, ass_link_col_fun)

# Add a 'group' column to each node. Here I decide to put all of them in the same group to make them grey
nodes$group <- as.factor(c("my_unique_group"))
links[1:5, ]
nodes[1:5, ]
```

```{r}
# Give a color for each group
my_color <- 'd3.scaleOrdinal() .domain(["a", "b", "c", "d", "e", "f", 
"g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", 
"Carbon", "Nitrogen", "Sulfur", "Others", "my_unique_group"]) .range(["#4970a0", "#f8696b", "#FCCDE5",
"#6A3D9A", "#1F78B4", "#FDB462", "#80B1D3", "#BEBADA", "#c988d1", "#A6CEE3",
"#8DD3C7", "#B2DF8A", "#33A02C", "#FB9A99", "#FFFFB3", "#FDBF6F", "#E31A1C", "#d5416a", 
"#FB8072", "#d87a71","#6a75d5", "#76c2d7", "#d87a71", "#FF7F00", "#836834", 
"#c0c0c0"])'

# Make the Network. I call my colour scale with the colourScale argument
P <- sankeyNetwork(Links = links, Nodes = nodes, Source = "IDsource", Target = "IDtarget", 
                   Value = "value", NodeID = "name", colourScale = my_color, 
                   LinkGroup = "group", NodeGroup = "group", iterations = 0, 
                   fontFamily = 'Arial', fontSize = 15, 
                   nodeWidth = 20, nodePadding = 10, height = 700, width = 890,
                   sinksRight = F)
# save the widget
# Recommendation: htmlwidgets
library(htmlwidgets)
# saveWidget(P, file = paste0(sankey.plots.folder, "/sankey_plot.html"))
# library(webshot)
# install phantom:
# webshot::install_phantomjs()
# Make a webshot in pdf : high quality but can not choose printed zone
# webshot(paste0(sankey.plots.folder, "/sankey_plot.html"), delay = 0.2)
P
```

## Metabolic network plot with reactions
Generate a network plot with reactions as node, and edges are MAGs, and are colored by taxonomic group. Read input files
```{r}
ctrl_net_input_table <- file.path(wd_fun, "metabolic/control/METABOLIC_Figures_Input/functional_network_input.txt")
collap_net_input_table <- file.path(wd_fun, "metabolic/collapsed/METABOLIC_Figures_Input/functional_network_input.txt")
network.plots.folder <- file.path(save.dir, "MAGs/metabolic/network")
table_net_ctrl <- read.csv(ctrl_net_input_table, header = T, sep = "\t")
table_net_collap <- read.csv(collap_net_input_table, header = T, sep = "\t")
```

```{r, fig.width=15, fig.height=7}
#install.packages("ggraph")
pacman::p_load(ggraph, igraph, tidyverse, tidygraph)

my_graph_ctrl <- table_net_ctrl[, c(2,3,4,5)] %>% 
  graph_from_data_frame()
my_graph_collap <- table_net_collap[, c(2,3,4,5)] %>% 
  graph_from_data_frame()

edge_density(my_graph_ctrl, loops = TRUE)
edge_density(my_graph_collap, loops = TRUE)

deg_ctrl <- degree(my_graph_ctrl, mode = "all")
deg_collap <- degree(my_graph_collap, mode = "all")

# color
ctrl_color <- c("#d64e9e", "#6cd54c", "#dd49d1", "#432d7c", "#c6d179", "#8f379a", "#c988d1")
collap_color <- c("#d64e9e", "#6cd54c", "#dd49d1", "#a152dd", "#5139c2", "#ceaa3b", "#432d7c",
               "#c6d179", "#8f379a", "#70d68c", "#d9432f", "#6ad5be", "#d5416a",
               "#6a75d5", "#c988d1", "#598939")

# this generates the whole community plot
community.plot_ctrl <- table_net_ctrl[, c(2,3,4,5)] %>% 
  graph_from_data_frame() %>% 
  ggraph(layout = "linear", circular = TRUE) +
  geom_edge_arc(alpha = .25, 
                aes(width = Coverage.value.average., color=as.factor(Taxonomic.Group))) +
  # scale_edge_colour_manual(values = ctrl_color) +
  geom_node_point(aes(size = 0.02*deg_ctrl), color = "black", alpha = 0.75) +
  geom_node_text(aes(label = name),  color = "black", repel = TRUE)+
  #theme_graph()+
  labs(title = 'Metabolic connections within dataset', 
       subtitle = 'No scaling')
community.plot_collap <- table_net_collap[, c(2,3,4,5)] %>% 
  graph_from_data_frame() %>% 
  ggraph(layout = "linear", circular = TRUE) +
  geom_edge_arc(alpha = .25, 
                aes(width = Coverage.value.average., color = as.factor(Taxonomic.Group))) +
  # scale_edge_colour_manual(values = collap_color) +
  geom_node_point(aes(size = 0.02*deg_collap), color = "black", alpha = 0.75) +
  geom_node_text(aes(label = name),  color = "black", repel = TRUE)+
  #theme_graph()+
  labs(title = 'Metabolic connections within dataset', 
       subtitle = 'No scaling')

# save community.plot
# plot.name <- paste0(network.plots.folder, "/SurCommunityPlot.PDF")
# print(plot.name)
# cairo_pdf(filename = plot.name, width = 8, height = 5.5, onefile = TRUE)
# community.plot_sur
# dev.off()
# 
# plot.name <- paste0(network.plots.folder, "/SubCommunityPlot.PDF")
# print(plot.name)
# cairo_pdf(filename = plot.name, width = 8, height = 5.5, onefile = TRUE)
# community.plot_sub
# dev.off()
# 
cowplot::plot_grid(community.plot_ctrl, community.plot_collap)
```
















## PCoA ordination analysis using the relative abundance of MAGs and Bray-Curtis dissimilarity index

```{r}
MAG_abund_tab <- abundance_tab
MAG_abund_tab <- MAG_abund_tab[, metadata$Sample_name]
MAG_dist <- as.matrix(vegdist(t(MAG_abund_tab), "bray" ))
adonis2(MAG_dist ~ Group, data = metadata)
mrpp(MAG_dist, metadata$Group, perm = 999)
anosim(MAG_dist, metadata$Group, perm = 999)
```

```{r, fig.align='center', fig.width=4.5, fig.height=4}
ord.MAG <-  cmdscale(MAG_dist,  k = 2, eig = T, add = T)
pcoa_MAG_plot <- data.frame(Group = metadata$Group, scores(ord.MAG)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  ggplot(aes(x = Dim1, y = Dim2, shape = Group, color = Group)) + 
  geom_point(size = 1, alpha = 0.8) + 
  stat_ellipse(geom = "polygon", aes(fill = Group), alpha = 0.2, show.legend = FALSE, level = 0.95) +
  scale_colour_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  scale_fill_manual(values = c("#f8766d", "#a3a500", "#00b0f6")) +
  labs(x=paste("PCoA1 (", format(100 * ord.MAG$eig[1] / sum(ord.MAG$eig), digits = 3), "%)", sep = ""),
       y=paste("PCoA2 (", format(100 * ord.MAG$eig[2] / sum(ord.MAG$eig), digits = 3), "%)", sep = "")) +
  theme(axis.title = element_text(size = 8, colour = "black"),
        axis.text = element_text(size = 6, colour = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
# ggsave(file.path(save.dir, "./figs/beta/PCoA_MAG_bray.pdf"),
#        pcoa_MAG_plot, width = 89, height = 59, units = "mm")
pcoa_MAG_plot
```

## Explore the difference in genome size among layers
```{r, fig.align='center', fig.width=4.5, fig.height=4.5}
library(ggpubr)
library(ggplot2)
mags.att.file <- "data/metagenome/MAGs/MAGs_attributes.txt"
mags.att <- read.table(mags.att.file, header = TRUE, sep = "\t", row.names = 1,
                  as.is = TRUE, stringsAsFactors = FALSE, comment.char = "",
                  check.names = FALSE)
#Group the MAGs into uncollapsed and collapsed MAGs
abun_uncollapsed <- abundance_tab[, grep("_C", colnames(abundance_tab), value = T)]
abun_uncollapsed <- abun_uncollapsed[rowSums(abun_uncollapsed) > 0, , drop = FALSE]

abun_collapsed <- abundance_tab[, grep("_T", colnames(abundance_tab), value = T)]
abun_collapsed <- abun_collapsed[rowSums(abun_collapsed) > 0, , drop = FALSE]

#Compare the difference in genome size among uncollapsed and collapsed samples
library(rstatix)
library(gghalves)

genome_size_uncolla <- mags.att %>% 
  filter(Genome_id %in% rownames(abun_uncollapsed)) %>%
  select(Genome_size)

genome_size_colla <- mags.att %>% 
  filter(Genome_id %in% rownames(abun_collapsed)) %>%
  select(Genome_size)

genome_size_df <- rbind(data.frame(Group = rep("Uncollapsed", length(genome_size_uncolla)), 
                                   Genome_size = genome_size_uncolla),
                        data.frame(Group = rep("Collapsed", length(genome_size_colla)), 
                                   Genome_size = genome_size_colla))

my_comparisons_group <- list(c('Uncollapsed', 'Collapsed'))
genome_size_group <- genome_size_df %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  ggplot(aes(Group, Genome_size, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.2), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.3, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  ggpubr::stat_compare_means(comparisons = my_comparisons_group, 
                     method = "wilcox.test", p.adjust.method = "BH", 
                     bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00) +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00", "#1F77B4", "#AEC7E8", "#FF7F0E", "#FFBB78")) +
  theme_bw() +
  theme(axis.title = element_text(colour = "black", size = 14),
        axis.text.x = element_text(colour = "black", size = 12),
        axis.text.y = element_text(colour = "black", size = 12),
        strip.text = element_text(colour = "black", size = 12),
        legend.key.size = unit(1,"line"),
        panel.grid = element_blank(),
        panel.spacing = unit(0, "lines"),
        legend.position = "none")
genome_size_group
```


```{r}
# determine the effect size of the permafrost thawing on each MAG
mags_abun <- abundance_tab %>% t() %>% as.data.frame() %>% rownames_to_column(var = "Sample_id") %>%
  mutate(Gully_id = case_when(grepl("G1", Sample_id) ~ "EB",
                              grepl("G2", Sample_id) ~ "ML",
                              grepl("G3", Sample_id) ~ "RS",
                              grepl("G4", Sample_id) ~ "SLH",
                              grepl("G5", Sample_id) ~ "HSX",
                              grepl("G6", Sample_id) ~ "HH")) %>%
  mutate(Group = case_when(grepl("_C", Sample_id) ~ "Uncollapsed",
                           grepl("_T", Sample_id) ~ "Collapsed")) %>%
  relocate(c(Gully_id, Group), .before = 1) %>%
  select(-Sample_id) %>% data.frame()

mags_scale <- mags_abun %>% 
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size
library(lme4)
library(lmerTest)
#Function for add a significant symbols according to the P values
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0,0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
#Function for linear mixed model
summary.model <- function(model){
  presult <- car::Anova(model, type = 2)
  coefs <- coef(summary(model))[, "Estimate"]  ##four coefs
  names(coefs) <- paste0(names(coefs), ".mean")
  
  SEvalues <- coef(summary(model))[, "Std. Error"]  ##standard errors
  names(SEvalues) <- paste0(names(SEvalues), ".se")
  
  tvalues <- coef(summary(model))[, "t value"]  ##t values
  names(tvalues) <- paste0(names(tvalues), ".t")
  
  chisqP <- c(presult[, 1], presult[, 3])
  names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))
  
  results <- c(coefs, SEvalues, tvalues, chisqP)
  return(results)
}

lmm.fun <- function(data) {
  df <- NULL
  for(i in 3:ncol(data)) {
    fm <- lmer(data[, i] ~ Group + (1|Gully_id), data = data)
    tmp <- summary.model(fm)
    if (is.null(df)){
      df <- tmp
    } else {
      df <- rbind(df, tmp)
    }
  }
  rownames(df)<-colnames(data)[-c(1:2)]
  df <- df %>% data.frame(check.names = F) %>%
    mutate(sig = p.stars(Group.P), 
           eff.group = case_when(GroupCollapsed.mean > 0 & Group.P < 0.05 ~ "Positive",
                                GroupCollapsed.mean < 0 & Group.P < 0.05 ~ "Negative",
                                Group.P > 0.05 ~ "Neutral")) %>%
    mutate(itol_label = case_when(eff.group == "Negative" ~ "TL|0|10|#79ceb8|Negative",
                                  eff.group == "Positive" ~ "TR|0|10|#e95f5c|Positive",
                                  eff.group == "Neutral" ~ ""))
  return(df)
}

mags_result_lmm <- lmm.fun(mags_scale)
# write.table(mags_result_lmm, file = "E:/thermokarst_gully/result/tables/MAGs/diff_in_mags.csv", 
#             sep = ",", col.names = NA, qmethod = "double")
```
Explore the composition of the MAGs within significant positive and negative groups
```{r, fig.align='center', fig.width=4.5, fig.height=4.5}
mags_tax_group_df <- mags_result_lmm %>% rownames_to_column("ID") %>% 
  left_join(tax_bin, by = "ID") %>%
  select(Class, eff.group) %>% 
  filter(eff.group %in% c("Positive", "Negative")) %>%
  mutate(Number = rep(1, nrow(.))) %>%
  group_by(Class, eff.group) %>%
  summarize(across(everything(), sum)) %>%
  mutate(Class = sapply(Class, function(x) gsub("c__", '', x, perl = TRUE))) %>%
  mutate(Class = factor(Class),
         Class = fct_reorder(Class, Number, .desc = T))

# loading the appropriate libraries
library(ggplot2)
library(webr)
library(dplyr)
PieDonut(mags_tax_group_df, aes(eff.group, Class, count = Number), 
         r0 = 0.45, r1 = 0.9, explode = 1)

```

```{r, fig.align='center', fig.width=7, fig.height=4.5}
microtrait_results <- readRDS(file.path(wd_fun, "/MAGs/microtraits/thermokarst_gully.microtraitresults.rds"))

abun.tab <- abundance_tab %>% 
  mutate(Control = rowMeans(select(., grep('_C', colnames(abundance_tab), value = T)))) %>%
  mutate(Collapsed = rowMeans(select(., grep('_T', colnames(abundance_tab), value = T)))) %>%
  rownames_to_column("Genome_id") %>%
  select(c(Genome_id, Control, Collapsed)) %>%
  pivot_longer(cols = -c(Genome_id), names_to = "Group", values_to = 'rel_abun') %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  filter(Group == "Control") %>% #Only select the values for MAGs in uncollapsed group
  select(Genome_id, rel_abun)
  
mags_result_eff.size <- mags_result_lmm %>% rownames_to_column("Genome_id")
growth_rate_file <- "E:/thermokarst_gully/data/metagenome/MAGs/growth_rate_results.csv"
growth_rate <- read.table(growth_rate_file, header = TRUE, sep = ",", row.names = 1,
                  as.is = TRUE, stringsAsFactors = FALSE, comment.char = "",
                  check.names = FALSE)
MAGs_all_aatr <- abun.tab %>% left_join(mags.att, by = "Genome_id") %>% left_join(growth_rate, by = "Genome_id") %>% left_join(mags_result_eff.size, by = "Genome_id") %>%
  mutate(`Max growth rate` = 1/d,
         `Codon usage bias` = 1/ConsistencyHE) %>%
  select(Genome_id, rel_abun, Genome_size, GC, `Codon usage bias`, 
         `Max growth rate`, GroupCollapsed.mean, eff.group)
# Combine all MAGs attributes
microtrait_results_metadata = microtrait::add.metadata(microtrait_results, MAGs_all_aatr, genome_metadata_idcol = "Genome_id")
# saveRDS(genomeset_results_wmetadata, file.path(base_dir, paste0(dataset, ".microtraitresults.rds")))

library(microtrait)
microtrait_results_metadata_norm = microtrait_results_metadata %>% trait.normalize(normby = "genome_length")

strategy_vars <- c("rel_abun", "Genome_size", "GC", "Max growth rate",  
                   "Resource Acquisition:Substrate uptake:N compound transport", 
                   "Resource Acquisition:Substrate degradation:complex carbohydrate depolymerization",
                   "Resource Acquisition:Substrate degradation:simple compound degradation",
                   "Resource Acquisition:Substrate assimilation:C1 compounds",
                   "Resource Acquisition:Substrate assimilation:N compounds",
                   "Resource Use:Chemotrophy:chemoorganoheterotrophy:aerobic respiration",
                   "Resource Use:Chemotrophy:chemolithoheterotrophy:anaerobic respiration",
                   "Resource Use:Chemotrophy:chemoorganoheterotrophy:fermentation",
                   "Stress Tolerance:Specific:temperature:low",
                   "Stress Tolerance:Specific:desiccation/osmotic/salt stress",
                   "Stress Tolerance:Specific:oxidative stress")
                    
# write a function to change the facet labels
strategy_names <- list(
  "rel_abun" = "Mean relative abundance (%)",
  "Genome_size" = "Genome size (Mb)",
  "GC" = "GC content (%)",
  "Max growth rate" = expression(paste("Max growth rate", " (h"^-1, ")")),
  "Resource Acquisition:Substrate uptake:N compound transport" = "N compound transport",
  "Resource Acquisition:Substrate degradation:complex carbohydrate depolymerization" = "Complex carbohydrate depolymerization",
  "Resource Acquisition:Substrate degradation:simple compound degradation" = "Simple compound degradation",
  "Resource Acquisition:Substrate assimilation:C1 compounds" = "C1 compounds assimilation",
  "Resource Acquisition:Substrate assimilation:N compounds" = "N compounds assimilation",
  "Resource Use:Chemotrophy:chemoorganoheterotrophy:aerobic respiration" = "Aerobic respiration", 
  "Resource Use:Chemotrophy:chemolithoheterotrophy:anaerobic respiration" = "Anaerobic respiration", 
  "Resource Use:Chemotrophy:chemoorganoheterotrophy:fermentation" = "Fermentation",
  "Stress Tolerance:Specific:temperature:low" = "Low temperature stress",
  "Stress Tolerance:Specific:desiccation/osmotic/salt stress" = "Desiccation/osmotic/salt stress", 
  "Stress Tolerance:Specific:oxidative stress" = "Oxidative stress")

facet_labeller <- function(variable,value){
  return(strategy_names[value])
}

# ggplot2
my_comparisons <- list(c('Negative', 'Positive'))
microtraits_comparison <- microtrait_results_metadata_norm$trait_matrixatgranularity1 %>% 
  mutate(Genome_size = Genome_size/1000000)%>%
  select(c("eff.group", all_of(strategy_vars))) %>%
  filter(eff.group %in% c('Negative', 'Positive')) %>%
  gather(mag.attrs, value, -c("eff.group")) %>%
  mutate(eff.group = factor(eff.group, levels = c('Negative', 'Positive'))) %>%
  mutate(mag.attrs = factor(mag.attrs, levels = strategy_vars)) %>%
  ggplot(aes(eff.group, value, fill = eff.group)) +
  geom_half_violin(position = position_nudge(x = 0.25), 
                   side = "r", width = 0.8, color = NA) +
  # geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = eff.group), size = 2, 
              width = 0.2, stroke = 0, pch = 21) +
  stat_summary(fun = median, geom = "crossbar", width = 0.35) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", 
                     bracket.size = 0.5,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~mag.attrs, scales = "free_y", ncol = 5, labeller = facet_labeller) + #
  main_theme +
  theme(legend.position = "none",
        strip.text.x = element_text(margin = margin(0.05, 0, 0.05, 0, "cm")))
# ggsave(file.path(save.dir, "./figs/MAGs/microtraits_comparison.pdf"),
#        microtraits_comparison, width = 7.8, height = 5, units = "in")

microtraits_comparison
```

```{r}


```


```{r}
trait_data %>%
  pivot_longer(cols = everything(), names_to = "traits", values_to = "value") %>% 
  ggplot(aes(x = value)) +
  geom_histogram(binwidth = 0.5, fill = "steelblue") +
  labs(title = "Multiple Histograms", x = "Value", y = "Frequency") +
  facet_wrap(~ traits, nrow = 5, scales = 'free_x') +
  theme_minimal()
```





```{r}
#toxonomic analysis
scaled_trait_data <- log(trait_data + 1) # Standardize the data
dist_traits <- dist(scaled_trait_data)
ord.traits <- cmdscale(as.matrix(dist_traits))

round(ord.traits$eig*100/sum(ord.traits$eig),1)[c(1,2)]

fit.total <- envfit(ord.traits, cbind(env.table[,c('vegetable_type')], env_se_total), perm=999, na.rm = TRUE)
fit.total
data.scores.total = as.data.frame(scores(ord.traits))
data.scores.total <- cbind(vegetable_type = env.table[,c('vegetable_type')], data.scores.total)
fit_coord_cont.total = as.data.frame(scores(fit.total, "vectors")) 
fit_coord_cat.total = as.data.frame(scores(fit.total, "factors")) 

gg.total <- ggplot(data = data.scores.total, aes(x = Dim1, y = Dim2)) + 
  geom_point(data = data.scores.total, colour = '#E76BF3', size = 3.5, alpha = 0.8) + 
  geom_segment(aes(x = 0, y = 0, xend = Dim1, yend = Dim2), 
               data = fit_coord_cont.total, size =1, alpha = 0.5, colour = "grey30") +
  labs(x = 'Dim1 (11.9%)', y = 'Dim (9.1%)') + 
  geom_text(data = fit_coord_cont.total, aes(x = Dim1, y = Dim2), colour = "black", 
            size = 3.5, label = row.names(fit_coord_cont.total)) + 
  theme(axis.title = element_text(size = 14, colour = "black"),
        axis.text = element_text(size = 12, colour = "black"),
        panel.grid = element_blank(),
        panel.background = element_blank(), 
        panel.border = element_rect(fill = NA, colour = "black"))
```


```{r, fig.align='center', fig.width=5, fig.height=2.3}
abun.tab <- abundance_tab %>% 
  mutate(Control = rowMeans(select(., grep('_C', colnames(abundance_tab), value = T)))) %>%
  mutate(Collapsed = rowMeans(select(., grep('_T', colnames(abundance_tab), value = T)))) %>%
  rownames_to_column("Genome_id") %>%
  select(c(Genome_id, Control, Collapsed)) %>%
  pivot_longer(cols = -c(Genome_id), names_to = "Group", values_to = 'rel_abun') %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  filter(Group == "Control") %>% #Only select the values for MAGs in uncollapsed group
  select(Genome_id, rel_abun)
  
mags_result_eff.size <- mags_result_lmm %>% rownames_to_column("Genome_id")
growth_rate_file <- "E:/thermokarst_gully/data/metagenome/MAGs/growth_rate_results.csv"
growth_rate <- read.table(growth_rate_file, header = TRUE, sep = ",", row.names = 1,
                  as.is = TRUE, stringsAsFactors = FALSE, comment.char = "",
                  check.names = FALSE)
MAGs_all_aatr <- abun.tab %>% left_join(mags.att, by = "Genome_id") %>% left_join(growth_rate, by = "Genome_id") %>% left_join(mags_result_eff.size, by = "Genome_id") %>%
  mutate(`Max growth rate` = 1/d,
         `Codon usage bias` = 1/ConsistencyHE) %>%
  select(Genome_id, rel_abun, Genome_size, GC, `Codon usage bias`, 
         `Max growth rate`, GroupCollapsed.mean, eff.group)

attri_vars <- c("rel_abun", "Genome_size", "GC", "Codon usage bias", "Max growth rate")

my_comparisons <- list(c('Negative', 'Positive'))
mags_atrr_plot <- MAGs_all_aatr %>% select(c("eff.group", attri_vars)) %>%
  filter(eff.group %in% c('Negative', 'Positive')) %>%
  gather(mag.attrs, value, -c("eff.group")) %>%
  mutate(eff.group = factor(eff.group, levels = c('Negative', 'Neutral', 'Positive'))) %>%
  mutate(mag.attrs = factor(mag.attrs, levels = attri_vars)) %>%
  ggplot(aes(eff.group, value, fill = eff.group)) +
  geom_half_violin(position = position_nudge(x = 0.25), 
                   side = "r", width = 0.8, alpha = 0.6, color = NA) +
  # geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = eff.group), size = 2, 
              alpha = 0.6, width = 0.2, stroke = 0, pch = 21) +
  stat_summary(fun = median, geom = "crossbar", width = 0.35) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.5,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  # labs(x = NULL, y = "Relative abundance (%)") +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~mag.attrs, scales = "free_y", ncol = 5) +
  main_theme +
  theme(legend.position = "none")
# ggsave(file.path(save.dir, "./figs/MAGs/mags_atrr_plot.pdf"),
#        mags_atrr_plot, width = 7.8, height = 2.4, units = "in")
mags_atrr_plot
```





```{r}
dat2 <- abundance_tab %>% 
  mutate(Control = rowMeans(select(., grep('_C', colnames(abun), value = T)))) %>%
  mutate(Collapsed = rowMeans(select(., grep('_T', colnames(abun), value = T)))) %>%
  rownames_to_column("ID") %>%
  select(c(ID, Control, Collapsed)) %>%
  pivot_longer(cols = -c(ID), names_to = "Group", values_to = 'rel_abun') %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed')))
# write the table for itol annotation
# write.csv(dat2, "E:/thermokarst_gully/result/MAGs/itol/abundance_annotation_coverm.csv")

#ZP plot
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 7),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 7),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.title = element_text(colour = 'black', size = 7),
        legend.text = element_text(colour = 'black', size = 6),
        legend.key.size = unit(0.5, 'cm'))
ZP_plot <- mags.att %>%
  mutate(`Network roles` = factor(`Network roles`, levels = c("Network hubs", "Module hubs",
                                                              "Connector hubs", "Peripheral species"))) %>%
  ggplot(aes(x = Pi, y = Zi)) +
  geom_point(shape = 19, alpha = 0.75, aes(colour = `Network roles`)) +
  scale_color_manual(values = c("#e95f5c", "#79ceb8", "#5cc3e8", "#ffdb00")) +
  scale_x_continuous(limits = c(0, 1.0)) +
  scale_y_continuous(limits = c(-2, 7)) +
  geom_vline(xintercept = 0.62) +
  geom_hline(yintercept = 2.5) +
  labs(x = "Among−module connectivity (Pi)",
       y = "Within−module connectivity (Zi)") +
  main_theme +
  theme(legend.position = c(0.2, 0.8))
# save the plot
# ggsave(file.path(save.dir, "./figs/MAGs/ZP_plot.pdf"),
#        ZP_plot, width = 90, height = 90, units = "mm")
ZP_plot
```
```{r}
annoatation_net <- mags.att %>%  
  mutate(net_itol_label = case_when(`Network roles` == "Connector hubs" ~ "EL|0|10|#33a02c|Connector hubs",
                                `Network roles` == "Peripheral species" ~ "RE|0|10|#a6cee3|Peripheral species",
                                `Network roles` == "Network hubs" ~ "DI|0|10|#b2df8a|Network hubs",
                               `Network roles` == "Module hubs" ~ "HH|0|10|#1f78b4|Module hubs"))
write.csv(annoatation_net, "E:/thermokarst_gully/result/MAGs/itol/annoatation_net1214.csv")
```





```{r, fig.align='center', fig.width=7.5, fig.height=4}
#difference in the core genomes relative abundance between the uncollapsed and collapsed samples
net_hubs <- mags.att %>% as_tibble() %>% filter(`Network roles` == "Network hubs") %>% pull(`shared name`)
net_cor <- abundance_tab[net_hubs, ] %>% t() %>% as.data.frame() %>% 
  rownames_to_column(var = "Sample_id") %>%
  mutate(Gully_id = gsub("_.+$", "", Sample_id)) %>%
  mutate(Group = case_when(grepl("_C", Sample_id) ~ "Uncollapsed",
                           grepl("_T", Sample_id) ~ "Collapsed"))

library(lme4)
library(lmerTest)
lmm_mags_modes <- lapply(net_hubs, function(x) {
  lmer(substitute(i ~ Group + (1|Gully_id), list(i = as.name(x))), data = net_cor)})
summary.model <- function(model){
  F.value <- anova(model)$'F value'
  p.value <- anova(model)$'Pr(>F)'
  p.stars <- function(p.values) {
    unclass(symnum(p.values, corr = FALSE, 
                   na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 1),
                   symbols = c("***", "**", "*", "ns")))}
  sig <- p.stars(p.value)
  results<-data.frame(F.value, p.value, sig)
  return(results)
}
df <- NULL
for(i in 1:length(net_hubs)) {
  tmp <- summary.model(lmm_mags_modes[[i]])
  if (is.null(df)){
    df <- tmp
  } else {
    df <- rbind(df, tmp)
  }
}

mags_result_lmm <- data.frame(Variables = net_hubs, group1 = rep("Uncollapsed", length(net_hubs)),
                            group2 = rep("Collapsed", length(net_hubs)), df)

library(gghalves)
annot_text <- tibble(Genome = factor(mags_result_lmm$Variables, levels = net_hubs),
                x1 = rep(0.5, length(net_hubs)),
                y1 = rep(0.95, length(net_hubs)),
                sig.labels = mags_result_lmm$sig,
                x2 = rep(0.1, length(net_hubs)),
                y2 = rep(1, length(net_hubs)),
                panel.labels = letters[as.numeric(Genome)])


mags_abun_plot <- net_cor %>% select(c("Group", net_hubs)) %>%
  gather(Genome, value, -c("Group")) %>%
  mutate(Group = factor(Group, levels = c("Uncollapsed", "Collapsed"))) %>%
  mutate(Genome = factor(Genome, levels = net_hubs)) %>%
  ggplot(aes(Group, value, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  labs(x = NULL, y = "Relative abundance (%)") +
  # add a sigment to denote the comparison between groups
  ggpp::geom_text_npc(data = annot_text, aes(npcx = x1, npcy = y1, label = sig.labels), inherit.aes = F) +
  ggpp::geom_text_npc(data = annot_text, aes(npcx = x2, npcy = y2, label = panel.labels), inherit.aes = F) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~Genome, scales = "free_y", ncol = 5) +
  main_theme +
  theme(legend.position = "none")


# save the plot
# ggsave(file.path(save.dir, "./figs/MAGs/mags_abun_plot.pdf"),
#        mags_abun_plot, width = 7.59, height = 2.74, units = "in")
mags_abun_plot
```

```{r}

```














## count the number of MAGs of each Class
```{r, fig.width=6, fig.height=4.5}
tax_tab %>% select(c("Domain")) %>%
  mutate(number = c(rep(1, nrow(tax_tab)))) %>%
  group_by(Domain) %>%
  summarise(across(everything(), sum)) %>%
  arrange(desc(number)) %>%
  mutate(Domain = gsub("d__", "", Domain)) %>%
  mutate(Domain = factor(Domain, levels = Domain))
```

```{r, fig.width=6, fig.height=4.5}
tax_tab %>% select(c("Phylum")) %>%
  mutate(number = c(rep(1, nrow(tax_tab)))) %>%
  group_by(Phylum) %>%
  summarise(across(everything(), sum)) %>%
  arrange(desc(number)) %>%
  mutate(Phylum = gsub("p__", "", Phylum)) %>%
  mutate(Phylum = factor(Phylum, levels = Phylum))
```

```{r, fig.width=6, fig.height=4.5}
tax_tab %>% select(c("Class")) %>%
  mutate(number = c(rep(1, nrow(tax_tab)))) %>%
  group_by(Class) %>%
  summarise(across(everything(), sum)) %>%
  arrange(desc(number)) %>%
  mutate(Class = gsub("c__", "", Class)) %>%
  mutate(Class = factor(Class, levels = Class)) %>%
  ggplot(aes(x = Class, y = number)) +
  geom_bar(stat = "identity", fill = "#0073c2", width = 0.65) +
  scale_y_continuous(limits = c(0, 38), expand = c(0, 0)) +
  labs(x = "Class", y = "Number of MAGs") +
  theme_bw() +
  theme(axis.title = element_text(color = "black", size = 12),
        axis.text.x = element_text(colour = "black", angle = 90, vjust = 0.5, hjust=1),
        axis.text.y = element_text(colour = "black"))
```

```{r}
dat2 <- abundance_tab %>% 
  mutate(Control = rowMeans(select(., grep('_C', colnames(abundance_tab), value = T)))) %>%
  mutate(Collapsed = rowMeans(select(., grep('_T', colnames(abundance_tab), value = T)))) %>%
  select(c(ID, Control, Collapsed)) %>%
  pivot_longer(cols = -c(ID), names_to = "Group", values_to = 'rel_abun') %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed')))
# write the table for itol annotation
# write.csv(dat2, "E:/thermokarst_gully/result/MAGs/itol/abundance_annotation.csv")
```




## Determine the ratio of presence of each metabolic pathway in each layers

```{r}
dat_3layers <- metabolic_tab %>% 
  filter(Function %in% c(C_N_metapathway, other_metapathway)) %>% 
  mutate(across(starts_with("s_"), ~ifelse(. == "Present", 1, 0))) %>%
  group_by(Category, Function) %>%
  summarise(across(everything(), sum)) %>%
  mutate(across(starts_with("s_"), ~ifelse(. >= 1, 1, 0))) %>%
  mutate(ratio_sur = rowSums(across(starts_with("s_SUR")), na.rm = T)/34,
         ratio_sub = rowSums(across(starts_with("s_SUB")), na.rm = T)/94,
         ratio_pl = rowSums(across(starts_with("s_PL")), na.rm = T)/146) %>%
  select(c('Category', 'Function', 'ratio_sur', 'ratio_sub', 'ratio_pl'))
knitr::kable(dat_3layers[1:10, ], "simple")
```

## relationship of bin with carbon flux
```{r}
bin_flux_file = file.path(wd_fun, '/bin_flux.csv')
bin_flux <- read_delim(bin_flux_file, col_names = T)
p <- ggplot(bin_flux, aes(x = s_SUB_15_15, y = Cumulative_5)) + 
  geom_point(size = 3.5, alpha = 0.8, colour = "#0000ff") +
  geom_smooth(method = "lm", formula = y ~ poly(x, 1, raw = TRUE), size = 1, se = T, colour = 'black') +
  ggpubr::stat_cor(aes(s_SUB_15_15, Cumulative_5, label = paste(..r.label.., ..p.label.., sep = "~`,`~")),
                   cor.coef.name = "r", p.accuracy = 0.001, r.accuracy = 0.01,
                   method = "pearson", label.x.npc = 0.1, label.y.npc = 0.9, size = 4) + 
  xlab("Abundance of s_SUB_15_15 (TPM)") + ylab(expression(CO[2])) +
  theme_bw() + 
  theme(panel.grid = element_blank(), 
        axis.title = element_text(color = "black", size = 10),
        axis.ticks.length = unit(0.2, "lines"), 
        axis.ticks = element_line(color = "black"), 
        axis.line = element_blank(), 
        axis.text.y = element_text(colour = "black", size = 8), 
        axis.text.x = element_text(colour = "black", size = 8), 
        strip.text = element_text(size = 14), legend.position = "none")
# ggsave(file.path(save.dir, "./figs/bin/bin_flux.pdf"),
#        p, width = 110, height = 120, units = "mm")
p
```

