---
title: "Multifuctionality_analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 450)
```
## Data input 
Set working directory
```{r}
setwd('e:/thermokarst_gully/')
save.dir <- file.path(getwd(),"result")
```
Loading packages
```{r}
# Averaging Approach
library(multifunc)

#for plotting
library(ggplot2)
library(patchwork)

#for data
library(tidyr)
library(dplyr)
library(purrr)
library(forcats)

#for analysis
library(car)
library(ggpubr)
library(ggpmisc)
library(vegan)
library(lme4)
```

```{r}
# read metadata
source("script/read_data_rdp.R")
# taxonomic diversity
div_table_16s <- read.table(file = file.path(save.dir, 'tables/div_table_16s.csv'),
             sep = ',',  header = T, row.names = 1, stringsAsFactors = F)
div_table_its <- read.table(file = file.path(save.dir, 'tables/div_table_its.csv'),
             sep = ',',  header = T, row.names = 1, stringsAsFactors = F)
# gene diversity
N_genes <- read.table(file = "E:/thermokarst_gully/data/metagenome/NCyc_output/all.result.TPM.txt", sep = "\t", header = TRUE, row.names = 1)
colnames(N_genes) <- metadata$Sample_name
P_genes <- read.table(file = "E:/thermokarst_gully/data/metagenome/PCyc_output/all.result.TPM.txt", sep = "\t", header = TRUE, row.names = 1)
colnames(P_genes) <- metadata$Sample_name
S_genes <- read.table(file = "E:/thermokarst_gully/data/metagenome/SCyc_output/all.result.TPM.txt", sep = "\t", header = TRUE, row.names = 1)
colnames(S_genes) <- metadata$Sample_name


R.N <- specnumber(t(N_genes))
H.N <- diversity(t(N_genes), 'shannon')

R.P <- specnumber(t(P_genes))
H.P <- diversity(t(P_genes), 'shannon')

R.S <- specnumber(t(S_genes))
H.S <- diversity(t(S_genes), 'shannon')

R.kegg <- specnumber(t(ko_tpm_table))
H.kegg <- diversity(t(ko_tpm_table), 'shannon')

multi_func_df <- data.frame(Ric_16s = div_table_16s$Observed, Shan_16s = div_table_16s$Shannon,
                            Ric_its = div_table_its$Observed, Shan_its = div_table_its$Shannon,
                            Ric_N = R.N, Shan_N = H.N, Ric_P = R.P, Shan_P = H.P, 
                            Ric_S = R.S, Shan_S = H.S, Ric_kegg = R.kegg, Shan_kegg = H.kegg,
                            metadata)
# write.table(multi_func_df, file.path(save.dir, 'tables/multi_func_df.csv'),
#             sep=',',  col.names = T, row.names = F , quote=FALSE)
```


```{r}
#Read in data
allVars <- c("BG",  "NAG", "LAP", "PPO", "H2O2", "WHC", "SOC", "NO3_N", "NH4_N", "AGB", "plant_pathogen_control", "AP", "Soil_respiration")

multi_func_df <- multi_func_df[, c("Plant_shannon", "Plant_richness", "Ric_N", "Shan_N", "Ric_P", "Shan_P", "Ric_S", "Shan_S", "Ric_16s", "Shan_16s", "Ric_its", "Shan_its", "Ric_kegg", "Shan_kegg", "Sample_name", "Gully_id", "Group", allVars)]

# varIdx<-which(names(all_biodepth) %in% allVars)

#add on the new functions along with the averaged multifunctional index
multi_ave_func_df <- cbind(multi_func_df, getStdAndMeanFunctions(multi_func_df, allVars, standardizeZScore))
#germany<-cbind(germany, getStdAndMeanFunctions(germany, vars, standardizeZScore))
```


```{r, fig.align='center', fig.width=8.9, fig.height=6.2}
library(gghalves)
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 7),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 7),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.title = element_text(colour = 'black', size = 7),
        legend.text = element_text(colour = 'black', size = 6),
        legend.key.size = unit(0.5, 'cm'))
my_comparisons <- list(c('Control', 'Collapsed'))
p1 <- multi_ave_func_df %>%
  select(c("Plant_richness", "Ric_N", "Ric_P", "Ric_S", "Ric_16s", "Ric_its", "Ric_kegg", "Group", "meanFunction")) %>%
  pivot_longer(-c(Group), names_to = "diversity", values_to = "value") %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  mutate(diversity = factor(diversity, levels = c("meanFunction", "Plant_richness", "Ric_kegg", "Ric_16s", 
                                                  "Ric_its", "Ric_N", "Ric_P", "Ric_S"))) %>%
  ggplot(aes(Group, value, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~diversity, scales = "free_y", ncol = 4) +
  main_theme + theme(legend.position = "none")

save.dir.multifunc <- file.path(save.dir,"figs/multifunction")
if (!dir.exists(save.dir.multifunc)) {
  dir.create(save.dir.multifunc)
}
ggsave(file.path(save.dir.multifunc, "./diversity_comparison.pdf"),
       p1, width = 8.9, height = 6.2, units = "in")
p1
```

```{r}
# Threshold Approach
thermoThresh <- getFuncsMaxed(multi_func_df, allVars, threshmin = 0.05, threshmax = 0.99, 
                              threshstep = 0.01, prepend = c("plot","Diversity"), maxN = 13)

# mfuncthermoLinear75 <- glm(funcMaxed ~ Ric_kegg, data = subset(thermoThresh, thermoThresh$thresholds == "0.75"), family = quasipoisson(link = "identity"), start = c(0.5,0.5))
# 
# Anova(mfuncthermoLinear75, test.statistic = "F")
# 
# summary(mfuncthermoLinear75)

gcPlot <- subset(thermoThresh, thermoThresh$thresholds %in% qw(0.25, 0.5, 0.75, 0.90)) #note, using qw as %in% is a string comparison operator

gcPlot$percent<-paste(100*gcPlot$thresholds, "%", sep = "")
```

Test the difference of multifunctionality at multiple thresholds using wilcox test.
```{r, fig.align='center', fig.width=3.5, fig.height=3.2}
p2 <- thermoThresh %>%
  select(Sample_name, Gully_id, Group, funcMaxed) %>%
  group_by(Sample_name, Gully_id, Group) %>%
  dplyr::summarise(across(, mean, na.rm = TRUE)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  ggplot(aes(Group, funcMaxed, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = expression("Number of Functions" >= Threshold)) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  main_theme + theme(legend.position = "none")

ggsave(file.path(save.dir.multifunc, "./Multithrehold_MultiFun_comparison.pdf"),
       p2, width = 3.5, height = 3.2, units = "in")
p2
```

Test the difference of multifunctionality at 25%, 50%, 75%, and 90% thresholds using wilcox test.
```{r, fig.align='center', fig.width=8.9, fig.height=3.2}
p3 <- gcPlot %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  ggplot(aes(Group, funcMaxed, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2, height = 0) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)), oob = scales::squish) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = expression("Number of Functions" >= Threshold)) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~percent, scales = "free_y", ncol = 4) +
  main_theme + theme(legend.position = "none")

ggsave(file.path(save.dir.multifunc, "./Multithrehold_25_50_75_90_MultiFun_comparison.pdf"),
       p3, width = 8.9, height = 3.2, units = "in")
p3
```
```{r, fig.align='center', fig.width=4.5, fig.height=4}
ggplot(data = gcPlot, aes(x = Ric_kegg, y = funcMaxed)) +
  geom_point(shape = 19, size = 1, colour ='tomato3', alpha = 0.8) +
  ylab(expression("Number of Functions" >= Threshold)) +
  xlab("Gene numbers") +
  stat_smooth(method = "glm", 
              method.args = list(family = gaussian(link = "identity")), 
              lwd = 0.8, fill = NA, colour = 'black', na.rm = F) +
  facet_wrap(~percent, scales = "free_y", ncol = 2)+
  main_theme + theme(legend.position = "none")
```

```{r, fig.align='center', fig.width=4, fig.height=2.5}
thermoThresh$percent <- 100 * thermoThresh$thresholds
ggplot(data = thermoThresh, aes(x = Ric_kegg, y = funcMaxed, group = percent)) +
  ylab(expression("Number of Functions" >= Threshold)) +
  xlab("Gene numbers") +
  stat_smooth(method = "glm", 
              method.args = list(family = gaussian(link = "identity")), 
              lwd = 0.8, fill = NA, aes(color = percent), na.rm = T) +
  theme_bw(base_size = 14) +
  scale_color_gradient(name = "Percent of \nMaximum", low = "#79ceb8", high = "#e95f5c") +
  main_theme
```


```{r}
# Multiple Threshold Approach
thermoLinearSlopes<-getCoefTab(funcMaxed ~ Ric_kegg,
                                data = thermoThresh, 
                                coefVar = "Ric_kegg",
                                family = gaussian(link = "identity"))
# Plot the values of the diversity slope at
# different levels of the threshold
thermoSlopes <- ggplot(thermoLinearSlopes, aes(x=thresholds*100,
                                                y = estimate,
                                                ymax = estimate + 1.96 * std.error,
                                                ymin = estimate - 1.96 * std.error)) +
  geom_ribbon(fill = "grey50") +
  geom_point() +
  ylab("Change in Number of Functions \nper Addition of 1 Species\n") +
  xlab("\nThreshold (%)") +
  geom_abline(intercept = 0, slope = 0, lwd = 1, linetype = 2) +
  theme_bw(base_size = 14) +
  main_theme
thermoSlopes
```

### test the difference in abundance of main taxa using liner models
```{r}
# determine the effect size of the permafrost thawing on each phylum
env_scale <- metadata %>% 
  select(c("Group", "Gully_id", allVars)) %>%
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
env_S1 <- sapply(3:ncol(env_scale), function(j) {
    message("Now j=", j, " in ", ncol(env_scale), ". ", date())
    if (length(unique(env_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(env_scale[, j] ~ Group + (1 | Gully_id), data = env_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(env_S1)<-colnames(env_scale)[-c(1:2)]
data.frame(env_S1)
```

Effect size plot
```{r, fig.align='center', fig.width=3.5, fig.height=4.5}
# rep_str = c("Plant_richness" = "Plant Richness",
#             "Plant_shannon" = "Plant Shannon",
#             "Soil_moisture" = "Soil moisture",
#             "H2O2" = expression(paste("H"[2], "O"[2])))
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
diff_in_env_plot <- env_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = colnames(metadata)[-c(1:4)])) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, color = variables)) +
  geom_hline(aes(yintercept =0), size=0.7,  colour="gray2")+
  geom_point(size = 3) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size")+
  # scale_color_manual(values=c("#D55E00","#446DA9","#446DA9","#446DA9","#446DA9","#D55E00","#D55E00")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0, xmax = 5.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#ff6666") +
  annotate("rect", xmin = 5.5, xmax = 6.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#f19837") +
  annotate("rect", xmin = 6.5, xmax = 7.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#e56eee") +
  annotate("rect", xmin = 7.5, xmax = 8.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#5fb236") +
  annotate("rect", xmin = 8.5, xmax = 12.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#1ca9c9") +
  annotate("rect", xmin = 12.5, xmax = 13.7, ymin = -2, ymax = 2, alpha = 0.2, fill = "#0000ff") +
  main_theme +
  theme(legend.position = "none",
        strip.background = element_rect(fill = c("#FFF6E1")))

if (!dir.exists(file.path(save.dir, "figs/env/"))) {
  dir.create(file.path(save.dir, "figs/env/"))
}
ggsave(file.path(save.dir, "/figs/env/diff_in_env_plot.pdf"),
       diff_in_env_plot, width = 3, height = 3.5, units = "in")
diff_in_env_plot
```

```{r, fig.align='center', fig.width=10, fig.height=6}
#plot it
p4 <- multi_ave_func_df %>%
  select(c("Plant_richness", "Ric_N", "Ric_P", "Ric_S", "Ric_16s", "Ric_its", "Ric_kegg", "Group", "meanFunction")) %>%
  pivot_longer(-c(Group, meanFunction), names_to = "diversity", values_to = "value") %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  mutate(diversity = factor(diversity, levels = c("Plant_richness", "Ric_kegg", "Ric_16s", 
                                                  "Ric_its", "Ric_N", "Ric_P", "Ric_S"))) %>%
  ggplot(aes(x = value, y = meanFunction)) +
  geom_point(size=2, alpha=0.8, aes(colour = "#5cc3e8")) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  # scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  # scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), 
                                 after_stat(p.value.label), sep = "*\", \"*")),
               size = 2.4) +
  xlab("\nDiversity") + 
  ylab("Average Value of Standardized Functions\n") +
  facet_wrap(~diversity, scales = "free_x", ncol = 4) +
  main_theme + theme(legend.position = "none")
ggsave(file.path(save.dir.multifunc, "./cor_diversity_multifunc.pdf"),
       p4, width = 10, height = 6, units = "in")
p4
```


```{r, fig.align='center', fig.width=10, fig.height=6}
#plot it
p5 <- multi_ave_func_df %>%
  select(c("Plant_richness", "Ric_N", "Ric_P", "Ric_S", "Ric_16s", "Ric_its", "Ric_kegg", "Group", "meanFunction")) %>%
  pivot_longer(-c(Group, meanFunction), names_to = "diversity", values_to = "value") %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  mutate(diversity = factor(diversity, levels = c("Plant_richness", "Ric_kegg", "Ric_16s", 
                                                  "Ric_its", "Ric_N", "Ric_P", "Ric_S"))) %>%
  ggplot(aes(x = value, y = meanFunction, fill = Group)) +
  geom_point(size=2, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), 
                                 after_stat(p.value.label), 
                                 sep = "*\", \"*"), colour = Group),
               size = 2.4) +
  xlab("\nDiversity") + 
  ylab("Average Value of Standardized Functions\n") +
  facet_wrap(~diversity, scales = "free_x", ncol = 4) +
  main_theme
ggsave(file.path(save.dir.multifunc, "./cor_group_diversity_multifunc.pdf"),
       p5, width = 10, height = 6, units = "in")
p5
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
ggplot(aes(x = Shan_16s, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) +
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nBacterial Shannon") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
ggplot(aes(x = Ric_its, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nFungal Richness") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
ggplot(aes(x = Shan_its, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nFungal Shannon") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r}
kegg_richness_group_plot <- multi_func_df  %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  ggplot(aes(Group, Ric_kegg, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = "\nKEGG gene Richness") +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  main_theme
# ggsave(file.path(save.dir, "./figs/alpha/AVD_plot.pdf"), 
#        AVD_16s_plot, width = 89, height = 89, units = "mm")
kegg_richness_group_plot
```

```{r}
kegg_shannon_group_plot <- multi_func_df  %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  ggplot(aes(Group, Shannon_kegg, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = "\nKEGG gene Shannon") +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  main_theme
# ggsave(file.path(save.dir, "./figs/alpha/AVD_plot.pdf"), 
#        AVD_16s_plot, width = 89, height = 89, units = "mm")
kegg_shannon_group_plot
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
#plot it
ggplot(aes(x = Ric_kegg, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nKEGG gene Richness") + 
  ylab("Average Value of Standardized Functions\n")
```


```{r, fig.align='center', fig.width=5.5, fig.height=5}
#plot it
ggplot(aes(x = Shannon_kegg, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nKEGG gene Shannon") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
library(ggpmisc)
multi_func_df <- multi_func_df %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed")))

ggplot(aes(x = Ric_kegg, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nKEGG gene Richness") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
ggplot(aes(x = Shannon_kegg, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nKEGG gene Shannon") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
#plot it
ggplot(aes(x = Ric_N, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nNitrogen Richness") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
ggplot(aes(x = Ric_N, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nNitrogen Richness") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
#plot it
ggplot(aes(x = Shan_N, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nNitrogen Shannon") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
ggplot(aes(x = Shan_N, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nNitrogen Shannon") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
#plot it
ggplot(aes(x = Ric_P, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nP Richness") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
ggplot(aes(x = Ric_P, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nP Richness") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
#plot it
ggplot(aes(x = Shan_P, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nP Shannon") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
ggplot(aes(x = Shan_P, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nP Shannon") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
#plot it
ggplot(aes(x = Ric_S, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nS Richness") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
ggplot(aes(x = Ric_S, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nSurful Richness") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
```

```{r, fig.align='center', fig.width=5.5, fig.height=5}
#plot it
ggplot(aes(x = Shan_S, y = meanFunction), data = multi_func_df) + 
  geom_point(size = 3) + theme_bw(base_size = 15) +
  stat_smooth(method = "lm", colour = "black", size = 2) + 
  stat_cor(method = "pearson",label.x.npc = 0.65,label.y.npc = 0.95) +
  xlab("\nS Shannon") + 
  ylab("Average Value of Standardized Functions\n")
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
ggplot(aes(x = Shan_S, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nSurful Shannon") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
```


```{r, fig.align='center', fig.width=7.5, fig.height=5}
# devtools::install_github("aphalo/ggpmisc")
# devtools::install_github("aphalo/ggpp")
library(ggpmisc)
multi_func_df <- multi_func_df %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed")))

p1 <-ggplot(aes(x = Ric_16s, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nBacterial Richness") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
p1
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
library(ggpmisc)
p2 <- ggplot(aes(x = Shan_16s, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nBacterial Shannon") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
p2
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
library(ggpmisc)
p3 <- ggplot(aes(x = Ric_its, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nFungal Richness") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
p3
```

```{r, fig.align='center', fig.width=7.5, fig.height=5}
library(ggpmisc)
p4 <- ggplot(aes(x = Shan_its, y = meanFunction, fill = Group), data = multi_func_df) +
  geom_point(size=3.5, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), sep = "*\", \"*"), colour = Group)) +
  xlab("\nFungal Shannon") + 
  ylab("Average Value of Standardized Functions\n") +
  theme_bw() +
  theme(panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=10),
        axis.ticks.length = unit(0.2,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_blank(), 
        axis.text.y = element_text(colour='black',size=10),
        axis.text.x = element_text(colour='black', size = 10),
        strip.text = element_text(size = 10),
        legend.position = c(0.85, 0.85))
p4
```

```{r, fig.align='center', fig.width=8.9, fig.height=8.9}
cowplot::plot_grid(p1, p2, p3, p4, ncol = 2)
```

#reshape for plotting everything with ggplot2
germanyMeanForPlotting <- germany %>%
  select(Diversity, biomassY3.std:meanFunction) %>%
  pivot_longer(cols = c(biomassY3.std:meanFunction),
               names_to = "variable") %>%
  mutate(variable = fct_inorder(variable))
#nice names for plotting
levels(germanyMeanForPlotting$variable) <- c('Aboveground Biomass', 'Root Biomass', 'Cotton Decomposition', 'Soil Nitrogen', 'Plant Nitrogen', 'Mean Multifuncion Index')

#plot it
ggplot(aes(x=Diversity, y=value),data=germanyMeanForPlotting)+geom_point(size=3)+
  facet_grid(~variable) +
  theme_bw(base_size=15)+
  stat_smooth(method="lm", colour="black", size=2) + 
  xlab("\nSpecies Richness") +
  ylab("Standardized Value of Function\n")


#statistical fit
aveFit<-lm(meanFunction ~ Diversity, data=germany)
Anova(aveFit)

summary(aveFit)
