---
title: "Multifuctionality_analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 450)
```
## Data input 
Set working directory
```{r}
setwd('e:/thermokarst_gully/')
save.dir <- file.path(getwd(),"result")
```
Loading packages
```{r}
# Averaging Approach
library(multifunc)

#for plotting
library(ggplot2)
library(patchwork)

#for data
library(tidyr)
library(dplyr)
library(purrr)
library(forcats)

#for analysis
library(car)
library(ggpubr)
library(ggpmisc)
library(vegan)
library(lme4)
```

Read data
```{r}
# read metadata
# source("script/read_data_all.R")
# taxonomic diversity
div_table <- read.table(file = file.path(save.dir, 'tables/div_table_all.csv'),
             sep = ',',  header = T, row.names = 1, stringsAsFactors = F)
# gene diversity
N_genes <- read.table(file = "E:/thermokarst_gully/data/metagenome/NCyc_output/all.result.TPM.txt", sep = "\t", header = TRUE, row.names = 1)
colnames(N_genes) <- metadata$Sample_name
P_genes <- read.table(file = "E:/thermokarst_gully/data/metagenome/PCyc_output/all.result.TPM.txt", sep = "\t", header = TRUE, row.names = 1)
colnames(P_genes) <- metadata$Sample_name
S_genes <- read.table(file = "E:/thermokarst_gully/data/metagenome/SCyc_output/all.result.TPM.txt", sep = "\t", header = TRUE, row.names = 1)
colnames(S_genes) <- metadata$Sample_name


R.N <- specnumber(t(N_genes))
# H.N <- diversity(t(N_genes), 'shannon')

R.P <- specnumber(t(P_genes))
# H.P <- diversity(t(P_genes), 'shannon')

R.S <- specnumber(t(S_genes))
# H.S <- diversity(t(S_genes), 'shannon')

R.kegg <- specnumber(t(ko_tpm_table))
# H.kegg <- diversity(t(ko_tpm_table), 'shannon')

multi_func_df <- data.frame(div_table, Ric_N = R.N, Ric_P = R.P, Ric_S = R.S, Ric_kegg = R.kegg, metadata)
# write.table(multi_func_df, file.path(save.dir, 'tables/multi_func_df.csv'),
#             sep=',',  col.names = T, row.names = F , quote=FALSE)
```

## Determine the multifunctional index
### Averaged multifunctional index
```{r}
#Read in data
source("script/read_data_all.R")
div_table <- read.table(file = "E:/thermokarst_gully/data/multifun/multifun_richness_table_all.csv",
             sep = ',',  header = T, row.names = 1, stringsAsFactors = F)

allVars <- c("BG",  "NAG", "LAP", "PPO", "H2O2", "WHC", "SOC", "NO3_N", "NH4_N", "AGB", "plant_pathogen_control", "AP", "Soil_respiration")

# multi_func_df <- multi_func_df[, c("Plant_richness", "Soil_multidiv", "Ric_16s", "Ric_its", "Ric_pro", "Ric_anim", "Ric_N", "Ric_P", "Ric_S", "Ric_kegg", "Actinobacteria",	"Alphaproteobacteria",	"Acidobacteria",	"Firmicutes",	"Betaproteobacteria", "Planctomycetes",	"Gammaproteobacteria",	"Bacteroidetes",	"Deltaproteobacteria",	"Ascomycota",	"Basidiomycota",	"Mortierellomycota",	"Cercozoa",	"Conosa",	"Ciliophora",	"Chlorophyta",	"Apicomplexa",	"Ochrophyta",	"Pseudofungi",	"Lobosa",	"Sagenista",	"Opalozoa",	"Nematoda",	"Rotifera",	"Arthropoda",	"Annelida",	"Tardigrada", "Sample_name", "Gully_id", "Group", allVars)]

# varIdx<-which(names(all_biodepth) %in% allVars)

#add on the new functions along with the averaged multifunctional index
multi_ave_func_df <- cbind(div_table, getStdAndMeanFunctions(div_table, allVars, standardizeZScore))
```

### Multifunctionality using multithreshold Approach
```{r}
# Threshold Approach
thermoThresh <- getFuncsMaxed(div_table, allVars, threshmin = 0.05, threshmax = 0.99, 
                              threshstep = 0.01, prepend = c("plot", "Diversity"), maxN = 13)

# mfuncthermoLinear75 <- glm(funcMaxed ~ Ric_kegg, data = subset(thermoThresh, thermoThresh$thresholds == "0.75"), family = quasipoisson(link = "identity"), start = c(0.5,0.5))
# 
# Anova(mfuncthermoLinear75, test.statistic = "F")
# 
# summary(mfuncthermoLinear75)
```

Test the difference of multifunctionality between the contol and collapsed groups based on the average multifunctional index and multiple thresholds multifunctionality using wilcox test.
```{r, fig.align='center', fig.width=3.5, fig.height=2.5}
multifunctionality_tab <- thermoThresh %>%
  select(Sample_name, Gully_id, Group, funcMaxed) %>%
  group_by(Sample_name, Gully_id, Group) %>%
  dplyr::summarise(across(, mean, na.rm = TRUE)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  data.frame(., Meanfunction = multi_ave_func_df$meanFunction)
# plot
library(gghalves)
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 7),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 7),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6),
        legend.title = element_text(colour = 'black', size = 7),
        legend.text = element_text(colour = 'black', size = 6),
        legend.key.size = unit(0.5, 'cm'))
my_comparisons <- list(c('Control', 'Collapsed'))
p1 <- multifunctionality_tab %>%
  select(c("Group", "Meanfunction", "funcMaxed")) %>%
  pivot_longer(-c(Group), names_to = "Functionality", values_to = "value") %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  mutate(Functionality = factor(Functionality, levels = c("Meanfunction", "funcMaxed"))) %>%
  ggplot(aes(Group, value, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = NULL) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~Functionality, scales = "free_y", ncol = 4) +
  main_theme + theme(legend.position = "none")

save.dir.multifunc <- file.path(save.dir,"figs/multifunction")
if (!dir.exists(save.dir.multifunc)) {
  dir.create(save.dir.multifunc)
}
# ggsave(file.path(save.dir.multifunc, "./multifunctionality_comparison.pdf"),
#        p1, width = 3.5, height = 2.5, units = "in")
p1
```
Test the difference of multifunctionality at 25%, 50%, 75%, and 90% thresholds using wilcox test.
```{r, fig.align='center', fig.width=8.9, fig.height=3.2}
gcPlot <- subset(thermoThresh, thermoThresh$thresholds %in% qw(0.25, 0.5, 0.75, 0.90)) #note, using qw as %in% is a string comparison operator
gcPlot$percent<-paste(100*gcPlot$thresholds, "%", sep = "")

p2 <- gcPlot %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  ggplot(aes(Group, funcMaxed, fill = Group)) +
  geom_half_violin(position = position_nudge(x = 0.25), side = "r", width = 0.8, color = NA) +
  geom_boxplot(width = 0.4, size = 0.75, outlier.color = NA) +
  geom_jitter(aes(fill = Group), shape = 21, size = 1.5, width = 0.2, height = 0) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)), oob = scales::squish) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.3,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = NULL, y = expression("Number of Functions" >= Threshold)) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  facet_wrap(~percent, scales = "free_y", ncol = 4) +
  main_theme + theme(legend.position = "none")

# ggsave(file.path(save.dir.multifunc, "./Multithrehold_25_50_75_90_MultiFun_comparison.pdf"),
#        p2, width = 8.9, height = 3.2, units = "in")
p2
```

### Test the difference in single function between control and collapsed groups using liner models
```{r}
# determine the effect size of the permafrost thawing on each phylum
env_scale <- metadata %>% 
  select(c("Group", "Gully_id", allVars)) %>%
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
env_S1 <- sapply(3:ncol(env_scale), function(j) {
    message("Now j=", j, " in ", ncol(env_scale), ". ", date())
    if (length(unique(env_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(env_scale[, j] ~ Group + (1 | Gully_id), data = env_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(env_S1)<-colnames(env_scale)[-c(1:2)]
data.frame(env_S1)
```

Effect size plot
```{r, fig.align='center', fig.width=3.5, fig.height=4.5}
rep_str = c("Plant_richness" = "Plant Richness",
            "Plant_shannon" = "Plant Shannon",
            "Soil_moisture" = "Soil moisture",
            "Soil_respiration" = "Soil respiration",
            "H2O2" = expression(paste("H"[2], "O"[2])),
            "NH4_N" = expression(paste("NH"[4]^"+", "-N")),
            "NO3_N" = expression(paste("NO"[3]^"-", "-N")),
            "plant_pathogen_control" = "Plant pathogen control"
)
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
single_Fun_comparison <- env_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = colnames(metadata)[-c(1:4)])) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, color = variables)) +
  geom_hline(aes(yintercept = 0), size = 0.7,  colour = "gray2")+
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size")+
  # scale_color_manual(values=c("#D55E00","#446DA9","#446DA9","#446DA9","#446DA9","#D55E00","#D55E00")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "top", labels = rep_str) +
  annotate("rect", xmin = 0.4, xmax = 5.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#ff6666") +
  annotate("rect", xmin = 5.5, xmax = 6.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#f19837") +
  annotate("rect", xmin = 6.5, xmax = 7.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#e56eee") +
  annotate("rect", xmin = 7.5, xmax = 8.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#5fb236") +
  annotate("rect", xmin = 8.5, xmax = 12.5, ymin = -2, ymax = 2, alpha = 0.2, fill = "#1ca9c9") +
  annotate("rect", xmin = 12.5, xmax = 13.7, ymin = -2, ymax = 2, alpha = 0.2, fill = "#0000ff") +
  main_theme +
  theme(legend.position = "none",
        strip.background = element_rect(fill = c("#FFF6E1")))

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir.multifunc, "./single_Fun_comparison.pdf"),
#        single_Fun_comparison, width = 2.7, height = 5, units = "in")
single_Fun_comparison
```
## 
Ordinary least squares linear regression between multidiversity (standardized between 0 and 1) of soil organisms and multifunctionality,
```{r}

```




```{r, fig.align='center', fig.width=4.5, fig.height=4}
ggplot(data = gcPlot, aes(x = Ric_kegg, y = funcMaxed)) +
  geom_point(shape = 19, size = 1, colour ='tomato3', alpha = 0.8) +
  ylab(expression("Number of Functions" >= Threshold)) +
  xlab("Gene numbers") +
  stat_smooth(method = "glm", 
              method.args = list(family = gaussian(link = "identity")), 
              lwd = 0.8, fill = NA, colour = 'black', na.rm = F) +
  facet_wrap(~percent, scales = "free_y", ncol = 2)+
  main_theme + theme(legend.position = "none")
```

```{r, fig.align='center', fig.width=4, fig.height=2.5}
thermoThresh$percent <- 100 * thermoThresh$thresholds
ggplot(data = thermoThresh, aes(x = Ric_kegg, y = funcMaxed, group = percent)) +
  ylab(expression("Number of Functions" >= Threshold)) +
  xlab("Gene numbers") +
  stat_smooth(method = "glm", 
              method.args = list(family = gaussian(link = "identity")), 
              lwd = 0.8, fill = NA, aes(color = percent), na.rm = T) +
  theme_bw(base_size = 14) +
  scale_color_gradient(name = "Percent of \nMaximum", low = "#79ceb8", high = "#e95f5c") +
  main_theme
```


```{r}
# Multiple Threshold Approach
thermoLinearSlopes<-getCoefTab(funcMaxed ~ Ric_kegg,
                                data = thermoThresh, 
                                coefVar = "Ric_kegg",
                                family = gaussian(link = "identity"))
# Plot the values of the diversity slope at
# different levels of the threshold
thermoSlopes <- ggplot(thermoLinearSlopes, aes(x=thresholds*100,
                                                y = estimate,
                                                ymax = estimate + 1.96 * std.error,
                                                ymin = estimate - 1.96 * std.error)) +
  geom_ribbon(fill = "grey50") +
  geom_point() +
  ylab("Change in Number of Functions \nper Addition of 1 Species\n") +
  xlab("\nThreshold (%)") +
  geom_abline(intercept = 0, slope = 0, lwd = 1, linetype = 2) +
  theme_bw(base_size = 14) +
  main_theme
thermoSlopes
```

```{r, fig.align='center', fig.width=12, fig.height=4.5}
#plot it
p4 <- multi_ave_func_df %>%
  select(c("Plant_richness", "Soil_multidiv",  "Ric_16s", "Ric_its", "Ric_pro", "Ric_anim", 
           "Ric_N", "Ric_P", "Ric_S", "Ric_kegg", "Group", "meanFunction")) %>%
  pivot_longer(-c(Group, meanFunction), names_to = "diversity", values_to = "value") %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  mutate(diversity = factor(diversity, levels = c("Plant_richness", "Soil_multidiv", "Ric_16s", "Ric_its", "Ric_pro", "Ric_anim", "Ric_N", "Ric_P", "Ric_S", "Ric_kegg"))) %>%
  ggplot(aes(x = value, y = meanFunction)) +
  geom_point(size = 2, alpha = 0.8, aes(colour = "#5cc3e8")) +
  geom_smooth(method = "lm", size = 1, se = T, colour = 'black') +
  # scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  # scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour = 'black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), 
                                 after_stat(p.value.label), sep = "*\", \"*")),
               size = 2.4) +
  xlab("\nDiversity") + 
  ylab("Average Value of Standardized Functions\n") +
  facet_wrap(~diversity, scales = "free_x", ncol = 5) +
  main_theme + theme(legend.position = "none")
# ggsave(file.path(save.dir.multifunc, "./cor_diversity_multifunc.pdf"),
#        p4, width = 10, height = 6, units = "in")
p4
```



```{r, fig.align='center', fig.width=12, fig.height=5.5}
#plot it
p5 <- multi_ave_func_df %>%
  select(c("Plant_richness", "Soil_multidiv", "Ric_16s", "Ric_its", "Ric_pro", "Ric_anim", 
           "Ric_N", "Ric_P", "Ric_S", "Ric_kegg", "Group", "meanFunction")) %>%
  pivot_longer(-c(Group, meanFunction), names_to = "diversity", values_to = "value") %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  mutate(diversity = factor(diversity, levels = c("Plant_richness", "Soil_multidiv", "Ric_16s", "Ric_its", "Ric_pro", "Ric_anim", "Ric_N", "Ric_P", "Ric_S", "Ric_kegg"))) %>%
  ggplot(aes(x = value, y = meanFunction, fill = Group)) +
  geom_point(size=2, alpha=0.8, aes(colour = Group)) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), 
                                 after_stat(p.value.label), 
                                 sep = "*\", \"*"), colour = Group),
               size = 2.4) +
  xlab("\nDiversity") + 
  ylab("Average Value of Standardized Functions\n") +
  facet_wrap(~diversity, scales = "free_x", ncol = 5) +
  main_theme
# ggsave(file.path(save.dir.multifunc, "./cor_group_diversity_multifunc.pdf"),
#        p5, width = 10, height = 6, units = "in")
p5
```

```{r, fig.align='center', fig.width=12, fig.height=12.5}
#plot it
p6 <- multi_ave_func_df %>%
  select(c("Actinobacteria",	"Alphaproteobacteria",	"Acidobacteria",	"Firmicutes",	"Betaproteobacteria", "Planctomycetes",	"Gammaproteobacteria",	"Bacteroidetes",	"Deltaproteobacteria",	"Ascomycota",	"Basidiomycota",	"Mortierellomycota",	"Cercozoa",	"Conosa",	"Ciliophora",	"Chlorophyta",	"Apicomplexa",	"Ochrophyta",	"Pseudofungi",	"Lobosa",	"Sagenista",	"Opalozoa",	"Nematoda",	"Rotifera",	"Arthropoda",	"Annelida",	"Tardigrada", "Group", "meanFunction")) %>%
  pivot_longer(-c(Group, meanFunction), names_to = "diversity", values_to = "value") %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  mutate(diversity = factor(diversity, levels = c("Actinobacteria",	"Alphaproteobacteria",	"Acidobacteria",	"Firmicutes",	"Betaproteobacteria", "Planctomycetes",	"Gammaproteobacteria",	"Bacteroidetes",	"Deltaproteobacteria",	"Ascomycota",	"Basidiomycota",	"Mortierellomycota",	"Cercozoa",	"Conosa",	"Ciliophora",	"Chlorophyta",	"Apicomplexa",	"Ochrophyta",	"Pseudofungi",	"Lobosa",	"Sagenista",	"Opalozoa",	"Nematoda",	"Rotifera",	"Arthropoda",	"Annelida",	"Tardigrada"))) %>%
  ggplot(aes(x = value, y = meanFunction)) +
  geom_point(size=2, alpha=0.8, aes(colour = "#5cc3e8")) +
  geom_smooth(method="lm", size=1, se=T, colour='black') +
  # scale_colour_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  # scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  stat_poly_line(colour='black') +
  stat_poly_eq(aes(label = paste(after_stat(eq.label),
                                 after_stat(rr.label), 
                                 after_stat(p.value.label), sep = "*\", \"*")),
               size = 2.4) +
  xlab("\nRichness") + 
  ylab("Average Value of Standardized Functions\n") +
  facet_wrap(~diversity, scales = "free_x", ncol = 5) +
  main_theme + theme(legend.position = "none")
# ggsave(file.path(save.dir.multifunc, "./cor_diversity_multifunc.pdf"),
#        p4, width = 10, height = 6, units = "in")
p6
```

```{r}
#linner mixed model for matrix to matrix
lmm.mat.cal <- function(y, x, method){
  y <- as.matrix(y)
  x <- as.matrix(x)
  df<-NULL
  for(i in colnames(y)){
    for(j in colnames(x)){
      a <- y[, i, drop = F]
      b <- x[, j, drop = F]
      mode <- lmer(a ~ b + (1 | Gully_id), data = multi_ave_func_df, na.action=na.omit)
      coeff <- summary(mode)$coefficients[2, 1]
      r.square <- MuMIn::r.squaredGLMM(mode)[1]
      if (coeff>0) {
        r = sqrt(r.square)
      } else {
        r = (-1) * sqrt(r.square)
      }
      p.value = car::Anova(mode, type = 2)[, 3]
      tmp <- c(i, j, coeff, r, r.square, p.value)
      if(is.null(df)){
        df <- tmp  
      }
      else{
        df <- rbind(df, tmp)
      }    
    }
  }
  df <- data.frame(row.names = NULL, df) %>% mutate_at(c(3:6), as.numeric)
  colnames(df) <- c("Fun", "Diversity", "coefficents", "Correlation", "r.square", "Pvalue")
  df$Pvalue <- as.numeric(as.character(df$Pvalue))
  df$AdjPvalue <- rep(0, dim(df)[1])
  df$Correlation <- as.numeric(as.character(df$Correlation))
  #You can adjust the p-values for multiple comparison using Benjamini & Hochberg (1995):
  for(i in unique(df$Diversity)){
      sel <- df$Diversity == i
      df$AdjPvalue[sel] <- p.adjust(df$Pvalue[sel], method = "BH")
      }
  #Now we generate the labels for signifant values
  df$Significance <- cut(df$AdjPvalue, breaks = c(-Inf, 0.001, 0.01, 0.05, Inf), label = c("***", "**", "*", ""))
  df$Diversity <- factor(df$Diversity, ordered = T, levels = rev(colnames(x)))
  df$Fun <- factor(df$Fun, ordered = T, levels = colnames(y))
  return(df)
}

diver_tab <- multi_ave_func_df[ ,c("Plant_richness", "Soil_multidiv", "Ric_16s", "Actinobacteria",	"Alphaproteobacteria",	"Acidobacteria",	"Firmicutes",	"Betaproteobacteria", "Planctomycetes",	"Gammaproteobacteria",	"Bacteroidetes",	"Deltaproteobacteria", "Ric_its", "Ascomycota",	"Basidiomycota",	"Mortierellomycota", "Ric_pro", "Cercozoa",	"Conosa",	"Ciliophora",	"Chlorophyta",	"Apicomplexa",	"Ochrophyta",	"Pseudofungi",	"Lobosa",	"Sagenista",	"Opalozoa",	"Ric_anim", "Nematoda",	"Rotifera",	"Arthropoda",	"Annelida",	"Tardigrada", "Ric_kegg", "Ric_N", "Ric_P", "Ric_S")] %>% scale()

multifun_tab <- multi_ave_func_df[ ,c("meanFunction", "WHC", "BG", "NAG", "LAP", "PPO", "H2O2", "SOC", "NO3_N", "NH4_N", "AP",  "AGB", "plant_pathogen_control", "Soil_respiration")]  %>% scale()


lmm.matrix <- lmm.mat.cal(multifun_tab, diver_tab)
# lmm.matrix$Correlation[lmm.matrix$AdjPvalue >= 0.05] <- 0
```

```{r, fig.align='center', fig.width=8.5, fig.height=8.5}
#plot
heatmap_lmm_plot <- ggplot(aes(x = Fun, y = Diversity, fill = coefficents), data = lmm.matrix)+
  geom_tile() + scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C")+
  geom_text(aes(label = Significance), color="black", size = 3.5)+
  labs(y = NULL, x = 'Multifunctionality', fill = 'Standardized coefficients') +
  scale_y_discrete(position = "right") +
  theme(panel.background = element_rect(fill = 'white', colour ='black'),
        #panel.grid=element_blank(), 
        axis.title = element_text(color = 'black', size = 7),
        axis.ticks.length = unit(0.2, "lines"), axis.ticks = element_line(color = 'black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour = 'black', size = 6),
        axis.text.x = element_text(colour = 'black', size = 6, angle = 45, hjust = 1),
        legend.title = element_text(size = 7),
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.25, "cm"),
        legend.background = element_rect(colour = "white"))
```

### Test the difference in single function between control and collapsed groups using liner models
```{r}
# determine the effect size of the permafrost thawing for the diversity indexes
div_index <- c("Plant_richness", "Soil_multidiv", "Ric_16s", "Actinobacteria",	"Alphaproteobacteria",	"Acidobacteria",	"Firmicutes",	"Betaproteobacteria", "Planctomycetes",	"Gammaproteobacteria",	"Bacteroidetes",	"Deltaproteobacteria", "Ric_its", "Ascomycota",	"Basidiomycota",	"Mortierellomycota", "Ric_pro", "Cercozoa",	"Conosa",	"Ciliophora",	"Chlorophyta",	"Apicomplexa",	"Ochrophyta",	"Pseudofungi",	"Lobosa",	"Sagenista",	"Opalozoa",	"Ric_anim", "Nematoda",	"Rotifera",	"Arthropoda",	"Annelida",	"Tardigrada", "Ric_kegg", "Ric_N", "Ric_P", "Ric_S")

div_scale <- multi_ave_func_df %>% 
  select(c("Group", "Gully_id", div_index)) %>%
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
div_S1 <- sapply(3:ncol(div_scale), function(j) {
    message("Now j=", j, " in ", ncol(div_scale), ". ", date())
    if (length(unique(div_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(div_scale[, j] ~ Group + (1 | Gully_id), data = div_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(div_S1)<-colnames(div_scale)[-c(1:2)]
data.frame(div_S1)
```

```{r, fig.align='center', fig.width=3.5, fig.height=7.8}
rep_str = c("Plant_richness" = "Plant Richness",
            "Plant_shannon" = "Plant Shannon",
            "Soil_moisture" = "Soil moisture",
            "Soil_respiration" = "Soil respiration",
            "H2O2" = expression(paste("H"[2], "O"[2])),
            "NH4_N" = expression(paste("NH"[4]^"+", "-N")),
            "NO3_N" = expression(paste("NO"[3]^"-", "-N")),
            "plant_pathogen_control" = "Plant pathogen control"
)
p.stars <- function(p.values) {
  unclass(symnum(p.values, corr = FALSE, 
                 na = FALSE, cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1),
                 symbols = c("***", "**", "*", ".", " ")))}
single_div_comparison <- div_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = rev(div_index))) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, color = variables)) +
  geom_hline(aes(yintercept = 0), size = 0.7,  colour = "gray2")+
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size")+
  # scale_color_manual(values=c("#D55E00","#446DA9","#446DA9","#446DA9","#446DA9","#D55E00","#D55E00")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "top") +
  annotate("rect", xmin = 0.4, xmax = 4.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#ff6666") +
  annotate("rect", xmin = 4.5, xmax = 10.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 10.5, xmax = 21.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 21.5, xmax = 25.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#5fb236") +
  annotate("rect", xmin = 25.5, xmax = 35.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#1ca9c9") +
  annotate("rect", xmin = 35.5, xmax = 36.5, ymin = -2, ymax = 2, alpha = 0.1, fill = "#0000ff") +
  annotate("rect", xmin = 36.5, xmax = 37.7, ymin = -2, ymax = 2, alpha = 0.1, fill = "#004225") +
  main_theme +
  theme(legend.position = "none",
        strip.background = element_rect(fill = c("#FFF6E1")),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

# if (!dir.exists(file.path(save.dir, "figs/env/"))) {
#   dir.create(file.path(save.dir, "figs/env/"))
# }
# ggsave(file.path(save.dir.multifunc, "./single_div_comparison.pdf"),
#        single_div_comparison, width = 2.7, height = 5, units = "in")
```

```{r, fig.align='center', fig.width=4.5, fig.height=6.5}
### determine the average relative abundance of each phylum within each group
abun_table <- read.table(file = "E:/thermokarst_gully/data/multifun/multifun_abundance_table_all.csv",
             sep = ',',  header = T, row.names = 1, stringsAsFactors = F)
# determine the mean, standard deviations, standard errors for each phylum
taxa_name <- c("Actinobacteria",	"Alphaproteobacteria",	"Acidobacteria",	"Firmicutes",	"Betaproteobacteria", "Planctomycetes",	"Gammaproteobacteria",	"Bacteroidetes",	"Deltaproteobacteria", "Ascomycota",	"Basidiomycota",	"Mortierellomycota", "Cercozoa",	"Conosa",	"Ciliophora",	"Chlorophyta",	"Apicomplexa",	"Ochrophyta",	"Pseudofungi",	"Lobosa",	"Sagenista",	"Opalozoa",	"Nematoda",	"Rotifera",	"Arthropoda",	"Annelida",	"Tardigrada")

phyla_tabin_control <- abun_table %>%
  select(c("Group", taxa_name)) %>%
  mutate_if(is.numeric, ~ . * 100) %>%
  group_by(Group) %>% 
  summarise(across(where(is.numeric), list(mean = ~mean(., na.rm = T), 
                                           sd = ~sd(., na.rm = T), 
                                           n = ~n()))) %>% 
  pivot_longer(cols = -c(Group), 
               names_to = c('Phylum', '.value'), 
               names_sep = '_') %>% 
  mutate(se = sd/sqrt(n)) %>%
  filter(Group == "Control") %>%
  arrange(mean) %>%
  rbind(., data.frame(Group = c(rep("Control", 10)),
                      Phylum = c("Plant_richness", "Soil_multidiv", "Ric_16s", "Ric_its", "Ric_pro", "Ric_anim", "Ric_kegg", "Ric_N", "Ric_P", "Ric_S"), 
                      mean = c(rep(0, 10)), 
                      sd = c(rep(0, 10)), 
                      n = c(rep(30, 10)), 
                      se = c(rep(0, 10)))) %>%
    mutate(Phylum = factor(Phylum, levels = rev(div_index)))
  
  
avg_abun_in_control_plot <- phyla_tabin_control %>%
  ggplot(aes(x = Phylum, y = mean)) +
  geom_bar(position = 'dodge', stat = 'identity', 
           fill= '#FFF6E1', colour = 'black', width = 0.7) +
  geom_errorbar(aes(ymin = mean, ymax = mean + se), width=.2,  position = position_dodge(0.7)) +
  labs(x = NULL, y = 'Mean relative abundance (%)') +
  main_theme + coord_flip() + scale_y_reverse(limits=c(90, 0), expand=c(0, 0)) +
  annotate("rect", xmin = 0.4, xmax = 4.5, ymin = 0, ymax = 90, alpha = 0.1, fill = "#ffffff") +
  annotate("rect", xmin = 4.5, xmax = 10.5, ymin = 0, ymax = 90, alpha = 0.1, fill = "#f19837") +
  annotate("rect", xmin = 10.5, xmax = 21.5, ymin = 0, ymax = 90, alpha = 0.1, fill = "#e56eee") +
  annotate("rect", xmin = 21.5, xmax = 25.5, ymin = 0, ymax = 90, alpha = 0.1, fill = "#5fb236") +
  annotate("rect", xmin = 25.5, xmax = 35.5, ymin = 0, ymax = 90, alpha = 0.1, fill = "#1ca9c9") +
  annotate("rect", xmin = 35.5, xmax = 36.5, ymin = 0, ymax = 90, alpha = 0.1, fill = "#ffffff") +
  annotate("rect", xmin = 36.5, xmax = 37.7, ymin = 0, ymax = 90, alpha = 0.1, fill = "#ffffff") +
  theme(axis.ticks.y = element_blank(),
        axis.text.y = element_blank())
```


```{r, fig.align='center', fig.width=9.8, fig.height=6.5}
library(aplot) # Decorate a 'ggplot' with Associated Information
diff_richness_abun_plot <- avg_abun_in_control_plot %>%
  insert_right(single_div_comparison, width = 1.5) %>%
  insert_right(heatmap_lmm_plot, width = 2)
# ggsave(file.path(save.dir.comp, "./diff_comp_16s_plot.pdf"),
#        diff_comp_16s_plot, width = 4, height = 4.5, units = "in")
diff_richness_abun_plot
```

# Structural Equation Modeling
```{r}
# determine the effect size of the permafrost thawing for the environmental variables
env_index <- c("pH", "Soil_moisture", "Clay.Silt")

env_scale <- metadata %>% 
  select(c("Group", "Gully_id", env_index)) %>%
  mutate(across(where(is.numeric), scale)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  select(where(~ !any(is.na(.))))

# codes for calculating the effect size refer to wu et al. 2022:https://github.com/Linwei-Wu/warming_soil_biodiversity.
env_S1 <- sapply(3:ncol(env_scale), function(j) {
    message("Now j=", j, " in ", ncol(env_scale), ". ", date())
    if (length(unique(env_scale[, j])) < 3) {
        result <- rep(NA, 10)
    } else {
        fm1 <- lmer(env_scale[, j] ~ Group + (1 | Gully_id), data = env_scale)

        presult <- car::Anova(fm1, type = 2)
        coefs <- coef(summary(fm1))[, "Estimate"]  ##four coefs
        names(coefs) <- paste0(names(coefs), ".mean")

        SEvalues <- coef(summary(fm1))[, "Std. Error"]  ##standard errors
        names(SEvalues) <- paste0(names(SEvalues), ".se")

        tvalues <- coef(summary(fm1))[, "t value"]  ##t values
        names(tvalues) <- paste0(names(tvalues), ".t")

        chisqP <- c(presult[, 1], presult[, 3])
        names(chisqP) <- c(paste0(row.names(presult), ".chisq"), paste0(row.names(presult), ".P"))

        result <- c(coefs, tvalues, SEvalues, chisqP)
    }
    result
})
colnames(env_S1)<-colnames(env_scale)[-c(1:2)]
data.frame(env_S1)
```
# PLot 
```{r}
env_comparison <- env_S1 %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column(., "variables") %>%
  mutate(sig = as.vector(unlist(lapply(Group.P, p.stars)))) %>%
  mutate(variables = factor(variables, levels = rev(env_index))) %>%
  ggplot(aes(x = variables, y = GroupCollapsed.mean, color = variables)) +
  geom_hline(aes(yintercept = 0), size = 0.7,  colour = "gray2")+
  geom_point(size = 2) +
  geom_errorbar(aes(ymin = GroupCollapsed.mean - GroupCollapsed.se, 
                    ymax = GroupCollapsed.mean + GroupCollapsed.se), 
                width = 0, position = position_dodge(width = 0.7), cex = 0.9) +
  geom_text(aes(label = sig, x = variables, y = (GroupCollapsed.mean/abs(GroupCollapsed.mean))*(abs(GroupCollapsed.mean) + GroupCollapsed.se)*1.2),
            position = position_dodge(0.1), vjust = 0.55) +
  labs(x = NULL, y = "Effect size")+
  # scale_color_manual(values=c("#D55E00","#446DA9","#446DA9","#446DA9","#446DA9","#D55E00","#D55E00")) +
  scale_y_continuous(expand = c(0, 0), limit = c(-2, 2)) +
  theme_bw() + coord_flip() + scale_x_discrete(position = "top") +
  main_theme +
  theme(legend.position = "none")
env_comparison
```

# PCA
```{r}
#load the package
library(vegan)
library(adespatial)
#set the environmental variables
env_vars <- metadata %>%
  select(c("Soil_moisture", "Clay.Silt"))
  

soil_table <- decostand(env_vars, 'standardize')
soil_dist <- vegdist(soil_table, 'euclidean', upper = F)
decorana(soil_dist)
sol <- rda(substrates_table, trace = F)
sol
soil_PC1 <- scores(sol)$sites[, 1]

sem_df <- data.frame(metadata[, c("Gully_id", "Group")], Soil_PC1 = soil_PC1, 
                     multi_ave_func_df[, c("Plant_richness", "Soil_multidiv", "Ric_kegg", "meanFunction")])
```



```{r}
#load the package
library(piecewiseSEM)
library(lme4)
library(lmerTest)
library(nlme)
#Model
gully_sem <- psem(
  lme(Soil_PC1 ~ Group, random = ~1|Gully_id, data = sem_df, method = "ML"),
  lme(Plant_richness ~ Group + Soil_PC1 + Soil_multidiv, random = ~1|Gully_id, data = sem_df, method = "ML"), 
  lme(Soil_multidiv ~ Group + Soil_PC1 + Plant_richness + Ric_kegg, random = ~1|Gully_id, data = sem_df, method = "ML"),
  lme(Ric_kegg ~ Group + Soil_PC1 + Plant_richness + Soil_multidiv, random = ~1|Gully_id, data = sem_df, method = "ML"),
  lme(meanFunction ~ Group + Soil_PC1 + Plant_richness + Soil_multidiv + Ric_kegg, random = ~1|Gully_id, data = sem_df, method = "ML"),
  data = sem_df
)


summary(gully_sem)


sem_mod <- psem(
  lm(LCBD ~ MAP + FrI + Salinity +ALT, data = env_div_agg),
  lm(ALT ~ MAP, data = env_div_agg),
  lm(FrI ~ MAP + ALT, data = env_div_agg),
  lm(Salinity ~ MAP + FrI, data = env_div_agg))
summary(sem_mod)
sem_mod <- psem(
  lm(LCBD ~ MAP + FrI + Salinity +ALT, data = env_div_agg),
  lm(FrI ~ MAP + ALT, data = env_div_agg),
  lm(Salinity ~ MAP + FrI, data = env_div_agg))
summary(sem_mod)
sem_mod <- psem(
  lm(LCBD ~ MAP + FrI + Salinity, data = env_div_agg),
  lm(FrI ~ MAP + ALT, data = env_div_agg),
  lm(Salinity ~ MAP + FrI, data = env_div_agg))
summary(sem_mod)
sem_mod <- psem(
  lm(LCBD ~ MAP + FrI, data = env_div_agg),
  lm(FrI ~ MAP + ALT, data = env_div_agg),
  lm(Salinity ~ MAP + FrI, data = env_div_agg))
summary(sem_mod)
sem_mod <- psem(
  lm(LCBD ~ MAP + FrI, data = env_div_agg),
  lm(FrI ~ MAP, data = env_div_agg),
  lm(Salinity ~ MAP + FrI, data = env_div_agg))
summary(sem_mod)
sem_mod <- psem(
  lm(LCBD ~ MAP + FrI, data = env_div_agg),
  lm(FrI ~ MAP, data = env_div_agg),
  lm(Salinity ~ MAP, data = env_div_agg))
summary(sem_mod)
fisherC(sem_mod)
rsquared(sem_mod)

coeffs_sem <- coefs(sem_mod, standardize = "scale")[,c(1,2,8)]
coeffs_sem
## Direct effect of climate variables on total community composition.
clim_dir_effect <- coeffs_sem[1,3]
## Ind effect of climate variables on total community composition.
clim_ind_effect <- coeffs_sem[3,3]*coeffs_sem[2,3]

total_clim_effect <- clim_dir_effect + clim_ind_effect

## Direct effect of ALT on total community composition.
ALT_dir_effect <- 0
## Ind effect of ALT on total community composition.
ALT_ind_effect <- 0
total_ALT_effect <- ALT_dir_effect + ALT_ind_effect

## Direct effect of DOM variables on total community composition.
DOM_dir_effect <- coeffs_sem[2,3]
## Ind effect of DOM variables on total community composition.
DOM_ind_effect <- 0
total_DOM_effect <- DOM_dir_effect + DOM_ind_effect

## Direct effect of physio_chemical variables on total community composition.
physio_chemical_dir_effect <- 0
## Ind effect of climate variables on total community composition.
physio_chemical_ind_effect <- 0
total_physio_chemical_effect <- physio_chemical_dir_effect + physio_chemical_ind_effect
## bar plot
sem_effect <- data.frame(variable = c(rep('Climate elements',3),
                                      rep('ALT', 3),
                                      rep('DOM properties', 3),
                                      rep('physio-chemical factors', 3)),
                         type_of_effects = c(rep(c('Total effects', 'Direct effects', 
                                                   'Indirect effects'), 4)),
                         value = c(total_clim_effect, clim_dir_effect, clim_ind_effect,
                                   total_ALT_effect, ALT_dir_effect, ALT_ind_effect,
                                   total_DOM_effect, DOM_dir_effect, DOM_ind_effect,
                                   total_physio_chemical_effect, physio_chemical_dir_effect,
                                   physio_chemical_ind_effect))
sem_effect$type_of_effects <- factor(sem_effect$type_of_effects, ordered = T,
                                     levels = c('Total effects', 'Direct effects', 'Indirect effects'))
sem_effect$variable <- factor(sem_effect$variable, ordered = T,
                              levels = c('Climate elements', 'DOM properties', 'physio-chemical factors', 'ALT'))
sem_effect_plot <-  ggplot(data = sem_effect, 
                           aes(x = variable, y = value, fill = variable)) +
  geom_bar(stat="identity", width = 0.6) +
  geom_hline(yintercept = 0) +
  scale_y_continuous(limits = c(-0.8, 0.5)) +
  labs(x = 'Variables', y = 'Standardized effects from SEM') +
  facet_wrap(~ type_of_effects) +
  theme_bw() +
  theme(legend.position = 'none',
        axis.title = element_text(color='black',size=14),
        strip.text = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.text.y = element_text(colour='black',size=12),
        axis.text.x = element_text(colour='black', size = 12, angle = 45, hjust = 1),
        legend.title=element_text(size = 12),
        legend.text=element_text(size=9),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "white"))

sem_effect_plot
```


```{r}
bowen <- read.csv("https://raw.githubusercontent.com/jslefche/sem_book/master/data/bowen.csv")

bowen <- na.omit(bowen)

library(nlme)

bowen_sem <- psem(
  lme(observed_otus ~ status, random = ~1|Genotype, data = bowen, method = "ML"),
  lme(RNA.DNA ~ status + observed_otus, random = ~1|Genotype, data = bowen, method = "ML"), 
  lme(below.C ~ observed_otus + status, random = ~1|Genotype, data = bowen, method = "ML"), 
  lme(abovebiomass_g ~ RNA.DNA + observed_otus + belowCN + status, random = ~1|Genotype, data = bowen, method = "ML"),
  data = bowen
)
```


```{r}
library(nlme)
library(lme4)

shipley_psem <- psem(

  lme(DD ~ lat, random = ~ 1 | site / tree, na.action = na.omit,
  data = shipley),

  lme(Date ~ DD, random = ~ 1 | site / tree, na.action = na.omit,
  data = shipley),

  lme(Growth ~ Date, random = ~ 1 | site / tree, na.action = na.omit,
  data = shipley),

  glmer(Live ~ Growth + (1 | site) + (1 | tree),
  family = binomial(link = "logit"), data = shipley)

  )

summary(shipley_psem, .progressBar = FALSE)
```



