---
title: "Amplicon_analysis"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 150)
```
## Data input 
Set work directory
```{r}
setwd('e:/thermokarst_gully/')
wd_16s <- file.path(getwd(),"data/16S/rdp")
# if (!dir.exists(wd_16s)) {
#   dir.create(wd_16s)
# }
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```
Loading packages
```{r}
pacman::p_load(phyloseq, ape, vegan, Biostrings, microbiome, tidyverse)
```
Data input
```{r}
source("script/read_data_rdp.R")
```

##  Network analysis
```{r}
# loading packages
pacman::p_load(microeco, file2meco, igraph, WGCNA, rgexf)
# construct the phyloseq projects for control and collapsed groups
ctrl_phylo <- subset_samples(phylo, Group == "Control")
ctrl_phylo <- prune_taxa(taxa_sums(ctrl_phylo) > 0, ctrl_phylo) 
collap_phylo <- subset_samples(phylo, Group == "Collapsed")
collap_phylo <- prune_taxa(taxa_sums(collap_phylo) > 0, collap_phylo)

# transform the phyloseq objects to microeco objects
meco_ctrl_obj <- phyloseq2meco(ctrl_phylo)
meco_collap_obj <- phyloseq2meco(collap_phylo)

# determine the adjacency matrixes
t1 <- trans_network$new(dataset = meco_ctrl_obj, cal_cor = "WGCNA", taxa_level = "OTU", filter_thres = 0.0001, cor_method = "spearman")
t2 <- trans_network$new(dataset = meco_collap_obj, cal_cor = "WGCNA", taxa_level = "OTU", filter_thres = 0.0001, cor_method = "spearman")

# use arbitrary coefficient threshold to contruct network
t1$cal_network(COR_p_thres = 0.01, COR_cut = 0.7)
t2$cal_network(COR_p_thres = 0.01, COR_cut = 0.7)

# invoke igraph cluster_fast_greedy function for this undirected network 
t1$cal_module(method = "cluster_fast_greedy")
t2$cal_module(method = "cluster_fast_greedy")

# require rgexf package to be installed
t1$save_network(filepath = file.path(save.dir, "./figs/ctrl_network.gexf"))
t2$save_network(filepath = file.path(save.dir, "./figs/collapsed_network.gexf"))

# calculate network attributes
t1$cal_network_attr()
t2$res_network_attr()
```


```{r, fig.align='center', fig.width=6.5, fig.height=4.5}
# get node properties
t1$get_node_table(node_roles = TRUE)
t2$get_node_table(node_roles = TRUE)
# return t1$res_node_table

# get edge properties
t1$get_edge_table()
t2$get_edge_table()
# return t1$res_edge_table 
t1$get_adjacency_matrix()
t2$get_adjacency_matrix()
# return t1$res_adjacency_matrix

# add_label = TRUE can be used to directly add text label for points
t1$plot_taxa_roles(use_type = 1)
t2$plot_taxa_roles(use_type = 1)
```

```{r, fig.align='center', fig.width=6.5, fig.height=5}
# plot node roles with phylum information
t1$plot_taxa_roles(use_type = 2)
t2$plot_taxa_roles(use_type = 2)
```
### Network robustness test
```{r}
# Network attack analysis
# Libraries
pacman::p_load(SpiecEasi, reshape2, phyloseq, igraph)
# get the igraph objects
ig.ctrl <- t1$res_network
ig.collapsed <- t2$res_network
# calculating the natural connectivity from adjacency matrix
ncc <- function(ig) {
  evals <- eigen(ig)$value
  nc <- log(mean(exp(evals)))
}

# calculating the natural connectivity from adjacency matrix of a graph
natcon <- function(ig) {
  adj <- get.adjacency(ig)
  evals <- eigen(adj)$value
  nc <- log(mean(exp(evals)))
}

# targeted attack ordered by betweenness
nc.attackbetweenness <- function(ig) {
  hubord <- order(rank(betweenness(ig)), decreasing=TRUE)
  sapply(1:round(vcount(ig)*.8), function(i) {
    ind <- hubord[1:i]
    tmp <- delete_vertices(ig, V(ig)$name[ind])
    natcon(tmp)
  })
}

# Targeted attack ordered by node degree.
nc.attackdegree <- function(ig) {
  hubord <- order(rank(degree(ig)), decreasing=TRUE)
  sapply(1:round(vcount(ig)*.8), function(i) {
    ind <- hubord[1:i]
    tmp <- delete_vertices(ig, V(ig)$name[ind])
    natcon(tmp)
  })
}


# Node removals
attack<-function (adj.mat, node.sup)
{
  n.nodes <- dim(adj.mat)[1]
  adj.mat[node.sup, ] <- rep(0, n.nodes)
  adj.mat[, node.sup] <- rep(0, n.nodes)
  nc<-ncc(adj.mat)
  list(new.mat = adj.mat, nc=nc)
}


ctrl.nc.deg.rmt <- nc.attackdegree(ig.ctrl)
ctrl.nc.bet.rmt <- nc.attackbetweenness(ig.ctrl)

collapsed.nc.deg.rmt <- nc.attackdegree(ig.collapsed)
collapsed.nc.bet.rmt <- nc.attackbetweenness(ig.collapsed)

ctrl.nc.deg.rmt <- cbind(rm_pro = c(1:length(ctrl.nc.deg.rmt))/vcount(ig.ctrl), nat.connet = ctrl.nc.deg.rmt)
ctrl.nc.bet.rmt <- cbind(rm_pro = c(1:length(ctrl.nc.bet.rmt))/vcount(ig.ctrl), nat.connet = ctrl.nc.bet.rmt)

collapsed.nc.deg.rmt <- cbind(rm_pro = c(1:length(collapsed.nc.deg.rmt))/vcount(ig.collapsed), nat.connet = collapsed.nc.deg.rmt)
collapsed.nc.bet.rmt <- cbind(rm_pro = c(1:length(collapsed.nc.bet.rmt))/vcount(ig.collapsed), nat.connet = collapsed.nc.bet.rmt)


write.table(ctrl.nc.deg.rmt, file.path(save.dir, "./tables/ctrl-deg-att-rmt.txt"),
            sep="\t", quote=F, row.names = T, col.names = T)
write.table(ctrl.nc.bet.rmt, file.path(save.dir, "./tables/ctrl-bet-att-rmt.txt"),
            sep="\t", quote=F, row.names = T, col.names = T)
write.table(collapsed.nc.deg.rmt, file.path(save.dir, "./tables/collapsed-deg-att-rmt.txt"), 
            sep="\t", quote=F, row.names = T, col.names = T)
write.table(collapsed.nc.bet.rmt, file.path(save.dir, "./tables/collapsed-bet-att-rmt.txt"),
            sep="\t", quote=F, row.names = T, col.names = T)
```


```{r, fig.align='center', fig.width=6, fig.height=3.5}
robut.df.rmt <- data.frame(rbind(cbind(Group = rep('Control', sum(nrow(ctrl.nc.deg.rmt), nrow(ctrl.nc.bet.rmt))),
                                       type = c(rep('Proportion of removed nodes', nrow(ctrl.nc.deg.rmt)), 
                                                rep('Proportion of removed betweenness', nrow(ctrl.nc.bet.rmt))),
                                       nat.connet = rbind(ctrl.nc.deg.rmt, ctrl.nc.bet.rmt)),
                                 cbind(Group = rep('Collapsed', sum(nrow(collapsed.nc.deg.rmt), nrow(collapsed.nc.bet.rmt))),
                                       type = c(rep('Proportion of removed nodes', nrow(collapsed.nc.deg.rmt)), 
                                                rep('Proportion of removed betweenness', nrow(collapsed.nc.bet.rmt))),
                                       nat.connet = rbind(collapsed.nc.deg.rmt, collapsed.nc.bet.rmt))))
robut.df.rmt$Group <- factor(robut.df.rmt$Group, ordered = T, 
                              levels = c('Control', 'Collapsed'))
robut.df.rmt$type <- factor(robut.df.rmt$type, ordered = T, 
                            levels = c('Proportion of removed nodes', 'Proportion of removed betweenness'))

robut.df.rmt$rm_pro <- as.numeric(robut.df.rmt$rm_pro)
robut.df.rmt$nat.connet <- as.numeric(robut.df.rmt$nat.connet)

robutness_test_plot_rmt <- ggplot(robut.df.rmt, aes(x = rm_pro, y = nat.connet, colour = Group))+
  geom_point() +
  scale_color_manual(values= c('#1b9e77', '#d95f02')) +
  labs(x = NULL, y = 'Natural connectivity') +
  facet_wrap(~type, scales = 'free') +
  theme_bw() +
  theme(strip.text = element_text(size = 10),
        axis.line = element_line(colour = "black"),
        axis.title = element_text(size = 12, colour = 'black'),
        axis.text = element_text(size = 10, colour = 'black'),
        axis.text.x = element_text(colour='black', size = 10, 
                                   angle = 45, hjust = 1),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 10),
        legend.position = c(0.9, 0.8),
        legend.background = element_blank(),
        panel.grid = element_blank())
robutness_test_plot_rmt
```
