---
title: "gene analysis within each site"
author: "Luyao Kang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = F}
knitr::opts_chunk$set(
  collapse = T, echo = T, comment = "#>", message = F, warning = F,
	fig.align = "center", fig.width= 4.5, fig.height = 3, dpi = 300)
```

## Data input
Set work directory
```{r}
setwd('e:/thermokarst_gully/')
wd_16s <- file.path(getwd(),"data/16S/rdp")
wd_fun <- file.path(getwd(),"data/metagenome")
save.dir <- file.path(getwd(),"result")
```

Loading packages
```{r}
pacman::p_load(tidyverse, ggrepel, EnhancedVolcano, edgeR, DESeq2, ggpubr)
```

Data input
```{r}
source("script/read_data_rdp.R")
```

```{r}
alpha_div <- estimate_richness(phylo_rare, measures = c("Observed", "Chao1", 'Shannon', 'Simpson'))
library(picante)
pd <- pd(t(otu), tree, include.root = F) # estimate the phylogenetic diversity
tax_dist <-vegdist(t(otu), "bray" )
mod1<-betadisper(tax_dist,metadata$Group)
div_table <- cbind(metadata[, c('Gully_id', 'Group')], alpha_div, Faith = pd$PD, Distance = mod1$distances, Evenness = alpha_div$Shannon/log(alpha_div$Observed)) %>%
  mutate(Group = factor(Group, levels = c('Control', 'Collapsed'))) %>%
  mutate(Gully_id = factor(Gully_id, levels = c('EB', 'ML', 'RS', 'SLH', 'HSX', 'HH')))
```

Box plot for alpha index using ggplot2

```{r, fig.align='center', fig.width=7.5, fig.height=6.2}
library(ggpubr)
library(ggplot2)
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 9),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 9),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 8),
        axis.text.x = element_text(colour = 'black', size = 8),
        legend.position = "none")
# box plot
my_comparisons <- list(c('Control', 'Collapsed'))
div_gully_plot <- div_table %>% 
  select(c("Gully_id", "Group", "Chao1", "Evenness", "Shannon", "Faith", "Distance")) %>%
  gather(diversity, values, -c("Gully_id", "Group")) %>%
  mutate(diversity = factor(diversity, levels = c("Chao1", "Evenness", "Shannon", "Faith", "Distance"))) %>%
  ggplot(aes(x = Group, y = values)) + 
  geom_boxplot(width = 0.5, aes(fill = Group), outlier.shape = NA) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.15))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.5,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = 'Group', y = NULL, fill= 'Group') +
  facet_grid(diversity ~ Gully_id, scales = "free", space = "free_x") +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  main_theme +
  theme(panel.spacing = unit(0, "lines"),
        axis.text.x = element_text(angle = 45, hjust = 1))
ggsave(file.path(save.dir, "./group_site_result/div_gully_plot.pdf"),
       div_gully_plot, width = 7.5, height = 6.2, units = "in")
div_gully_plot
```

Test the differnece in genes
```{r}
## Arrange the log fold change table
sel_ko <- read_csv(file.path(wd_fun, 'CN_ko_input.csv'), col_names = T)

C_N_ko_count_tab <- ko_count_table[rownames(ko_count_table) %in% sel_ko$KO, ]
C_N_ko_count_tab <- cbind(KO = rownames(C_N_ko_count_tab), C_N_ko_count_tab)

C_N_ko_tpm_tab <- ko_tpm_table[rownames(ko_tpm_table) %in% sel_ko$KO, ]
C_N_ko_tpm_tab <- cbind(KO = rownames(C_N_ko_tpm_tab), C_N_ko_tpm_tab)
```

## heatmap
```{r, fig.align='center', fig.width=8, fig.height=12}
C_N_path1 <- C_N_ko_tpm_tab %>% 
  inner_join(sel_ko, c("KO" = "KO")) %>% 
  select(-c(1)) %>%
  group_by(Category, Subcategory, Enzyme_protein_encoded) %>%
  summarise(across(everything(), sum)) %>% data.frame() %>%
  mutate(Subcategory = factor(Subcategory, levels = unique(sel_ko$Subcategory), ordered = T))
C_N_path2 <- C_N_path1 %>% select(c(3:63)) %>% 
  tibble::column_to_rownames('Enzyme_protein_encoded')
```

## edgeR analysis
```{r}
edgeR_fun <- function(gully_id) {
  # loading package
  library("edgeR")
  # Load Expression Data
  RNASeq <- C_N_path2[, meta_dat %>% as_tibble() %>% filter(Gully_id == gully_id) %>% pull(Sample_name)]
  # Load sub-type information
  classDefinitions_RNASeq <- meta_dat %>% as_tibble() %>% filter(Gully_id == gully_id)
  # Filter Data (RNA-seq read counts are converted to CPM values 
  # and genes with CPM > 1 in at least 50 of the samples are 
  # retained for further study, a gene mush have at least 50 measurements 
  # with more than 1 CPM in one of the classes to be included in the analysis)
  cpms <- cpm(RNASeq)
  keep <- rowSums(cpms > 1) >= 1
  counts <- RNASeq[keep,]
  # Normalization and Dispersion
  # create data structure to hold counts and subtype information for each sample.
  d <- DGEList(counts = counts, group = classDefinitions_RNASeq$Group)
  # Normalize the data
  d <- calcNormFactors(d)
  # calculate dispersion
  d <- estimateCommonDisp(d)
  d <- estimateTagwiseDisp(d)
  # Compare Collapsed to the Control group.
  classes <- factor(classDefinitions_RNASeq$Group)
  modelDesign <- model.matrix(~ 0 + classes)
  contrast_collap <- makeContrasts(
  collavsrest = "classesCollapsed-classesControl", levels = modelDesign)
  fit_glm <- glmFit(d, modelDesign)
  collavsrest <- glmLRT(fit_glm , contrast = contrast_collap)
  tt_collavsrest <- topTags(collavsrest, n = nrow(d))
  logFC_table <- tt_collavsrest$table %>%
    rownames_to_column(var = "pathway") %>%
    mutate(Expression = case_when(logFC >= log(2) & FDR <= 0.05 ~ "Up-regulated",
                           logFC <= -log(2) & FDR <= 0.05 ~ "Down-regulated",
                           TRUE ~ "Unchanged")) %>%
    mutate(sig = case_when(PValue <= 0.001 ~ "***",
                         PValue > 0.001 & PValue <= 0.01 ~ "**",
                         PValue > 0.01 & PValue <= 0.05 ~ "*"))
 return(logFC_table) 
}
```

```{r}
EB_logFC_table <- edgeR_fun("EB")
ML_logFC_table <- edgeR_fun("ML")
RS_logFC_table <- edgeR_fun("RS")
SLH_logFC_table <- edgeR_fun("SLH")
HSX_logFC_table <- edgeR_fun("HSX")
HH_logFC_table <- edgeR_fun("HH")

EB_logFC_table <- data.frame(EB_logFC_table, Gully_id = rep("EB", nrow(EB_logFC_table)))
ML_logFC_table <- data.frame(ML_logFC_table, Gully_id = rep("ML", nrow(ML_logFC_table)))
RS_logFC_table <- data.frame(RS_logFC_table, Gully_id = rep("RS", nrow(RS_logFC_table)))
SLH_logFC_table <- data.frame(SLH_logFC_table, Gully_id = rep("SLH", nrow(SLH_logFC_table)))
HSX_logFC_table <- data.frame(HSX_logFC_table, Gully_id = rep("HSX", nrow(HSX_logFC_table)))
HH_logFC_table <- data.frame(HH_logFC_table, Gully_id = rep("HH", nrow(HH_logFC_table)))

logFC_table <- rbind(EB_logFC_table, ML_logFC_table, RS_logFC_table, 
                     SLH_logFC_table, HSX_logFC_table, HH_logFC_table)
```

## arrange the log fold change table
```{r}
C_names <- c("Alpha-amylase", "Glucoamylase", "Pullulanase", "Isopullulanase",
             "Arabinofuranosidase", "Beta_mannanase", "Xylanase", "Xylose isomerase",
             "Beta-glucosidase",  "Endoglucanase", "Exoglucanase",
             "Acetylglucosaminidase", "Endochitinase", "Exochitinase",
             "Pectinase", 
             "Aryl-aldehyde oxidase", "Isocitrate lyase", "Limonene-1, 2-epoxide hydrolase", "Malate synthase", "Vanillate demethylase", 
             "Glyoxal oxidase", "Phenol oxidase (tyrosinase)",
             "Methyl coenzyme M reductase", "Particulate methane monooxygenase", "Soluable methane monooxygenase",
             "Carbon monoxide dehydrogenase", "Tetrahydrofolate formylase", "ATP citrate lyase", 
             "Propionyl-CoA carboxylase", "RuBisCo")
N_names <- c("amoA",	"amoB",	"amoC",	"hao",	"nxrA",	"nxrB", 
             "narG",	"narH",	"narI", "napA",	"napB", "nrfA",	"nrfH", "nirB",	"nirD",
             "nasA",	"nasB", "narB",	"NR", "NIT-6", "nirA",
             "nirK", "nirS",	"norB",	"norC",	"nosZ",
             	"nifD", "nifK", "nifH", 
             "nrtA",	"nrtB", "nrtC",	"nrtD", 
             "nmo", "gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC", "glnA")
S_names <- c("sat", "cysC", "cysD", "cysNC", "csyH", "cysJ", "cysN", "aprA", "aprB", "dsrA", "dsrB",
             "asrA", "asrB", "asrC", "sir", "sor", "sreA", "sreB", "sreC", "hydA", "hydD", "hydB",
             "hydG", "psrA", "psrB", "psrC", "sqr", "fccA", "fccB", "soxD", "soxX", "soxA", "soxB",
             "soxC", "soxY", "soxZ", "ttrA", "ttrB", "ttrC", "phsA", "phsB", "phsC")
Other_names <- c("Cyc1", "Cyc2", "MtrA", "MtrB", "MtrC", "arsC (grx)", "arsC (trx)", 
                 "aioA", "arsM", "ygfM", "xdhD", "YgfK")
```



```{r, fig.align='center', fig.width=8.9, fig.height=4.5}
C_enrich_heatmap <- logFC_table %>% filter(pathway %in% C_names) %>%
  mutate(pathway = factor(pathway, levels = C_names, ordered = T)) %>% 
  mutate(Category = rep("Carbon", nrow(.))) %>%
  mutate(Gully_id = factor(Gully_id, levels = c("EB", "ML", "RS", "SLH", "HSX", "HH"))) %>%
  ggplot(aes(x = pathway, y = Category, fill = logFC)) +
  geom_tile(height = 1, width = 1, size = 1, color = "black") + 
  scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C") +  # low="#2C7BB6", mid="white", high="#D7191C" or low = "#009E73", mid = "white", high = "#E69F00"
  geom_text(aes(label = sig), 
            color = "black", size = 2.5) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  labs(y = NULL, x = NULL, fill = expression(paste("log"[2], "FC"))) +
  facet_grid(Gully_id ~ .) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 8, colour = 'black'),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.text = element_text(size = 7),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        # legend.key.height= unit(0.35, 'cm'),
        # legend.key.width= unit(0.35, 'cm'),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6)) +
  coord_fixed(ratio=1)
ggsave(file.path(save.dir, "./group_site_result/C_enrich_heatmap.pdf"),
       C_enrich_heatmap, width = 8.9, height = 4.5, units = "in")
C_enrich_heatmap
```

```{r, fig.align='center', fig.width=8.9, fig.height=3.5}
N_enrich_heatmap <- logFC_table %>% filter(pathway %in% N_names) %>%
  mutate(pathway = factor(pathway, levels = N_names, ordered = T)) %>% 
  mutate(Category = rep("Nitrogen", nrow(.))) %>%
  mutate(Gully_id = factor(Gully_id, levels = c("EB", "ML", "RS", "SLH", "HSX", "HH"))) %>%
  ggplot(aes(x = pathway, y = Category, fill = logFC)) +
  geom_tile(height = 1, width = 1, size = 1, color = "black") + 
  scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C") +  # low="#2C7BB6", mid="white", high="#D7191C" or low = "#009E73", mid = "white", high = "#E69F00"
  geom_text(aes(label = sig), 
            color = "black", size = 2.5) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  labs(y = NULL, x = NULL, fill = expression(paste("log"[2], "FC"))) +
  facet_grid(Gully_id ~ .) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 8, colour = 'black'),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.text = element_text(size = 7),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        # legend.key.height= unit(0.35, 'cm'),
        # legend.key.width= unit(0.35, 'cm'),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6)) +
  coord_fixed(ratio=1)
ggsave(file.path(save.dir, "./group_site_result/N_enrich_heatmap1.pdf"),
       N_enrich_heatmap, width = 8.9, height = 3.5, units = "in")
N_enrich_heatmap
```

```{r, fig.align='center', fig.width=8.9, fig.height=3.5}
S_enrich_heatmap <- logFC_table %>% filter(pathway %in% S_names) %>%
  mutate(pathway = factor(pathway, levels = S_names, ordered = T)) %>% 
  mutate(Category = rep("Sulfur", nrow(.))) %>%
  mutate(Gully_id = factor(Gully_id, levels = c("EB", "ML", "RS", "SLH", "HSX", "HH"))) %>%
  ggplot(aes(x = pathway, y = Category, fill = logFC)) +
  geom_tile(height = 1, width = 1, size = 1, color = "black") + 
  scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C") +  # low="#2C7BB6", mid="white", high="#D7191C" or low = "#009E73", mid = "white", high = "#E69F00"
  geom_text(aes(label = sig), 
            color = "black", size = 2.5) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  labs(y = NULL, x = NULL, fill = expression(paste("log"[2], "FC"))) +
  facet_grid(Gully_id ~ .) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 8, colour = 'black'),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.text = element_text(size = 7),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        # legend.key.height= unit(0.35, 'cm'),
        # legend.key.width= unit(0.35, 'cm'),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6)) +
  coord_fixed(ratio=1)
ggsave(file.path(save.dir, "./group_site_result/S_enrich_heatmap.pdf"),
       S_enrich_heatmap, width = 8.9, height = 3.5, units = "in")
S_enrich_heatmap
```

```{r, fig.align='center', fig.width=6, fig.height=4.5}
Others_enrich_heatmap <- logFC_table %>% filter(pathway %in% Other_names) %>%
  mutate(pathway = factor(pathway, levels = Other_names, ordered = T)) %>% 
  mutate(Category = rep("Others", nrow(.))) %>%
  mutate(Gully_id = factor(Gully_id, levels = c("EB", "ML", "RS", "SLH", "HSX", "HH"))) %>%
  ggplot(aes(x = pathway, y = Category, fill = logFC)) +
  geom_tile(height = 1, width = 1, size = 1, color = "black") + 
  scale_fill_gradient2(low = "#2C7BB6", mid = "white", high = "#D7191C") +  # low="#2C7BB6", mid="white", high="#D7191C" or low = "#009E73", mid = "white", high = "#E69F00"
  geom_text(aes(label = sig), 
            color = "black", size = 2.5) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  labs(y = NULL, x = NULL, fill = expression(paste("log"[2], "FC"))) +
  facet_grid(Gully_id ~ .) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 8, colour = 'black'),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        strip.text = element_text(size = 7),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        # legend.key.height= unit(0.35, 'cm'),
        # legend.key.width= unit(0.35, 'cm'),
        legend.key.size = unit(0.35, 'cm'),
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6)) +
  coord_fixed(ratio=1)
ggsave(file.path(save.dir, "./group_site_result/Others_enrich_heatmap.pdf"),
       Others_enrich_heatmap, width = 4.5, height = 3.5, units = "in")
Others_enrich_heatmap
```











###################################
## Wilcoxon Test
Select the pathways or genes that we focus
```{r}
C_names <- c("Alpha-amylase", "Glucoamylase", "Pullulanase", "Isopullulanase",
             "Arabinofuranosidase", "Beta_mannanase", "Xylanase", "Xylose isomerase",
             "Beta-glucosidase",  "Endoglucanase", "Exoglucanase",
             "Acetylglucosaminidase", "Endochitinase", "Exochitinase",
             "Pectinase", 
             "Aryl-aldehyde oxidase", "Isocitrate lyase", "Limonene-1, 2-epoxide hydrolase", "Malate synthase", "Vanillate demethylase", 
             "Glyoxal oxidase", "Phenol oxidase (tyrosinase)",
             "Methyl coenzyme M reductase", "Particulate methane monooxygenase", "Soluable methane monooxygenase",
             "Carbon monoxide dehydrogenase", "Tetrahydrofolate formylase", "ATP citrate lyase", 
             "Propionyl-CoA carboxylase", "RuBisCo")
N_names <- c("amoA",	"amoB",	"amoC",	"hao",	"nxrA",	"nxrB", "narG",	"narH",	"narI", "napA",	
             "napB", "nirK", "nirS",	"norB",	"norC",	"nosZ", "nrfA",	"nrfH", "nirB",	"nirD",	
             "nasA",	"nasB", "narB",	"NR", "NIT-6", "nirA",	"nifD", "nifK", "nifH", "nrtA",	"nrtB", "nrtC",	
             "nrtD", "nmo", "gdh_K00261", "gdh_K00262", "gdh_K15371", "glsA", "ureA", "ureC", "glnA")
S_names <- c("sat", "cysC", "cysD", "cysNC", "csyH", "cysJ", "cysN", "aprA", "aprB", "dsrA", "dsrB",
             "asrA", "asrB", "asrC", "sir", "sor", "sreA", "sreB", "sreC", "hydA", "hydD", "hydB",
             "hydG", "psrA", "psrB", "psrC", "sqr", "fccA", "fccB", "soxD", "soxX", "soxA", "soxB",
             "soxC", "soxY", "soxZ", "ttrA", "ttrB", "ttrC", "phsA", "phsB", "phsC")
Other_names <- c("Cyc1", "Cyc2", "MtrA", "MtrB", "MtrC", "arsC (grx)", "arsC (trx)", 
                 "aioA", "arsM", "ygfM", "xdhD", "YgfK")
```

## Test the difference in genes involved in carbon degradation
```{r, fig.align='center', fig.width=8, fig.height=4.7}
carbon_dat <- C_N_path1 %>%
  filter(Category == "Carbon") %>%
  select(-c(Category, Subcategory)) %>%
  pivot_longer(cols = -Enzyme_protein_encoded, names_to = "Sample", values_to = "Value") %>%
  mutate(Group = case_when(grepl("_C", Sample) ~ "Control", 
                           grepl("_T", Sample) ~ "Collapsed")) %>%
  mutate(Gully_id = case_when(grepl("G1_", Sample) ~ "EB", 
                              grepl("G2_", Sample) ~ "ML",
                              grepl("G3_", Sample) ~ "RS",
                              grepl("G4_", Sample) ~ "SLH",
                              grepl("G5_", Sample) ~ "HSX",
                              grepl("G6_", Sample) ~ "HH")) %>%
  dplyr::rename(pathway = Enzyme_protein_encoded)

# box plot
library(ggplot2)
library(ggpubr)
main_theme = theme_bw() + 
  theme(panel.grid = element_blank(),
        panel.border = element_rect(size = 0.5),
        strip.text = element_text(colour = 'black', size = 12),
        strip.background = element_rect(colour = 'black', fill = 'grey'),
        axis.title = element_text(color = 'black',size = 12),
        axis.ticks = element_line(color = "black", linewidth = 0.5),
        axis.text.y = element_text(colour = 'black', size = 10),
        axis.text.x = element_text(colour = 'black', size = 10),
        legend.position = "none")

my_comparisons <- list(c('Control', 'Collapsed'))

EB_C_enrich <- carbon_dat %>% filter(Gully_id == "EB") %>%
  mutate(pathway = factor(pathway, levels = C_names, ordered = T)) %>%
  mutate(Group = factor(Group, levels = c("Control", "Collapsed"))) %>%
  ggplot(aes(x = pathway, y = Value))+
  geom_boxplot(width = 0.5, aes(fill = Group), outlier.shape = NA) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.15))) +
  stat_compare_means(comparisons = my_comparisons, paired = F,
                     p.adjust.method = "BH", label = "p.signif", bracket.size = 0.5,
                     size = 3.5, tip.length = 0.00, method = "wilcox.test") +
  labs(x = 'Pathway', y = NULL, fill= 'Group') +
  scale_fill_manual(values = c("#79ceb8", "#e95f5c", "#5cc3e8", "#ffdb00")) +
  # scale_y_break(c(10,45),
  #               scales=1.2, 
  #               ticklabels=c(50,100),
  #               space=0.2)+
  # scale_y_break(c(100,400),
  #               scales=2.5, 
  #               ticklabels=c(600,800,1200),
  #               space=0.2) +
  main_theme +
  theme(panel.spacing = unit(0, "lines"),
        axis.text.x = element_text(angle = 45, hjust = 1))
EB_C_enrich
```